// checkly/dist/formatters/alert-channels.js
import chalk from 'chalk';
import { renderDetailFields, renderAdaptiveTable, renderCommandHints, formatDate, truncateError, } from './render.js';
function boolLabel(value, format) {
    if (value === undefined)
        return format === 'terminal' ? chalk.dim('-') : '-';
    if (format === 'md')
        return value ? 'yes' : 'no';
    return value ? chalk.green('yes') : chalk.dim('no');
}
function normalizeType(type) {
    if (!type)
        return '-';
    return type.toLowerCase().replace(/_/g, ' ');
}
function redacted(format) {
    return format === 'terminal' ? chalk.dim('redacted') : 'redacted';
}
export function titleFromConfig(channel) {
    const config = channel.config ?? {};
    return channel.name
        ?? config.name
        ?? config.channel
        ?? config.address
        ?? config.serviceName
        ?? config.account
        ?? `${normalizeType(channel.type)} alert channel`;
}
function targetFromConfig(channel, format) {
    const config = channel.config ?? {};
    const secret = redacted(format);
    switch (channel.type) {
        case 'EMAIL':
            return String(config.address ?? '-');
        case 'SLACK':
            return config.channel ? String(config.channel) : `Slack webhook (${secret})`;
        case 'SLACK_APP':
            return Array.isArray(config.slackChannels) ? config.slackChannels.join(', ') : '-';
        case 'WEBHOOK':
            return config.name ? String(config.name) : `Webhook URL (${secret})`;
        case 'PAGERDUTY':
            return String(config.serviceName ?? config.account ?? `Service key (${secret})`);
        case 'OPSGENIE':
            return [
                config.name,
                config.region,
                config.priority,
            ].filter(Boolean).join(' / ') || `API key (${secret})`;
        case 'SMS':
        case 'CALL':
            return String(config.name ?? config.number ?? config.phoneNumber ?? '-');
        default:
            return String(config.name ?? '-');
    }
}
function enabledEvents(channel, format) {
    const events = [];
    if (channel.sendFailure)
        events.push('failure');
    if (channel.sendRecovery)
        events.push('recovery');
    if (channel.sendDegraded)
        events.push('degraded');
    if (events.length === 0)
        return format === 'terminal' ? chalk.dim('-') : '-';
    return events.join(', ');
}
function sslExpiryLabel(channel, format) {
    if (!channel.sslExpiry)
        return boolLabel(false, format);
    const threshold = channel.sslExpiryThreshold ?? 30;
    return `yes (${threshold} days)`;
}
function subscriptionCount(channel) {
    return String(channel.subscriptions?.length ?? 0);
}
function buildAlertChannelColumns(format) {
    if (format === 'md') {
        return [
            { header: 'Type', value: c => normalizeType(c.type) },
            { header: 'Name', value: titleFromConfig },
            { header: 'Subscriptions', value: subscriptionCount },
            { header: 'Created', value: (c, fmt) => formatDate(c.created_at ?? c.createdAt, fmt) },
            { header: 'ID', value: c => String(c.id) },
        ];
    }
    return [
        { header: 'Type', width: 14, value: c => normalizeType(c.type) },
        { header: 'Name', minWidth: 12, maxWidth: 34, value: c => titleFromConfig(c) },
        { header: 'Target', minWidth: 12, maxWidth: 28, value: (c, fmt) => targetFromConfig(c, fmt) },
        { header: 'Subs', width: 8, align: 'right', value: subscriptionCount },
        { header: 'Created', width: 24, value: (c, fmt) => formatDate(c.created_at ?? c.createdAt, fmt) },
        { header: 'ID', minWidth: 8, maxWidth: 38, value: c => chalk.dim(String(c.id)) },
    ];
}
export function formatAlertChannels(channels, format) {
    return renderAdaptiveTable(buildAlertChannelColumns(format), channels, format);
}
export function formatAlertChannelPaginationInfo(pagination) {
    const { page, limit, total } = pagination;
    const totalPages = Math.max(1, Math.ceil(total / limit));
    const start = total === 0 ? 0 : (page - 1) * limit + 1;
    const end = Math.min(page * limit, total);
    return chalk.dim(`Showing ${start}-${end} of ${total} alert channels (page ${page}/${totalPages})`);
}
export function formatAlertChannelNavigationHints(pagination) {
    const { page, limit, total } = pagination;
    const totalPages = Math.ceil(total / limit);
    const hints = [];
    if (page < totalPages) {
        hints.push({ label: 'Next page', command: `checkly alert-channels list --page ${page + 1}` });
    }
    if (page > 1) {
        hints.push({ label: 'Prev page', command: `checkly alert-channels list --page ${page - 1}` });
    }
    hints.push({ label: 'View channel', command: 'checkly alert-channels get <id>' });
    hints.push({ label: 'View logs', command: 'checkly alert-channels logs <id> --status failed' });
    return renderCommandHints(hints, { gap: 1 });
}
const alertChannelDetailFields = [
    { label: 'Type', value: c => normalizeType(c.type) },
    { label: 'Target', value: targetFromConfig },
    { label: 'Enabled events', value: enabledEvents },
    { label: 'SSL expiry', value: sslExpiryLabel },
    { label: 'Auto-subscribe', value: (c, fmt) => boolLabel(c.autoSubscribe, fmt) },
    { label: 'Subscriptions', value: subscriptionCount },
    { label: 'Created', value: (c, fmt) => formatDate(c.created_at ?? c.createdAt, fmt) },
    { label: 'Updated', value: (c, fmt) => formatDate(c.updated_at ?? c.updatedAt, fmt) },
    { label: 'ID', value: c => String(c.id) },
];
function subscriptionType(subscription) {
    if (subscription.checkId || subscription.check)
        return 'check';
    if (subscription.groupId || subscription.group)
        return 'group';
    return 'subscription';
}
function subscriptionName(subscription, format) {
    const name = subscription.checkName ?? subscription.groupName ?? subscription.check?.name ?? subscription.group?.name;
    if (name)
        return name;
    return format === 'terminal' ? chalk.dim('-') : '-';
}
function subscriptionId(subscription) {
    return String(subscription.checkId
        ?? subscription.check?.id
        ?? subscription.groupId
        ?? subscription.group?.id
        ?? subscription.id
        ?? '-');
}
function hasSubscriptionNames(subscriptions) {
    return subscriptions.some(subscription => Boolean(subscription.checkName
        ?? subscription.groupName
        ?? subscription.check?.name
        ?? subscription.group?.name));
}
function buildSubscriptionColumns(format, subscriptions) {
    const includeName = hasSubscriptionNames(subscriptions);
    if (format === 'md') {
        const columns = [
            { header: 'Type', value: subscriptionType },
            { header: 'ID', value: subscriptionId },
        ];
        if (includeName) {
            columns.splice(1, 0, { header: 'Name', value: (s, fmt) => subscriptionName(s, fmt) });
        }
        return columns;
    }
    const columns = [
        { header: 'Type', width: 14, value: subscriptionType },
        { header: 'ID', minWidth: 8, maxWidth: 38, value: s => chalk.dim(subscriptionId(s)) },
    ];
    if (includeName) {
        columns.splice(1, 0, {
            header: 'Name',
            minWidth: 12,
            maxWidth: 32,
            value: (s, fmt) => subscriptionName(s, fmt),
        });
    }
    return columns;
}
export function formatAlertChannelSubscriptions(channel, format) {
    const subscriptions = channel.subscriptions ?? [];
    if (subscriptions.length === 0)
        return 'No subscriptions configured.';
    return renderAdaptiveTable(buildSubscriptionColumns(format, subscriptions), subscriptions, format);
}
export function formatAlertChannelDetail(channel, format) {
    const lines = [];
    lines.push(renderDetailFields(titleFromConfig(channel), alertChannelDetailFields, channel, format));
    lines.push('');
    if (format === 'md') {
        lines.push('## Subscriptions');
        lines.push('');
    }
    else {
        lines.push(chalk.bold('SUBSCRIPTIONS'));
    }
    lines.push(formatAlertChannelSubscriptions(channel, format));
    return lines.join('\n');
}
function formatNotificationStatus(status, format) {
    const label = status.toLowerCase().replace(/_/g, ' ');
    if (format === 'md')
        return label;
    switch (status) {
        case 'SUCCESS':
            return chalk.green(label);
        case 'FAILURE':
        case 'RATE_LIMITED':
            return chalk.red(label);
        case 'IN_PROGRESS':
            return chalk.yellow(label);
        default:
            return chalk.dim(label);
    }
}
function notificationCheckLabel(log) {
    return String(log.check?.name
        ?? log.checkName
        ?? log.alertConfig?.checkName
        ?? log.checkId
        ?? '-');
}
function notificationResultLabel(log) {
    return log.checkResultId ?? '-';
}
function notificationMessage(log) {
    return log.notificationResult ?? '-';
}
function buildAlertNotificationColumns(format) {
    if (format === 'md') {
        return [
            { header: 'Time', value: l => formatDate(l.timestamp, format) },
            { header: 'Status', value: l => formatNotificationStatus(l.status, format) },
            { header: 'Type', value: l => normalizeType(l.type) },
            { header: 'Check', value: notificationCheckLabel },
            { header: 'Result', value: notificationResultLabel },
            { header: 'Message', value: l => truncateError(notificationMessage(l), 120) },
        ];
    }
    return [
        { header: 'Time', width: 22, value: l => formatDate(l.timestamp, format) },
        { header: 'Status', width: 14, value: l => formatNotificationStatus(l.status, format) },
        { header: 'Type', width: 12, value: l => normalizeType(l.type) },
        { header: 'Check', minWidth: 12, maxWidth: 28, value: notificationCheckLabel },
        { header: 'Result', minWidth: 8, maxWidth: 22, value: l => chalk.dim(notificationResultLabel(l)) },
        { header: 'Message', minWidth: 16, maxWidth: 80, value: l => truncateError(notificationMessage(l), 160) },
    ];
}
export function formatAlertNotificationLogs(logs, format) {
    if (logs.length === 0)
        return 'No alert channel logs found.';
    return renderAdaptiveTable(buildAlertNotificationColumns(format), logs, format);
}
