// checkly/dist/reporters/abstract-list.js
import chalk from 'chalk';
import indentString from 'indent-string';
import { DateTime } from 'luxon';
import { CheckStatus, formatCheckTitle, getTestSessionUrl, printLn, resultToCheckStatus } from './util.js';
import { testSessions } from '../rest/api.js';
export default class AbstractListReporter {
    _clearString = '';
    _isCancelling = false;
    _isDetaching = false;
    runLocation;
    checkFilesMap;
    numChecks;
    verbose;
    testSessionId;
    hasPrintedTestSessionReference = false;
    _isSchedulingDelayExceeded;
    showStreamingHeaders;
    constructor(runLocation, verbose, options = {}) {
        this.runLocation = runLocation;
        this.verbose = verbose;
        this.showStreamingHeaders = options.showStreamingHeaders ?? true;
    }
    onBegin(checks, testSessionId) {
        this.testSessionId = testSessionId;
        this.numChecks = checks.length;
        // Sort the check files and checks alphabetically. This makes sure that there's a consistent order between runs.
        // For `checkly trigger`, getSourceFile() is not defined so we use optional chaining.
        const sortedCheckFiles = [...new Set(checks.map(({ check }) => check.getSourceFile?.()))].sort();
        const sortedChecks = checks.sort(({ check: a }, { check: b }) => a.name.localeCompare(b.name));
        this.checkFilesMap = new Map(sortedCheckFiles.map(file => [file, new Map()]));
        sortedChecks.forEach(({ check, sequenceId }) => {
            const fileMap = this.checkFilesMap.get(check.getSourceFile?.());
            fileMap.set(sequenceId, {
                check,
                titleString: formatCheckTitle(CheckStatus.SCHEDULING, check),
                checkStatus: CheckStatus.SCHEDULING,
                numRetries: 0,
            });
        });
        this._printTestSessionReference();
    }
    onCheckInProgress(check, sequenceId) {
        const checkFile = this.checkFilesMap.get(check.getSourceFile?.()).get(sequenceId);
        if (checkFile.checkStatus !== CheckStatus.SCHEDULING) {
            // For simplicity, if a check is retrying, we leave it in that state when the next retry attempt starts.
            return;
        }
        checkFile.titleString = formatCheckTitle(CheckStatus.RUNNING, check);
        checkFile.checkStatus = CheckStatus.RUNNING;
    }
    onSchedulingDelayExceeded() {
        this._isSchedulingDelayExceeded = true;
    }
    onCheckAttemptResult(sequenceId, checkResult, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    links) {
        const checkFile = this.checkFilesMap.get(checkResult.sourceFile).get(sequenceId);
        checkFile.checkStatus = CheckStatus.RETRIED;
        checkFile.titleString = formatCheckTitle(CheckStatus.RETRIED, checkResult);
        checkFile.numRetries++;
    }
    onCheckEnd(sequenceId, checkResult, testResultId, links) {
        const checkStatus = this.checkFilesMap.get(checkResult.sourceFile).get(sequenceId);
        checkStatus.result = checkResult;
        checkStatus.links = links;
        checkStatus.testResultId = testResultId;
        checkStatus.checkStatus = resultToCheckStatus(checkResult);
        checkStatus.titleString = formatCheckTitle(checkStatus.checkStatus, checkResult, {
            includeSourceFile: false,
        });
    }
    onError(err) {
        printLn(chalk.red('Unable to run checks: ') + err.message);
    }
    onStreamLogs(check, sequenceId, logs) {
        const checkFile = this.checkFilesMap.get(check.getSourceFile?.()).get(sequenceId);
        const logList = logs || [];
        // Clearing the summary first, printing logs second and finally adding the summary again at the end will ensure
        // the summary sticks to the bottom of the users screen with logs streaming in above.
        this._clearSummary();
        // Display the check title if this is the first time we're streaming logs for this check
        const isFirstLogBatch = !checkFile.hasStreamedLogs;
        checkFile.hasStreamedLogs = true;
        if (isFirstLogBatch && this.showStreamingHeaders) {
            // For Playwright tests, we need to create a better display name
            const displayCheck = {
                ...check,
                sourceFile: check.getSourceFile?.() || 'Playwright Test',
                name: check.name || check.logicalId || 'Unknown Test',
            };
            printLn(formatCheckTitle(CheckStatus.RUNNING, displayCheck, { includeSourceFile: true }));
        }
        // Format and display each log with proper indentation and timestamp handling
        logList.forEach(logEntry => {
            // Format timestamp from Unix timestamp to HH:mm:ss.SSS format
            const timestamp = DateTime.fromMillis(logEntry.timestamp).toFormat('HH:mm:ss.SSS');
            // Handle logs that contain newlines by splitting and prefixing each line with timestamp
            const messageLines = logEntry.message.split('\n');
            messageLines.forEach(line => {
                // Each line gets its own timestamp for clarity
                const formattedLine = `[${timestamp}] ${line}`;
                printLn(indentString(formattedLine, 4));
            });
        });
        this._printSummary();
    }
    onCancel() {
        this._isCancelling = true;
        this._clearSummary();
        this._printSummary();
    }
    onDetach() {
        this._isDetaching = true;
        this._clearSummary();
        this._printSummary();
    }
    // Clear the summary which was printed by _printStatus from stdout
    // TODO: Rather than clearing the whole status bar, we could overwrite the exact lines that changed.
    // This might look a bit smoother and reduce the flickering effects.
    _clearSummary() {
        if (this._clearString) {
            printLn(this._clearString);
            this._clearString = '';
        }
    }
    _printSummary(opts = {}) {
        const counts = {
            numFailed: 0, numPassed: 0, numDegraded: 0,
            numRunning: 0, numRetrying: 0, scheduling: 0, numCancelled: 0,
        };
        const status = [];
        if (this.checkFilesMap.size === 1 && this.checkFilesMap.has(undefined)) {
            status.push(chalk.bold('Summary:'));
        }
        for (const [sourceFile, checkMap] of this.checkFilesMap.entries()) {
            if (sourceFile)
                status.push(sourceFile);
            for (const [, { titleString, result, checkStatus }] of checkMap.entries()) {
                if (checkStatus === CheckStatus.SCHEDULING) {
                    counts.scheduling++;
                }
                else if (checkStatus === CheckStatus.RETRIED) {
                    counts.numRetrying++;
                }
                else if (!result) {
                    counts.numRunning++;
                }
                else if (result.hasFailures) {
                    counts.numFailed++;
                }
                else if (result.isDegraded) {
                    counts.numDegraded++;
                }
                else if (result.isCancelled) {
                    counts.numCancelled++;
                }
                else {
                    counts.numPassed++;
                }
                status.push(sourceFile ? indentString(titleString, 2) : titleString);
            }
        }
        if (!opts.skipCheckCount) {
            if (this._isCancelling) {
                status.push('');
                status.push(chalk.yellow('Cancelling checks... Use --detach to start checks in the cloud and exit after scheduling.'));
                if (!this.hasPrintedTestSessionReference) {
                    status.push(...this._getTestSessionReferenceLines());
                    this.hasPrintedTestSessionReference = true;
                }
            }
            if (this._isDetaching) {
                status.push('');
                status.push(chalk.yellow('Checks will continue running in the cloud.'));
                if (!this.hasPrintedTestSessionReference) {
                    status.push(...this._getTestSessionReferenceLines());
                    this.hasPrintedTestSessionReference = true;
                }
            }
            status.push('');
            status.push([
                counts.scheduling ? chalk.bold.blue(`${counts.scheduling} scheduling`) : undefined,
                counts.numRunning ? chalk.bold.magenta(`${counts.numRunning} running`) : undefined,
                counts.numRetrying ? chalk.bold(`${counts.numRetrying} retrying`) : undefined,
                counts.numFailed ? chalk.bold.red(`${counts.numFailed} failed`) : undefined,
                counts.numDegraded ? chalk.bold.yellow(`${counts.numDegraded} degraded`) : undefined,
                counts.numPassed ? chalk.bold.green(`${counts.numPassed} passed`) : undefined,
                counts.numCancelled ? chalk.bold.grey(`${counts.numCancelled} cancelled`) : undefined,
                `${this.numChecks} total`,
            ].filter(Boolean).join(', '));
            if (this._isSchedulingDelayExceeded && counts.scheduling) {
                status.push('Still waiting to schedule some checks. This may take a minute or two.');
            }
        }
        status.push('');
        const statusString = status.join('\n');
        printLn(statusString);
        // Ansi escape code for erasing the line and moving the cursor up
        this._clearString = '\r\x1B[K\r\x1B[1A'.repeat(statusString.split('\n').length + 1);
    }
    _printBriefSummary() {
        const counts = { numFailed: 0, numDegraded: 0, numPassed: 0, numPending: 0, numCancelled: 0 };
        const status = [];
        for (const [, checkMap] of this.checkFilesMap.entries()) {
            for (const [, { result }] of checkMap.entries()) {
                if (!result) {
                    counts.numPending++;
                }
                else if (result.hasFailures) {
                    counts.numFailed++;
                }
                else if (result.isDegraded) {
                    counts.numDegraded++;
                }
                else if (result.isCancelled) {
                    counts.numCancelled++;
                }
                else {
                    counts.numPassed++;
                }
            }
        }
        status.push('');
        status.push([
            counts.numFailed ? chalk.bold.red(`${counts.numFailed} failed`) : undefined,
            counts.numDegraded ? chalk.bold.yellow(`${counts.numDegraded} degraded`) : undefined,
            counts.numPassed ? chalk.bold.green(`${counts.numPassed} passed`) : undefined,
            counts.numCancelled ? chalk.bold.grey(`${counts.numCancelled} cancelled`) : undefined,
            counts.numPending ? chalk.bold.magenta(`${counts.numPending} pending`) : undefined,
            `${this.numChecks} total`,
        ].filter(Boolean).join(', '));
        status.push('');
        const statusString = status.join('\n');
        printLn(statusString);
        // Ansi escape code for erasing the line and moving the cursor up
        this._clearString = '\r\x1B[K\r\x1B[1A'.repeat(statusString.split('\n').length + 1);
    }
    async _printTestSessionsUrl() {
        if (this.testSessionId) {
            this._printTestSessionReference();
            try {
                const { data: { link } } = await testSessions.getShortLink(this.testSessionId);
                printLn(`${chalk.white('Detailed session summary at:')} ${chalk.underline.cyan(link)}`, 2);
            }
            catch {
                printLn(`${chalk.white('Detailed session summary at:')} ${chalk.underline.cyan(getTestSessionUrl(this.testSessionId))}`, 2);
            }
        }
    }
    _getTestSessionReferenceLines() {
        if (!this.testSessionId) {
            return [];
        }
        return [
            `${chalk.white('Test session ID:')} ${this.testSessionId}`,
            `${chalk.white('Inspect session:')} ${chalk.cyan(`checkly test-sessions get ${this.testSessionId} --watch`)}`,
            `${chalk.white('Open session:')} ${chalk.underline.cyan(getTestSessionUrl(this.testSessionId))}`,
        ];
    }
    _printTestSessionReference() {
        if (this.hasPrintedTestSessionReference) {
            return;
        }
        const lines = this._getTestSessionReferenceLines();
        if (!lines.length) {
            return;
        }
        printLn(lines.join('\n'), 2);
        this.hasPrintedTestSessionReference = true;
    }
    _printTip(tip) {
        printLn(`Tip > ${chalk.white(tip)}`, 2);
    }
    _runLocationString() {
        if (this.runLocation.type === 'PUBLIC') {
            return this.runLocation.region;
        }
        else {
            return `private location ${this.runLocation.slugName}`;
        }
    }
}
