// checkly/dist/reporters/github.js
import * as fs from 'fs';
import * as path from 'path';
import AbstractListReporter from './abstract-list.js';
import { CheckStatus, formatDuration, getTestSessionUrl, printLn, resultToCheckStatus } from './util.js';
const outputFile = './checkly-github-report.md';
function nonNullable(value) {
    return value !== null && value !== undefined;
}
export class GithubMdBuilder {
    testSessionId;
    numChecks;
    runLocation;
    checkFilesMap;
    subHeader;
    tableHeaders;
    extraTableHeadersWithLinks;
    tableRows = [];
    hasFilenames;
    header = '# Checkly Test Session Summary';
    tableSeparatorFiller = '|:-';
    tableSeparator = '|';
    constructor(options) {
        this.testSessionId = options.testSessionId;
        this.numChecks = options.numChecks;
        this.runLocation = options.runLocation;
        this.checkFilesMap = options.checkFilesMap;
        this.subHeader = [];
        this.hasFilenames = !(options.checkFilesMap.size === 1 && options.checkFilesMap.has(undefined));
        this.tableHeaders = [
            'Result',
            'Name',
            'Check Type',
            this.hasFilenames ? 'Filename' : undefined,
            'Duration',
        ].filter(nonNullable);
        this.extraTableHeadersWithLinks = ['Link'];
        this.tableRows = [];
    }
    render() {
        this.subHeader.push(`Ran **${this.numChecks}** checks in **${this.runLocation}**.`);
        if (this.testSessionId) {
            this.subHeader.push(`[View detailed test session summary](${getTestSessionUrl(this.testSessionId)})`);
        }
        if (this.testSessionId) {
            this.tableHeaders = this.tableHeaders.concat(this.extraTableHeadersWithLinks);
        }
        for (const [, checkMap] of this.checkFilesMap.entries()) {
            for (const [, { result, testResultId }] of checkMap.entries()) {
                const checkStatus = resultToCheckStatus(result);
                const tableRow = [
                    `${checkStatus === CheckStatus.FAILED ? '❌ Fail' : checkStatus === CheckStatus.DEGRADED ? '⚠️ Degraded' : '✅ Pass'}`,
                    `${result.name}`,
                    `${result.checkType}`,
                    this.hasFilenames ? `\`${result.sourceFile}\`` : undefined,
                    `${formatDuration(result.responseTime)} `,
                ].filter(nonNullable);
                if (this.testSessionId && testResultId) {
                    const linkColumn = `[Full test report](${getTestSessionUrl(this.testSessionId)}/results/${testResultId})`;
                    tableRow.push(linkColumn);
                }
                this.tableRows.push(this.tableSeparator + tableRow.join(this.tableSeparator) + this.tableSeparator);
            }
        }
        const markdown = this.header + '\n'
            + this.subHeader.join('\n') + '\n'
            + this.tableSeparator + this.tableHeaders.join('|') + this.tableSeparator + '\n'
            + this.tableSeparatorFiller.repeat(this.tableHeaders.length) + this.tableSeparator + '\n'
            + this.tableRows.sort((a, b) => a < b ? 1 : -1).join('\n') + '\n';
        return markdown;
    }
}
export default class GithubReporter extends AbstractListReporter {
    onBegin(checks, testSessionId) {
        super.onBegin(checks, testSessionId);
        printLn(`Running ${this.numChecks} checks in ${this._runLocationString()}.`, 2);
    }
    onEnd() {
        this._printBriefSummary();
        const githubMdBuilder = new GithubMdBuilder({
            testSessionId: this.testSessionId,
            numChecks: this.numChecks,
            runLocation: this._runLocationString(),
            checkFilesMap: this.checkFilesMap,
        });
        const markDown = githubMdBuilder.render();
        const summaryFilename = process.env.CHECKLY_REPORTER_GITHUB_OUTPUT ?? outputFile;
        fs.mkdirSync(path.resolve(path.dirname(summaryFilename)), { recursive: true });
        fs.writeFileSync(summaryFilename, markDown);
        printLn(`Github summary saved in '${path.resolve(summaryFilename)}'.`, 2);
        this._printTestSessionsUrl();
    }
}
