// checkly/dist/constructs/agentic-check.js
import { Check } from './check.js';
import { Session } from './session.js';
import { CheckTypes } from '../constants.js';
import { InvalidPropertyValueDiagnostic } from './construct-diagnostics.js';
/**
 * Backwards-compatible default for checks that do not set a location and do
 * not inherit one from the project config.
 */
const DEFAULT_AGENTIC_CHECK_LOCATION = 'us-east-1';
/**
 * Creates an Agentic Check that uses AI to monitor websites and applications.
 *
 * Agentic checks use a prompt to define what should be verified, without
 * requiring traditional scripts. The AI agent interprets the prompt and
 * performs the checks.
 *
 * @example
 * ```typescript
 * new AgenticCheck('homepage-health', {
 *   name: 'Homepage Health Check',
 *   prompt: `
 *     Navigate to https://example.com and verify:
 *     1. The page loads with a 200 status
 *     2. The main heading is visible
 *     3. No console errors are present
 *   `,
 * })
 * ```
 */
export class AgenticCheck extends Check {
    prompt;
    agentRuntime;
    /**
     * Constructs the Agentic Check instance.
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props check configuration properties
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.prompt = props.prompt;
        this.agentRuntime = props.agentRuntime;
        // Preserve the old implicit single-region behavior for checks that do not
        // set locations directly and do not inherit project-level locations.
        this.locations ??= [DEFAULT_AGENTIC_CHECK_LOCATION];
        // Defensive overrides: even though these props are omitted from the type,
        // `Check.applyConfigDefaults()` may pull them in from the project-level
        // `checks` config defaults. Force them to values the platform currently
        // honors so the construct never claims to support something it doesn't.
        this.privateLocations = undefined;
        this.runParallel = false;
        this.retryStrategy = undefined;
        this.shouldFail = undefined;
        this.doubleCheck = undefined;
        this.triggerIncident = undefined;
        // Default frequency to 30m if the user did not specify one.
        if (this.frequency === undefined) {
            this.frequency = 30;
        }
        Session.registerConstruct(this);
        this.addSubscriptions();
    }
    describe() {
        return `AgenticCheck:${this.logicalId}`;
    }
    async validate(diagnostics) {
        await super.validate(diagnostics);
        if (!this.prompt || this.prompt.trim().length === 0) {
            diagnostics.add(new InvalidPropertyValueDiagnostic('prompt', new Error('"prompt" is required and must not be empty.')));
        }
        else if (this.prompt.length > 10_000) {
            diagnostics.add(new InvalidPropertyValueDiagnostic('prompt', new Error(`"prompt" must be at most 10000 characters, got ${this.prompt.length}.`)));
        }
        this.validateAgentRuntime(diagnostics);
    }
    // eslint-disable-next-line require-await
    async validateAgentRuntime(diagnostics) {
        if (this.agentRuntime?.skills) {
            for (const [index, skill] of this.agentRuntime.skills.entries()) {
                if (typeof skill !== 'string' || skill.trim().length === 0) {
                    diagnostics.add(new InvalidPropertyValueDiagnostic('agentRuntime.skills', new Error(`"agentRuntime.skills[${index}]" must be a non-empty string.`)));
                }
            }
        }
    }
    synthesize() {
        return {
            ...super.synthesize(),
            checkType: CheckTypes.AGENTIC,
            prompt: this.prompt,
            // Always emit `agentRuntime` so the backend has an explicit picture of
            // the skills the user wants. Omitted skills mean "the agent should not
            // have them", not "preserve whatever was there before".
            agentRuntime: {
                skills: this.agentRuntime?.skills ?? [],
            },
        };
    }
}
