// checkly/dist/constructs/alert-channel-subscription.js
import { Construct } from './construct.js';
import { Session } from './session.js';
/**
 * Creates an Alert Channel Subscription
 *
 * @remarks
 *
 * This class make use of the Alert Channel Subscriptions endpoints.
 */
export class AlertChannelSubscription extends Construct {
    alertChannelId;
    checkId;
    groupId;
    activated;
    static __checklyType = 'alert-channel-subscription';
    /**
     * Constructs the Alert Channel Syscription instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props alert channel subscription configuration properties
     */
    constructor(logicalId, props) {
        super(AlertChannelSubscription.__checklyType, logicalId);
        this.alertChannelId = props.alertChannelId;
        this.checkId = props.checkId;
        this.groupId = props.groupId;
        this.activated = props.activated;
        Session.registerConstruct(this);
    }
    describe() {
        return `AlertChannelSubscription:${this.logicalId}`;
    }
    allowInChecklyConfig() {
        return true;
    }
    synthesize() {
        return {
            alertChannelId: this.alertChannelId,
            checkId: this.checkId,
            groupId: this.groupId,
            activated: this.activated,
        };
    }
}
