// checkly/dist/constructs/api-check.js
import fs from 'node:fs/promises';
import { RuntimeCheck } from './check.js';
import { Session } from './session.js';
import { isContent, isEntrypoint } from './construct.js';
import { DeprecatedPropertyDiagnostic, InvalidPropertyValueDiagnostic } from './construct-diagnostics.js';
import { ApiCheckBundle } from './api-check-bundle.js';
import { validateResponseTimes } from './internal/common-diagnostics.js';
/**
 * Creates an API Check to monitor HTTP endpoints and APIs.
 *
 * API checks allow you to monitor REST APIs, GraphQL endpoints, and any HTTP-based service.
 * You can validate response status codes, response times, headers, and response body content.
 *
 * @example
 * ```typescript
 * // Basic API check
 * new ApiCheck('hello-api', {
 *   name: 'Hello API',
 *   request: {
 *     method: 'GET',
 *     url: 'https://api.example.com/hello',
 *     assertions: [
 *       AssertionBuilder.statusCode().equals(200)
 *     ]
 *   }
 * })
 *
 * // Advanced API check with POST request
 * new ApiCheck('user-api', {
 *   name: 'User API Check',
 *   frequency: Frequency.EVERY_5M,
 *   locations: ['us-east-1', 'eu-west-1'],
 *   request: {
 *     method: 'POST',
 *     url: 'https://api.example.com/users',
 *     headers: [{ key: 'Content-Type', value: 'application/json' }],
 *     body: JSON.stringify({ name: 'test-user' }),
 *     bodyType: 'JSON',
 *     assertions: [
 *       AssertionBuilder.statusCode().equals(201),
 *       AssertionBuilder.jsonBody('$.id').isNotNull(),
 *       AssertionBuilder.responseTime().lessThan(1000)
 *     ]
 *   },
 *   maxResponseTime: 5000,
 *   degradedResponseTime: 2000
 * })
 *
 * // Error validation check (shouldFail required for error status checks)
 * new ApiCheck('not-found-check', {
 *   name: 'Not Found Check',
 *   shouldFail: true,
 *   request: {
 *     method: 'GET',
 *     url: 'https://api.example.com/nonexistent',
 *     assertions: [
 *       AssertionBuilder.statusCode().equals(404)
 *     ]
 *   }
 * })
 * ```
 *
 * @see {@link https://www.checklyhq.com/docs/constructs/api-check/ | ApiCheck API Reference}
 * @see {@link https://www.checklyhq.com/docs/detect/synthetic-monitoring/api-checks/overview/ | API Checks Documentation}
 */
export class ApiCheck extends RuntimeCheck {
    request;
    localSetupScript;
    setupScript;
    localTearDownScript;
    tearDownScript;
    degradedResponseTime;
    maxResponseTime;
    /**
     * Constructs the API Check instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props check configuration properties
     *
     * {@link https://www.checklyhq.com/docs/constructs/api-check/ Read more in the docs}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.setupScript = props.setupScript;
        this.localSetupScript = props.localSetupScript;
        this.tearDownScript = props.tearDownScript;
        this.localTearDownScript = props.localTearDownScript;
        this.request = props.request;
        this.degradedResponseTime = props.degradedResponseTime;
        this.maxResponseTime = props.maxResponseTime;
        Session.registerConstruct(this);
        this.addSubscriptions();
        this.addPrivateLocationCheckAssignments();
    }
    describe() {
        return `ApiCheck:${this.logicalId}`;
    }
    supportsOnlyOnNetworkErrorRetryStrategy() {
        return true;
    }
    async validate(diagnostics) {
        await super.validate(diagnostics);
        if (this.setupScript) {
            if (!isEntrypoint(this.setupScript) && !isContent(this.setupScript)) {
                diagnostics.add(new InvalidPropertyValueDiagnostic('setupScript', new Error(`Either "entrypoint" or "content" is required.`)));
            }
            else if (isEntrypoint(this.setupScript) && isContent(this.setupScript)) {
                diagnostics.add(new InvalidPropertyValueDiagnostic('setupScript', new Error(`Provide exactly one of "entrypoint" or "content", but not both.`)));
            }
            else if (isEntrypoint(this.setupScript)) {
                const entrypoint = this.resolveContentFilePath(this.setupScript.entrypoint);
                try {
                    await fs.access(entrypoint, fs.constants.R_OK);
                }
                catch (err) {
                    diagnostics.add(new InvalidPropertyValueDiagnostic('setupScript', new Error(`Unable to access entrypoint file "${entrypoint}": ${err.message}`, { cause: err })));
                }
            }
        }
        if (this.localSetupScript) {
            diagnostics.add(new DeprecatedPropertyDiagnostic('localSetupScript', new Error(`Use "setupScript" instead.`)));
        }
        if (this.tearDownScript) {
            if (!isEntrypoint(this.tearDownScript) && !isContent(this.tearDownScript)) {
                diagnostics.add(new InvalidPropertyValueDiagnostic('tearDownScript', new Error(`Either "entrypoint" or "content" is required.`)));
            }
            else if (isEntrypoint(this.tearDownScript) && isContent(this.tearDownScript)) {
                diagnostics.add(new InvalidPropertyValueDiagnostic('tearDownScript', new Error(`Provide exactly one of "entrypoint" or "content", but not both.`)));
            }
            else if (isEntrypoint(this.tearDownScript)) {
                const entrypoint = this.resolveContentFilePath(this.tearDownScript.entrypoint);
                try {
                    await fs.access(entrypoint, fs.constants.R_OK);
                }
                catch (err) {
                    diagnostics.add(new InvalidPropertyValueDiagnostic('tearDownScript', new Error(`Unable to access entrypoint file "${entrypoint}": ${err.message}`, { cause: err })));
                }
            }
        }
        if (this.localTearDownScript) {
            diagnostics.add(new DeprecatedPropertyDiagnostic('localTearDownScript', new Error(`Use "tearDownScript" instead.`)));
        }
        if (this.runtimeId) {
            const runtime = Session.getRuntime(this.runtimeId);
            if (!runtime) {
                diagnostics.add(new InvalidPropertyValueDiagnostic('runtimeId', new Error(`"${this.runtimeId}" is not a known runtime.`)));
            }
        }
        await validateResponseTimes(diagnostics, this, {
            degradedResponseTime: 30_000,
            maxResponseTime: 30_000,
        });
    }
    async bundle(bundler) {
        const props = {};
        if (this.localSetupScript) {
            props.localSetupScript = this.localSetupScript;
        }
        if (this.setupScript) {
            if (isEntrypoint(this.setupScript)) {
                const { script, scriptPath, dependencies } = await ApiCheck.bundle(bundler, this.resolveContentFilePath(this.setupScript.entrypoint), this.runtimeId);
                props.localSetupScript = script;
                props.setupScriptPath = scriptPath;
                props.setupScriptDependencies = dependencies;
            }
            else {
                props.localSetupScript = this.setupScript.content;
            }
        }
        if (this.localTearDownScript) {
            props.localTearDownScript = this.localTearDownScript;
        }
        if (this.tearDownScript) {
            if (isEntrypoint(this.tearDownScript)) {
                const { script, scriptPath, dependencies } = await ApiCheck.bundle(bundler, this.resolveContentFilePath(this.tearDownScript.entrypoint), this.runtimeId);
                props.localTearDownScript = script;
                props.tearDownScriptPath = scriptPath;
                props.tearDownScriptDependencies = dependencies;
            }
            else {
                props.localTearDownScript = this.tearDownScript.content;
            }
        }
        return new ApiCheckBundle(this, props);
    }
    static async bundle(bundler, entrypoint, runtimeId) {
        const runtime = Session.getRuntime(runtimeId);
        if (!runtime) {
            throw new Error(`${runtimeId} is not supported`);
        }
        const parser = Session.getParser(runtime);
        const parsed = await parser.parse(entrypoint);
        // Maybe we can get the parsed deps with the content immediately
        const deps = [];
        for (const { filePath, content } of parsed.dependencies) {
            deps.push(Session.registerSharedFile({
                path: Session.relativePosixPath(filePath),
                content,
            }));
        }
        return {
            script: parsed.entrypoint.content,
            scriptPath: Session.relativePosixPath(parsed.entrypoint.filePath),
            dependencies: deps,
        };
    }
    synthesize() {
        return {
            ...super.synthesize(),
            checkType: 'API',
            request: this.request,
            localSetupScript: this.localSetupScript,
            localTearDownScript: this.localTearDownScript,
            degradedResponseTime: this.degradedResponseTime,
            maxResponseTime: this.maxResponseTime,
        };
    }
}
