// checkly/dist/constructs/api-check-codegen.js
import { Codegen, validateScript } from './internal/codegen/index.js';
import { expr, ident } from '../sourcegen/index.js';
import { buildRuntimeCheckProps } from './check-codegen.js';
import { valueForRequest } from './api-request-codegen.js';
const construct = 'ApiCheck';
export class ApiCheckCodegen extends Codegen {
    describe(resource) {
        return `API Check: ${resource.name}`;
    }
    gencode(logicalId, resource, context) {
        const filePath = context.filePath('resources/api-checks', resource.name, {
            tags: resource.tags,
            isolate: true,
            unique: true,
        });
        const file = this.program.generatedConstructFile(filePath.fullPath);
        file.namedImport(construct, 'checkly/constructs');
        file.section(expr(ident(construct), builder => {
            builder.new(builder => {
                builder.string(logicalId);
                builder.object(builder => {
                    if (resource.localSetupScript) {
                        const content = resource.localSetupScript;
                        validateScript(content);
                        const snippetFiles = context.findScriptSnippetFiles(content);
                        for (const snippetFile of snippetFiles) {
                            const localSnippetFile = this.program.generatedSupportFile(`${file.dirname}/snippets/${snippetFile.basename}`);
                            localSnippetFile.plainImport(localSnippetFile.relativePath(snippetFile));
                        }
                        builder.object('setupScript', builder => {
                            const scriptFile = this.program.staticSupportFile(`${file.dirname}/setup-script`, content);
                            builder.string('entrypoint', file.relativePath(scriptFile));
                        });
                    }
                    else if (resource.setupSnippetId) {
                        const snippetFile = context.lookupAuxiliarySnippetFile(resource.setupSnippetId);
                        if (!snippetFile) {
                            throw new Error(`Setup script refers to snippet #${resource.setupSnippetId} which is missing`);
                        }
                        const scriptFile = this.program.generatedSupportFile(`${file.dirname}/setup-script`);
                        scriptFile.plainImport(scriptFile.relativePath(snippetFile));
                        builder.object('setupScript', builder => {
                            builder.string('entrypoint', file.relativePath(scriptFile));
                        });
                    }
                    if (resource.localTearDownScript) {
                        const content = resource.localTearDownScript;
                        validateScript(content);
                        const snippetFiles = context.findScriptSnippetFiles(content);
                        for (const snippetFile of snippetFiles) {
                            const aliasFile = this.program.generatedSupportFile(`${file.dirname}/snippets/${snippetFile.basename}`);
                            aliasFile.plainImport(aliasFile.relativePath(snippetFile));
                        }
                        builder.object('tearDownScript', builder => {
                            const scriptFile = this.program.staticSupportFile(`${file.dirname}/teardown-script`, content);
                            builder.string('entrypoint', file.relativePath(scriptFile));
                        });
                    }
                    else if (resource.tearDownSnippetId) {
                        const snippetFile = context.lookupAuxiliarySnippetFile(resource.tearDownSnippetId);
                        if (!snippetFile) {
                            throw new Error(`Teardown script refers to snippet #${resource.tearDownSnippetId} which is missing`);
                        }
                        const scriptFile = this.program.generatedSupportFile(`${file.dirname}/teardown-script`);
                        scriptFile.plainImport(scriptFile.relativePath(snippetFile));
                        builder.object('tearDownScript', builder => {
                            builder.string('entrypoint', file.relativePath(scriptFile));
                        });
                    }
                    if (resource.degradedResponseTime !== undefined) {
                        builder.number('degradedResponseTime', resource.degradedResponseTime);
                    }
                    if (resource.maxResponseTime !== undefined) {
                        builder.number('maxResponseTime', resource.maxResponseTime);
                    }
                    buildRuntimeCheckProps(this.program, file, builder, resource, context);
                    builder.value('request', valueForRequest(this.program, file, context, resource.request));
                });
            });
        }));
    }
}
