// checkly/dist/constructs/browser-check-codegen.js
import { Codegen, validateScript } from './internal/codegen/index.js';
import { expr, ident } from '../sourcegen/index.js';
import { buildRuntimeCheckProps } from './check-codegen.js';
import { valueForPlaywrightConfig } from './playwright-config-codegen.js';
const construct = 'BrowserCheck';
export class BrowserCheckCodegen extends Codegen {
    describe(resource) {
        return `Browser Check: ${resource.name}`;
    }
    gencode(logicalId, resource, context) {
        const filePath = context.filePath('resources/browser-checks', resource.name, {
            tags: resource.tags,
            isolate: true,
            unique: true,
        });
        const file = this.program.generatedConstructFile(filePath.fullPath);
        file.namedImport(construct, 'checkly/constructs');
        file.section(expr(ident(construct), builder => {
            builder.new(builder => {
                builder.string(logicalId);
                builder.object(builder => {
                    builder.object('code', builder => {
                        validateScript(resource.script);
                        const snippetFiles = context.findScriptSnippetFiles(resource.script);
                        for (const snippetFile of snippetFiles) {
                            const localSnippetFile = this.program.generatedSupportFile(`${file.dirname}/snippets/${snippetFile.basename}`);
                            localSnippetFile.plainImport(localSnippetFile.relativePath(snippetFile));
                        }
                        const scriptFile = this.program.staticSpecFile(filePath.extless, resource.script);
                        builder.string('entrypoint', file.relativePath(scriptFile));
                    });
                    if (resource.sslCheckDomain !== undefined && resource.sslCheckDomain !== null && resource.sslCheckDomain !== '') {
                        builder.string('sslCheckDomain', resource.sslCheckDomain);
                    }
                    if (resource.playwrightConfig) {
                        builder.value('playwrightConfig', valueForPlaywrightConfig(resource.playwrightConfig));
                    }
                    buildRuntimeCheckProps(this.program, file, builder, resource, context);
                });
            });
        }));
    }
}
