// checkly/dist/constructs/check-codegen.js
import { Codegen } from './internal/codegen/index.js';
import { AgenticCheckCodegen } from './agentic-check-codegen.js';
import { valueForAlertEscalation } from './alert-escalation-policy-codegen.js';
import { ApiCheckCodegen } from './api-check-codegen.js';
import { BrowserCheckCodegen } from './browser-check-codegen.js';
import { CheckGroupCodegen, valueForCheckGroupFromId } from './check-group-codegen.js';
import { valueForFrequency } from './frequency-codegen.js';
import { HeartbeatMonitorCodegen } from './heartbeat-monitor-codegen.js';
import { valueForKeyValuePair } from './key-value-pair-codegen.js';
import { MultiStepCheckCodegen } from './multi-step-check-codegen.js';
import { valueForRetryStrategy } from './retry-strategy-codegen.js';
import { TcpMonitorCodegen } from './tcp-monitor-codegen.js';
import { UrlMonitorCodegen } from './url-monitor-codegen.js';
import { valueForPrivateLocationFromId } from './private-location-codegen.js';
import { valueForAlertChannelFromId } from './alert-channel-codegen.js';
import { DnsMonitorCodegen } from './dns-monitor-codegen.js';
import { IcmpMonitorCodegen } from './icmp-monitor-codegen.js';
export function buildCheckProps(program, genfile, builder, resource, context, options = {}) {
    builder.string('name', resource.name, { order: -1000 });
    if (resource.description != null) {
        builder.string('description', resource.description);
    }
    if (resource.activated !== undefined) {
        builder.boolean('activated', resource.activated);
    }
    if (resource.muted !== undefined && resource.muted !== false) {
        builder.boolean('muted', resource.muted);
    }
    if (resource.shouldFail !== undefined && resource.shouldFail !== false) {
        builder.boolean('shouldFail', resource.shouldFail);
    }
    if (resource.locations) {
        const locations = resource.locations;
        if (locations.length > 0) {
            builder.array('locations', builder => {
                for (const location of locations) {
                    builder.string(location);
                }
            });
        }
    }
    const privateLocationIds = (() => {
        try {
            return context.lookupCheckPrivateLocations(resource.id);
        }
        catch {
            return;
        }
    })();
    if (privateLocationIds !== undefined) {
        builder.array('privateLocations', builder => {
            for (const privateLocationId of privateLocationIds) {
                try {
                    const privateLocationVariable = context.lookupPrivateLocation(privateLocationId);
                    const id = context.importVariable(privateLocationVariable, genfile);
                    builder.value(id);
                }
                catch {
                    try {
                        const privateLocationVariable = context.lookupFriendPrivateLocation(privateLocationId);
                        const id = context.importFriendVariable(privateLocationVariable, genfile);
                        builder.value(id);
                    }
                    catch {
                        builder.value(valueForPrivateLocationFromId(genfile, privateLocationId));
                    }
                }
            }
        });
    }
    if (resource.tags) {
        const tags = resource.tags;
        if (tags.length > 0) {
            builder.array('tags', builder => {
                for (const tag of tags) {
                    builder.string(tag);
                }
            });
        }
    }
    if (resource.frequency !== undefined) {
        if (typeof resource.frequency === 'number') {
            builder.value('frequency', valueForFrequency(genfile, {
                frequency: resource.frequency,
                frequencyOffset: resource.frequencyOffset,
            }));
        }
        else {
            builder.value('frequency', valueForFrequency(genfile, resource.frequency));
        }
    }
    if (resource.groupId) {
        try {
            const groupVariable = context.lookupCheckGroup(resource.groupId);
            const id = context.importVariable(groupVariable, genfile);
            builder.value('group', id);
        }
        catch {
            try {
                const groupVariable = context.lookupFriendCheckGroup(resource.groupId);
                const id = context.importFriendVariable(groupVariable, genfile);
                builder.value('group', id);
            }
            catch {
                builder.value('group', valueForCheckGroupFromId(genfile, resource.groupId));
            }
        }
    }
    const alertChannelIds = (() => {
        try {
            return context.lookupCheckAlertChannels(resource.id);
        }
        catch {
            return;
        }
    })();
    if (alertChannelIds !== undefined) {
        builder.array('alertChannels', builder => {
            for (const alertChannelId of alertChannelIds) {
                try {
                    const alertChannelVariable = context.lookupAlertChannel(alertChannelId);
                    const id = context.importVariable(alertChannelVariable, genfile);
                    builder.value(id);
                }
                catch {
                    try {
                        const alertChannelVariable = context.lookupFriendAlertChannel(alertChannelId);
                        const id = context.importFriendVariable(alertChannelVariable, genfile);
                        builder.value(id);
                    }
                    catch {
                        builder.value(valueForAlertChannelFromId(genfile, alertChannelId));
                    }
                }
            }
        });
    }
    if (resource.alertSettings) {
        builder.value('alertEscalationPolicy', valueForAlertEscalation(genfile, resource.alertSettings));
    }
    if (resource.testOnly !== undefined && resource.testOnly !== false) {
        builder.boolean('testOnly', resource.testOnly);
    }
    if (!options.skipRetryStrategy) {
        builder.value('retryStrategy', valueForRetryStrategy(genfile, resource.retryStrategy));
    }
    if (resource.runParallel !== undefined && resource.runParallel !== false) {
        builder.boolean('runParallel', resource.runParallel);
    }
}
export function buildRuntimeCheckProps(program, genfile, builder, resource, context) {
    buildCheckProps(program, genfile, builder, resource, context);
    if (resource.runtimeId) {
        builder.string('runtimeId', resource.runtimeId);
    }
    if (resource.environmentVariables) {
        const variables = resource.environmentVariables;
        if (variables.length > 0) {
            builder.array('environmentVariables', builder => {
                for (const variable of variables) {
                    builder.value(valueForKeyValuePair(program, genfile, context, variable));
                }
            });
        }
    }
}
export class CheckCodegen extends Codegen {
    agenticCheckCodegen;
    apiCheckCodegen;
    browserCheckCodegen;
    checkGroupCodegen;
    heartbeatMonitorCodegen;
    multiStepCheckCodegen;
    tcpMonitorCodegen;
    urlMonitorCodegen;
    dnsMonitorCodegen;
    icmpMonitorCodegen;
    constructor(program) {
        super(program);
        this.agenticCheckCodegen = new AgenticCheckCodegen(program);
        this.apiCheckCodegen = new ApiCheckCodegen(program);
        this.browserCheckCodegen = new BrowserCheckCodegen(program);
        this.checkGroupCodegen = new CheckGroupCodegen(program);
        this.heartbeatMonitorCodegen = new HeartbeatMonitorCodegen(program);
        this.multiStepCheckCodegen = new MultiStepCheckCodegen(program);
        this.tcpMonitorCodegen = new TcpMonitorCodegen(program);
        this.urlMonitorCodegen = new UrlMonitorCodegen(program);
        this.dnsMonitorCodegen = new DnsMonitorCodegen(program);
        this.icmpMonitorCodegen = new IcmpMonitorCodegen(program);
    }
    describe(resource) {
        const { checkType } = resource;
        switch (checkType) {
            case 'AGENTIC':
                return this.agenticCheckCodegen.describe(resource);
            case 'BROWSER':
                return this.browserCheckCodegen.describe(resource);
            case 'API':
                return this.apiCheckCodegen.describe(resource);
            case 'TCP':
                return this.tcpMonitorCodegen.describe(resource);
            case 'MULTI_STEP':
                return this.multiStepCheckCodegen.describe(resource);
            case 'HEARTBEAT':
                return this.heartbeatMonitorCodegen.describe(resource);
            case 'URL':
                return this.urlMonitorCodegen.describe(resource);
            case 'DNS':
                return this.dnsMonitorCodegen.describe(resource);
            case 'ICMP':
                return this.icmpMonitorCodegen.describe(resource);
            default:
                throw new Error(`Unable to describe unsupported check type '${checkType}'.`);
        }
    }
    gencode(logicalId, resource, context) {
        const { checkType } = resource;
        switch (checkType) {
            case 'AGENTIC':
                this.agenticCheckCodegen.gencode(logicalId, resource, context);
                return;
            case 'BROWSER':
                this.browserCheckCodegen.gencode(logicalId, resource, context);
                return;
            case 'API':
                this.apiCheckCodegen.gencode(logicalId, resource, context);
                return;
            case 'TCP':
                this.tcpMonitorCodegen.gencode(logicalId, resource, context);
                return;
            case 'MULTI_STEP':
                this.multiStepCheckCodegen.gencode(logicalId, resource, context);
                return;
            case 'HEARTBEAT':
                this.heartbeatMonitorCodegen.gencode(logicalId, resource, context);
                return;
            case 'URL':
                this.urlMonitorCodegen.gencode(logicalId, resource, context);
                return;
            case 'DNS':
                this.dnsMonitorCodegen.gencode(logicalId, resource, context);
                return;
            case 'ICMP':
                this.icmpMonitorCodegen.gencode(logicalId, resource, context);
                return;
            default:
                throw new Error(`Unable to generate code for unsupported check type '${checkType}'.`);
        }
    }
}
