// checkly/dist/constructs/check-group-ref.js
import { CheckGroupV1 } from './check-group-v1.js';
import { Construct } from './construct.js';
import { validatePhysicalIdIsNumeric } from './internal/common-diagnostics.js';
import { Session } from './session.js';
/**
 * Creates a reference to an existing Check Group (both v1 and v2).
 *
 * References link existing resources to a project without managing them.
 *
 * Use {@link CheckGroupV2.fromId()} instead of instantiating this class directly.
 */
export class CheckGroupRef extends Construct {
    constructor(logicalId, physicalId) {
        super(CheckGroupV1.__checklyType, logicalId, physicalId, false);
        Session.registerConstruct(this);
    }
    describe() {
        return `CheckGroupRef:${this.logicalId}`;
    }
    async validate(diagnostics) {
        await super.validate(diagnostics);
        await validatePhysicalIdIsNumeric(diagnostics, 'CheckGroupV2', this.physicalId);
    }
    getCheckDefaults() {
        // The only value CheckGroup.getCheckDefaults() returns is frequency,
        // which exists purely for convenience, and only at CLI-level. It never
        // gets sent to the backend. If references are being used, no access to
        // this feature is required, and it can be ignored entirely.
        return {};
    }
    getBrowserCheckDefaults() {
        // See the comment for getCheckDefaults(), the same applies here.
        return {};
    }
    getMultiStepCheckDefaults() {
        // See the comment for getCheckDefaults(), the same applies here.
        return {};
    }
    allowInChecklyConfig() {
        return true;
    }
    synthesize() {
        return null;
    }
}
