// checkly/dist/constructs/check-group-v1.js
import path from 'node:path';
import { glob } from 'glob';
import { DeprecatedConstructDiagnostic, DeprecatedPropertyDiagnostic, InvalidPropertyValueDiagnostic } from './construct-diagnostics.js';
import { CheckTypes } from '../constants.js';
import { AlertChannelSubscription } from './alert-channel-subscription.js';
import { BrowserCheck } from './browser-check.js';
import { CheckGroupRef } from './check-group-ref.js';
import { Construct } from './construct.js';
import { MultiStepCheck } from './multi-step-check.js';
import { PrivateLocationGroupAssignment } from './private-location-group-assignment.js';
import { Ref } from './ref.js';
import { Session } from './session.js';
import { validateDeprecatedDoubleCheck } from './internal/common-diagnostics.js';
const defaultApiCheckDefaults = {
    headers: [],
    queryParameters: [],
    url: '',
    basicAuth: {
        username: '',
        password: '',
    },
};
/**
 * Creates a Check Group (v1).
 *
 * We strongly recommend upgrading to {@link CheckGroupV2} instead.
 *
 * The original CheckGroup v1 comes with implicit defaults that are not
 * intuitive and make it impossible to keep check-level behavior for
 * certain properties.
 *
 * The following properties will always use the group's values if set, or
 * implicit defaults if not set:
 *
 *   - {@link CheckGroupV1Props.alertEscalationPolicy}
 *   - {@link CheckGroupV1Props.retryStrategy}
 *   - {@link CheckGroupV1Props.runParallel}
 *
 * Please check the documentation for the individual properties to see their
 * behavior.
 *
 * For more information regarding the update, please see:
 *   https://feedback.checklyhq.com/changelog/checkly-groups-update-organize-checks-your-way
 *
 * @deprecated Use {@link CheckGroupV2} instead.
 */
export class CheckGroupV1 extends Construct {
    name;
    activated;
    muted;
    doubleCheck;
    runtimeId;
    locations;
    privateLocations;
    tags;
    concurrency;
    frequency;
    environmentVariables;
    alertChannels;
    localSetupScript;
    localTearDownScript;
    apiCheckDefaults;
    browserChecks;
    multiStepChecks;
    retryStrategy;
    runParallel;
    alertSettings;
    useGlobalAlertSettings;
    static __checklyType = 'check-group';
    /**
     * Constructs the CheckGroup instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props CheckGroup configuration properties
     *
     * {@link https://checklyhq.com/docs/cli/constructs-reference/#checkgroup Read more in the docs}
     */
    constructor(logicalId, props) {
        super(CheckGroupV1.__checklyType, logicalId);
        this.name = props.name;
        this.activated = props.activated;
        this.muted = props.muted;
        this.doubleCheck = props.doubleCheck;
        this.tags = props.tags;
        this.runtimeId = props.runtimeId;
        this.locations = props.locations ?? [];
        this.privateLocations = props.privateLocations;
        this.concurrency = props.concurrency;
        // `frequency` is not a CheckGroup resource property. Not present in synthesize()
        this.frequency = props.frequency;
        this.apiCheckDefaults = { ...defaultApiCheckDefaults, ...props.apiCheckDefaults };
        this.alertSettings = props.alertEscalationPolicy;
        this.useGlobalAlertSettings = !this.alertSettings;
        this.environmentVariables = props.environmentVariables ?? [];
        this.alertChannels = props.alertChannels ?? [];
        this.localSetupScript = props.localSetupScript;
        this.localTearDownScript = props.localTearDownScript;
        this.retryStrategy = props.retryStrategy;
        this.runParallel = props.runParallel;
        // `browserChecks` is not a CheckGroup resource property. Not present in synthesize()
        this.browserChecks = props.browserChecks;
        const fileAbsolutePath = Session.checkFileAbsolutePath;
        if (props.browserChecks?.testMatch) {
            this.__addChecks(fileAbsolutePath, props.browserChecks.testMatch, CheckTypes.BROWSER);
        }
        this.multiStepChecks = props.multiStepChecks;
        if (props.multiStepChecks?.testMatch) {
            this.__addChecks(fileAbsolutePath, props.multiStepChecks.testMatch, CheckTypes.MULTI_STEP);
        }
        Session.registerConstruct(this);
        this.__addSubscriptions();
        this.__addPrivateLocationGroupAssignments();
    }
    describe() {
        return `CheckGroupV1:${this.logicalId}`;
    }
    // eslint-disable-next-line require-await
    async onBeforeValidate(diagnostics) {
        diagnostics.add(new DeprecatedConstructDiagnostic('CheckGroup', new Error(`Please update to CheckGroupV2 which has more intuitive behavior.`
            + `\n\n`
            + `For more information, please see:\n`
            + `  https://feedback.checklyhq.com/changelog/checkly-groups-update-organize-checks-your-way`)));
    }
    async validateDoubleCheck(diagnostics) {
        await validateDeprecatedDoubleCheck(diagnostics, this);
    }
    async validate(diagnostics) {
        await super.validate(diagnostics);
        await this.onBeforeValidate(diagnostics);
        await this.validateDoubleCheck(diagnostics);
        if (this.environmentVariables) {
            this.environmentVariables.forEach(ev => {
                // only empty string is checked because the KeyValuePair.value doesn't allow undefined or null.
                if (ev.value === '') {
                    diagnostics.add(new InvalidPropertyValueDiagnostic('environmentVariables', new Error(`Value must not be empty.`)));
                }
            });
        }
        if (this.localSetupScript) {
            diagnostics.add(new DeprecatedPropertyDiagnostic('localSetupScript', new Error(`Use the setupScript property directly in your ApiChecks instead.`)));
        }
        if (this.localTearDownScript) {
            diagnostics.add(new DeprecatedPropertyDiagnostic('localTearDownScript', new Error(`Use the tearDownScript property directly in your ApiChecks instead.`)));
        }
    }
    /**
     * @param id - The numeric ID of the existing check group
     */
    static fromId(id) {
        return new CheckGroupRef(`check-group-${id}`, id);
    }
    __addChecks(fileAbsolutePath, testMatch, checkType) {
        const parent = path.dirname(fileAbsolutePath);
        const matched = glob.sync(testMatch, { nodir: true, cwd: parent });
        for (const match of matched) {
            const filepath = path.join(parent, match);
            const props = {
                group: this,
                name: match,
                code: {
                    entrypoint: filepath,
                },
                // the browserChecks props inherited from the group are applied in BrowserCheck.constructor()
            };
            const checkLogicalId = Session.relativePosixPath(filepath);
            if (checkType === CheckTypes.BROWSER) {
                new BrowserCheck(checkLogicalId, props);
            }
            else {
                new MultiStepCheck(checkLogicalId, props);
            }
        }
    }
    __addSubscriptions() {
        if (!this.alertChannels) {
            return;
        }
        for (const alertChannel of this.alertChannels) {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const subscription = new AlertChannelSubscription(`check-group-alert-channel-subscription#${this.logicalId}#${alertChannel.logicalId}`, {
                alertChannelId: Ref.from(alertChannel.logicalId),
                groupId: Ref.from(this.logicalId),
                activated: true,
            });
        }
    }
    __addPrivateLocationGroupAssignments() {
        if (!this.privateLocations) {
            return;
        }
        for (const privateLocation of this.privateLocations) {
            // slugName strings are processed in loadAllPrivateLocations()
            if (typeof privateLocation === 'string') {
                continue;
            }
            // use private location assignment for instances
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const assignment = new PrivateLocationGroupAssignment(`private-location-group-assignment#${this.logicalId}#${privateLocation.logicalId}`, {
                privateLocationId: Ref.from(privateLocation.logicalId),
                groupId: Ref.from(this.logicalId),
            });
        }
    }
    getCheckDefaults() {
        // TODO: investigate if make sense to add all other check's properties
        return {
            frequency: this.frequency,
        };
    }
    getBrowserCheckDefaults() {
        // TODO: investigate if make sense to add all other browser-check's properties
        return {
            frequency: this.browserChecks?.frequency,
        };
    }
    getMultiStepCheckDefaults() {
        return {
            frequency: this.multiStepChecks?.frequency,
        };
    }
    allowInChecklyConfig() {
        return true;
    }
    synthesize() {
        return {
            name: this.name,
            activated: this.activated,
            muted: this.muted,
            tags: this.tags,
            locations: this.locations,
            runtimeId: this.runtimeId,
            // private-location instances are assigned with loadAllPrivateLocations()
            privateLocations: undefined,
            concurrency: this.concurrency,
            localSetupScript: this.localSetupScript,
            localTearDownScript: this.localTearDownScript,
            apiCheckDefaults: this.apiCheckDefaults,
            environmentVariables: this.environmentVariables,
            // The backend doesn't actually support the `NO_RETRIES` type, it uses `null` instead.
            retryStrategy: this.retryStrategy?.type === 'NO_RETRIES'
                ? null
                : this.retryStrategy,
            // When `retryStrategy: NO_RETRIES` and `doubleCheck: undefined`, we want to let the user disable all retries.
            // The backend has a Joi default of `doubleCheck: true`, though, so we need special handling for this case.
            doubleCheck: this.doubleCheck === undefined && this.retryStrategy?.type === 'NO_RETRIES'
                ? false
                : this.doubleCheck,
            runParallel: this.runParallel,
            alertSettings: this.alertSettings,
            useGlobalAlertSettings: this.useGlobalAlertSettings,
        };
    }
}
