// checkly/dist/constructs/construct-codegen.js
import { Codegen } from './internal/codegen/index.js';
import { AlertChannelCodegen } from './alert-channel-codegen.js';
import { AlertChannelSubscriptionCodegen } from './alert-channel-subscription-codegen.js';
import { CheckCodegen } from './check-codegen.js';
import { CheckGroupCodegen } from './check-group-codegen.js';
import { DashboardCodegen } from './dashboard-codegen.js';
import { MaintenanceWindowCodegen } from './maintenance-window-codegen.js';
import { PrivateLocationCodegen } from './private-location-codegen.js';
import { PrivateLocationCheckAssignmentCodegen } from './private-location-check-assignment-codegen.js';
import { PrivateLocationGroupAssignmentCodegen } from './private-location-group-assignment-codegen.js';
import { StatusPageServiceCodegen } from './status-page-service-codegen.js';
import { StatusPageCodegen } from './status-page-codegen.js';
const resourceOrder = {
    'alert-channel-subscription': 800,
    'alert-channel': 810,
    'check-group': 700,
    'check': 600,
    'dashboard': 0,
    'maintenance-window': 0,
    'private-location-check-assignment': 900,
    'private-location-group-assignment': 900,
    'private-location': 910,
    'status-page': 500,
    'status-page-service': 510,
};
export function sortResources(resources) {
    return resources.sort((a, b) => {
        const ao = resourceOrder[a.type] ?? -1;
        const bo = resourceOrder[b.type] ?? -1;
        return bo - ao;
    });
}
export class ConstructCodegen extends Codegen {
    alertChannelCodegen;
    alertChannelSubscriptionCodegen;
    checkCodegen;
    checkGroupCodegen;
    dashboardCodegen;
    maintenanceWindowCodegen;
    privateLocationCodegen;
    privateLocationCheckAssignmentCodegen;
    privateLocationGroupAssignmentCodegen;
    statusPageCodegen;
    statusPageServiceCodegen;
    codegensByType;
    constructor(program) {
        super(program);
        this.alertChannelCodegen = new AlertChannelCodegen(program);
        this.alertChannelSubscriptionCodegen = new AlertChannelSubscriptionCodegen(program);
        this.checkCodegen = new CheckCodegen(program);
        this.checkGroupCodegen = new CheckGroupCodegen(program);
        this.dashboardCodegen = new DashboardCodegen(program);
        this.maintenanceWindowCodegen = new MaintenanceWindowCodegen(program);
        this.privateLocationCodegen = new PrivateLocationCodegen(program);
        this.privateLocationCheckAssignmentCodegen = new PrivateLocationCheckAssignmentCodegen(program);
        this.privateLocationGroupAssignmentCodegen = new PrivateLocationGroupAssignmentCodegen(program);
        this.statusPageCodegen = new StatusPageCodegen(program);
        this.statusPageServiceCodegen = new StatusPageServiceCodegen(program);
        this.codegensByType = {
            'alert-channel-subscription': this.alertChannelSubscriptionCodegen,
            'alert-channel': this.alertChannelCodegen,
            'check-group': this.checkGroupCodegen,
            'check': this.checkCodegen,
            'dashboard': this.dashboardCodegen,
            'maintenance-window': this.maintenanceWindowCodegen,
            'private-location-check-assignment': this.privateLocationCheckAssignmentCodegen,
            'private-location-group-assignment': this.privateLocationGroupAssignmentCodegen,
            'private-location': this.privateLocationCodegen,
            'status-page': this.statusPageCodegen,
            'status-page-service': this.statusPageServiceCodegen,
        };
    }
    describe(resource) {
        const codegen = this.codegensByType[resource.type];
        if (codegen === undefined) {
            throw new Error(`Unable to describe unsupported resource type '${resource.type}'.`);
        }
        return codegen.describe(resource.payload);
    }
    prepare(logicalId, resource, context) {
        const codegen = this.codegensByType[resource.type];
        if (codegen === undefined) {
            throw new Error(`Unable to generate code for unsupported resource type '${resource.type}'.`);
        }
        codegen.prepare(resource.logicalId, resource.payload, context);
    }
    gencode(logicalId, resource, context) {
        const codegen = this.codegensByType[resource.type];
        if (codegen === undefined) {
            throw new Error(`Unable to generate code for unsupported resource type '${resource.type}'.`);
        }
        codegen.gencode(resource.logicalId, resource.payload, context);
    }
}
