// checkly/dist/constructs/construct-diagnostics.js
import { Diagnostic, Diagnostics, ErrorDiagnostic, WarningDiagnostic } from './diagnostics.js';
export class InvalidPropertyValueDiagnostic extends ErrorDiagnostic {
    property;
    constructor(property, error) {
        super({
            title: `Invalid property value`,
            message: `The value provided for property "${property}" is not valid.`
                + `\n\n`
                + `Reason: ${error.message}`,
            error,
        });
        this.property = property;
    }
}
export class SubstitutedPropertyValueDiagnostic extends WarningDiagnostic {
    property;
    constructor(property, error) {
        super({
            title: `Substitution of property value`,
            message: `The value of property "${property}" was substituted with another value.`
                + `\n\n`
                + `Reason: ${error.message}`,
        });
        this.property = property;
    }
}
export class RequiredPropertyDiagnostic extends ErrorDiagnostic {
    property;
    constructor(property, error) {
        super({
            title: `Missing required property`,
            message: `Property "${property}" is required and must be set.`
                + `\n\n`
                + `Reason: ${error.message}`,
            error,
        });
        this.property = property;
    }
}
export class ConflictingPropertyDiagnostic extends ErrorDiagnostic {
    property1;
    property2;
    constructor(property1, property2, error) {
        super({
            title: `Conflicting property`,
            message: `Property "${property1}" cannot be set when "${property2}" is set.`
                + `\n\n`
                + `Hint: ${error.message}`,
            error,
        });
        this.property1 = property1;
        this.property2 = property2;
    }
}
export class DeprecatedPropertyDiagnostic extends WarningDiagnostic {
    property;
    constructor(property, error) {
        super({
            title: `Use of deprecated property`,
            message: `Property "${property}" is deprecated and will eventually be removed.`
                + `\n\n`
                + `Hint: ${error.message}`,
        });
        this.property = property;
    }
}
export class RemovedPropertyDiagnostic extends ErrorDiagnostic {
    property;
    constructor(property, error) {
        super({
            title: `Use of removed property`,
            message: `Property "${property}" has been removed.`
                + `\n\n`
                + `Hint: ${error.message}`,
            error,
        });
        this.property = property;
    }
}
export class UnsupportedPropertyDiagnostic extends ErrorDiagnostic {
    property;
    constructor(property, error) {
        super({
            title: `Use of unsupported property`,
            message: `Property "${property}" is not supported.`
                + `\n\n`
                + `Reason: ${error.message}`,
            error,
        });
        this.property = property;
    }
}
export class DeprecatedConstructDiagnostic extends WarningDiagnostic {
    construct;
    constructor(construct, error) {
        super({
            title: `Use of deprecated Construct`,
            message: `Construct "${construct}" is deprecated and will eventually be removed.`
                + `\n\n`
                + `Hint: ${error.message}`,
        });
        this.construct = construct;
    }
}
export class UnsupportedRuntimeFeatureDiagnostic extends ErrorDiagnostic {
    runtimeId;
    constructor(runtimeId, error) {
        super({
            title: `Use of unsupported runtime feature`,
            message: `Runtime "${runtimeId}" does not support the requested feature.`
                + `\n\n`
                + `Missing feature: ${error.message}`,
            error,
        });
        this.runtimeId = runtimeId;
    }
}
export class UnsatisfiedLocalPrerequisitesDiagnostic extends ErrorDiagnostic {
    constructor(error) {
        super({
            title: `Unsatisfied local prerequisites`,
            message: `Local environment does not satisfy the prerequisites for this `
                + `functionality.`
                + `\n\n`
                + `Cause: ${error.message}`,
            error,
        });
    }
}
export class ConstructDiagnostic extends Diagnostic {
    underlying;
    constructor(construct, underlying) {
        super({
            title: `[${construct.describe()}] ${underlying.title}`,
            message: underlying.message,
        });
        this.underlying = underlying;
    }
    isFatal() {
        return this.underlying.isFatal();
    }
    isBenign() {
        return this.underlying.isBenign();
    }
}
export class ConstructDiagnostics extends Diagnostics {
    construct;
    constructor(construct) {
        super();
        this.construct = construct;
    }
    add(diagnostic) {
        super.add(new ConstructDiagnostic(this.construct, diagnostic));
    }
}
