// checkly/dist/constructs/dashboard.js
import fs from 'node:fs/promises';
import { Construct, isContent, isEntrypoint } from './construct.js';
import { Session } from './session.js';
import { InvalidPropertyValueDiagnostic } from './construct-diagnostics.js';
import { DashboardBundle } from './dashboard-bundle.js';
/**
 * Creates a Dashboard
 *
 * @remarks
 *
 * This class make use of the Dashboard endpoints.
 */
export class Dashboard extends Construct {
    tags;
    customUrl;
    customDomain;
    logo;
    favicon;
    link;
    header;
    description;
    width;
    refreshRate;
    paginate;
    paginationRate;
    checksPerPage;
    useTagsAndOperator;
    hideTags;
    enableIncidents;
    expandChecks;
    showHeader;
    customCSS;
    isPrivate;
    showP95;
    showP99;
    static __checklyType = 'dashboard';
    /**
     * Constructs the Dashboard instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props dashboard configuration properties
     *
     * {@link https://www.checklyhq.com/docs/constructs/dashboard/ Read more in the docs}
     */
    constructor(logicalId, props) {
        super(Dashboard.__checklyType, logicalId);
        this.tags = props.tags;
        this.customUrl = props.customUrl;
        this.customDomain = props.customDomain;
        this.logo = props.logo;
        this.favicon = props.favicon;
        this.link = props.link;
        this.header = props.header;
        this.description = props.description;
        this.width = props.width;
        this.refreshRate = props.refreshRate;
        this.paginate = props.paginate;
        this.paginationRate = props.paginationRate;
        this.checksPerPage = props.checksPerPage;
        this.useTagsAndOperator = props.useTagsAndOperator;
        this.hideTags = props.hideTags;
        this.enableIncidents = props.enableIncidents;
        this.expandChecks = props.expandChecks;
        this.showHeader = props.showHeader;
        this.isPrivate = props.isPrivate;
        this.showP95 = props.showP95;
        this.showP99 = props.showP99;
        this.customCSS = props.customCSS;
        Session.registerConstruct(this);
    }
    describe() {
        return `Dashboard:${this.logicalId}`;
    }
    async validate(diagnostics) {
        await super.validate(diagnostics);
        if (!this.customUrl && !this.customDomain) {
            diagnostics.add(new InvalidPropertyValueDiagnostic('customUrl', new Error(`Required unless "customDomain" is set.`)));
            diagnostics.add(new InvalidPropertyValueDiagnostic('customDomain', new Error(`Required unless "customUrl" is set.`)));
        }
        if (this.customCSS) {
            if (!isEntrypoint(this.customCSS) && !isContent(this.customCSS)) {
                diagnostics.add(new InvalidPropertyValueDiagnostic('customCSS', new Error(`Either "entrypoint" or "content" is required.`)));
            }
            else if (isEntrypoint(this.customCSS) && isContent(this.customCSS)) {
                diagnostics.add(new InvalidPropertyValueDiagnostic('customCSS', new Error(`Provide exactly one of "entrypoint" or "content", but not both.`)));
            }
            else if (isEntrypoint(this.customCSS)) {
                const entrypoint = this.resolveContentFilePath(this.customCSS.entrypoint);
                try {
                    await fs.access(entrypoint, fs.constants.R_OK);
                }
                catch (err) {
                    diagnostics.add(new InvalidPropertyValueDiagnostic('customCSS', new Error(`Unable to access file "${entrypoint}": ${err.message}`, { cause: err })));
                }
            }
        }
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async bundle(bundler) {
        const customCSS = await (async () => {
            if (this.customCSS) {
                if (isEntrypoint(this.customCSS)) {
                    const entrypoint = this.resolveContentFilePath(this.customCSS.entrypoint);
                    const content = await fs.readFile(entrypoint);
                    return content.toString('utf8');
                }
                return this.customCSS.content;
            }
        })();
        return new DashboardBundle(this, {
            customCSS,
        });
    }
    synthesize() {
        return {
            tags: this.tags,
            customUrl: this.customUrl,
            customDomain: this.customDomain,
            logo: this.logo,
            favicon: this.favicon,
            link: this.link,
            header: this.header,
            description: this.description,
            width: this.width,
            refreshRate: this.refreshRate,
            paginate: this.paginate,
            paginationRate: this.paginationRate,
            checksPerPage: this.checksPerPage,
            useTagsAndOperator: this.useTagsAndOperator,
            hideTags: this.hideTags,
            enableIncidents: this.enableIncidents,
            expandChecks: this.expandChecks,
            showHeader: this.showHeader,
            isPrivate: this.isPrivate,
            showP95: this.showP95,
            showP99: this.showP99,
        };
    }
}
