// checkly/dist/constructs/diagnostics.js
/**
 * Abstract base class for diagnostic messages.
 * Diagnostics are used to report issues during validation and processing.
 */
export class Diagnostic {
    /** The title/summary of the diagnostic */
    title;
    /** Detailed message describing the diagnostic */
    message;
    /**
     * Creates a new diagnostic instance.
     * @param options The diagnostic configuration
     */
    constructor(options) {
        this.title = options.title;
        this.message = options.message;
    }
}
export class ErrorDiagnostic extends Diagnostic {
    error;
    constructor(options) {
        super(options);
        this.error = options.error;
    }
    isFatal() {
        return true;
    }
    isBenign() {
        return false;
    }
}
export class WarningDiagnostic extends Diagnostic {
    isFatal() {
        return false;
    }
    isBenign() {
        return false;
    }
}
export class NoticeDiagnostic extends Diagnostic {
    isFatal() {
        return false;
    }
    isBenign() {
        return true;
    }
}
export class Diagnostics {
    observations = [];
    #fatal = false;
    #benign = true;
    isFatal() {
        return this.#fatal;
    }
    isBenign() {
        return this.#benign;
    }
    add(diagnostic) {
        this.observations.push(diagnostic);
        this.#fatal ||= diagnostic.isFatal();
        this.#benign &&= diagnostic.isBenign();
    }
    extend(...diagnostics) {
        for (const diags of diagnostics) {
            for (const observation of diags.observations) {
                this.add(observation);
            }
        }
    }
}
