// checkly/dist/constructs/dns-monitor.js
import { Monitor } from './monitor.js';
import { Session } from './session.js';
import { validateResponseTimes } from './internal/common-diagnostics.js';
import { RequiredPropertyDiagnostic } from './construct-diagnostics.js';
/**
 * Creates a DNS Monitor
 */
export class DnsMonitor extends Monitor {
    request;
    degradedResponseTime;
    maxResponseTime;
    /**
     * Constructs the DNS Monitor instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props configuration properties
     *
     * {@link https://www.checklyhq.com/docs/constructs/dns-monitor/ Read more in the docs}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.request = props.request;
        this.degradedResponseTime = props.degradedResponseTime;
        this.maxResponseTime = props.maxResponseTime;
        Session.registerConstruct(this);
        this.addSubscriptions();
        this.addPrivateLocationCheckAssignments();
    }
    describe() {
        return `DnsMonitor:${this.logicalId}`;
    }
    async validate(diagnostics) {
        await super.validate(diagnostics);
        if (this.request.nameServer && this.request.port === undefined) {
            diagnostics.add(new RequiredPropertyDiagnostic('port', new Error(`A value for "port" is required when "nameServer" is set.`)));
        }
        if (!this.request.nameServer && this.request.port !== undefined) {
            diagnostics.add(new RequiredPropertyDiagnostic('nameServer', new Error(`A value for "nameServer" is required when "port" is set.`)));
        }
        await validateResponseTimes(diagnostics, this, {
            degradedResponseTime: 5_000,
            maxResponseTime: 5_000,
        });
    }
    synthesize() {
        return {
            ...super.synthesize(),
            checkType: 'DNS',
            request: this.request,
            degradedResponseTime: this.degradedResponseTime,
            maxResponseTime: this.maxResponseTime,
        };
    }
}
