// checkly/dist/constructs/email-alert-channel.js
import { AlertChannel } from './alert-channel.js';
import { Session } from './session.js';
/**
 * Creates an Email Alert Channel
 *
 * @remarks
 *
 * This class make use of the Alert Channel endpoints.
 */
export class EmailAlertChannel extends AlertChannel {
    address;
    /**
     * Constructs the Email Alert Channel instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props email alert channel configuration properties
     * {@link https://www.checklyhq.com/docs/constructs/email-alert-channel/ Read more in the docs}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.address = props.address;
        Session.registerConstruct(this);
    }
    describe() {
        return `EmailAlertChannel:${this.logicalId}`;
    }
    synthesize() {
        return {
            ...super.synthesize(),
            type: 'EMAIL',
            config: {
                address: this.address,
            },
        };
    }
}
