// checkly/dist/constructs/heartbeat-monitor.js
import { Monitor } from './monitor.js';
import { Session } from './session.js';
import { DateTime } from 'luxon';
import { CheckTypes } from '../constants.js';
function _customPeriodGraceValidation(heartbeat) {
    const now = DateTime.now();
    const addedTimePeriod = now.plus({ [heartbeat.periodUnit]: heartbeat.period });
    const addedGracePeriod = now.plus({ [heartbeat.graceUnit]: heartbeat.grace });
    const MAX_PERIOD_GRACE_DAYS = 365;
    const MIN_PERIOD_SECONDS = 30;
    if (addedTimePeriod.diff(now, 'days').days > MAX_PERIOD_GRACE_DAYS
        || addedTimePeriod.diff(now, 'seconds').seconds < MIN_PERIOD_SECONDS) {
        throw new Error('Period must be between 30 seconds and 365 days.');
    }
    if (addedGracePeriod.diff(now, 'days').days > MAX_PERIOD_GRACE_DAYS) {
        throw new Error('Grace must be less than 366 days.');
    }
}
/**
 * Creates a Heartbeat Monitor
 */
export class HeartbeatMonitor extends Monitor {
    heartbeat;
    /**
     * Constructs the Heartbeat Monitor instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props configuration properties
     * {@link https://www.checklyhq.com/docs/constructs/heartbeat-monitor/ Read more in the docs}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        _customPeriodGraceValidation(props);
        this.heartbeat = {
            period: props.period,
            periodUnit: props.periodUnit,
            grace: props.grace,
            graceUnit: props.graceUnit,
        };
        Session.registerConstruct(this);
        this.addSubscriptions();
    }
    describe() {
        return `HeartbeatMonitor:${this.logicalId}`;
    }
    synthesize() {
        return {
            ...super.synthesize(),
            checkType: CheckTypes.HEARTBEAT,
            heartbeat: this.heartbeat,
        };
    }
}
// Aliases for backwards compatibility.
export { HeartbeatMonitor as HeartbeatCheck, };
