// checkly/dist/constructs/icmp-monitor.js
import { Monitor } from './monitor.js';
import { Session } from './session.js';
/**
 * Creates an ICMP Monitor
 */
export class IcmpMonitor extends Monitor {
    request;
    degradedPacketLossThreshold;
    maxPacketLossThreshold;
    /**
     * Constructs the ICMP Monitor instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props configuration properties
     *
     * {@link https://www.checklyhq.com/docs/constructs/icmp-monitor/ Read more in the docs}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.request = props.request;
        this.degradedPacketLossThreshold = props.degradedPacketLossThreshold;
        this.maxPacketLossThreshold = props.maxPacketLossThreshold;
        Session.registerConstruct(this);
        this.addSubscriptions();
        this.addPrivateLocationCheckAssignments();
    }
    describe() {
        return `IcmpMonitor:${this.logicalId}`;
    }
    async validate(diagnostics) {
        await super.validate(diagnostics);
    }
    synthesize() {
        return {
            ...super.synthesize(),
            checkType: 'ICMP',
            request: this.request,
            degradedPacketLossThreshold: this.degradedPacketLossThreshold,
            maxPacketLossThreshold: this.maxPacketLossThreshold,
        };
    }
}
