// checkly/dist/constructs/maintenance-window.js
import { Construct } from './construct.js';
import { Session } from './session.js';
/**
 * Creates a Maintenance Window
 *
 * @remarks
 *
 * This class make use of the Maintenance Window endpoints.
 */
export class MaintenanceWindow extends Construct {
    name;
    tags;
    startsAt;
    endsAt;
    repeatInterval;
    repeatUnit;
    repeatEndsAt;
    static __checklyType = 'maintenance-window';
    /**
     * Constructs the Maintenance Window instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props maintenance window configuration properties
     *
     * {@link https://www.checklyhq.com/docs/constructs/maintenance-window/ Read more in the docs}
     */
    constructor(logicalId, props) {
        super(MaintenanceWindow.__checklyType, logicalId);
        this.name = props.name;
        this.tags = props.tags;
        this.startsAt = props.startsAt;
        this.endsAt = props.endsAt;
        this.repeatInterval = props.repeatInterval;
        this.repeatUnit = props.repeatUnit;
        this.repeatEndsAt = props.repeatEndsAt;
        Session.registerConstruct(this);
    }
    describe() {
        return `MaintenanceWindow:${this.logicalId}`;
    }
    synthesize() {
        return {
            name: this.name,
            tags: this.tags,
            startsAt: this.startsAt,
            endsAt: this.endsAt,
            repeatInterval: this.repeatInterval,
            repeatUnit: this.repeatUnit,
            repeatEndsAt: this.repeatEndsAt,
        };
    }
}
