// checkly/dist/constructs/multi-step-check.js
import fs from 'node:fs/promises';
import { RuntimeCheck } from './check.js';
import { Session } from './session.js';
import { isContent, isEntrypoint } from './construct.js';
import { CheckTypes } from '../constants.js';
import { InvalidPropertyValueDiagnostic, UnsupportedRuntimeFeatureDiagnostic } from './construct-diagnostics.js';
import { MultiStepCheckBundle } from './multi-step-check-bundle.js';
import { makeConfigDefaultsGetter } from './check-config.js';
/**
 * Creates a multi-step Check
 *
 * @remarks
 *
 * This class make use of the multi-step checks endpoints.
 */
export class MultiStepCheck extends RuntimeCheck {
    code;
    playwrightConfig;
    /**
     * Constructs the multi-step instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props check configuration properties
     * {@link https://www.checklyhq.com/docs/constructs/multistep-check/ Read more in the docs}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        const config = this.applyConfigDefaults(props);
        this.code = config.code;
        this.playwrightConfig = config.playwrightConfig;
        Session.registerConstruct(this);
        this.addSubscriptions();
        this.addPrivateLocationCheckAssignments();
    }
    describe() {
        return `MultiStepCheck:${this.logicalId}`;
    }
    async validate(diagnostics) {
        await super.validate(diagnostics);
        if (!isEntrypoint(this.code) && !isContent(this.code)) {
            diagnostics.add(new InvalidPropertyValueDiagnostic('code', new Error(`Either "entrypoint" or "content" is required.`)));
        }
        else if (isEntrypoint(this.code) && isContent(this.code)) {
            diagnostics.add(new InvalidPropertyValueDiagnostic('code', new Error(`Provide exactly one of "entrypoint" or "content", but not both.`)));
        }
        else if (isEntrypoint(this.code)) {
            const entrypoint = this.resolveContentFilePath(this.code.entrypoint);
            try {
                const stats = await fs.stat(entrypoint);
                if (stats.size === 0) {
                    diagnostics.add(new InvalidPropertyValueDiagnostic('code', new Error(`The entrypoint file "${entrypoint}" must not be empty.`)));
                }
            }
            catch (err) {
                diagnostics.add(new InvalidPropertyValueDiagnostic('code', new Error(`Unable to access entrypoint file "${entrypoint}": ${err.message}`, { cause: err })));
            }
        }
        const runtime = Session.getRuntime(this.runtimeId);
        if (runtime) {
            if (!runtime.multiStepSupport) {
                diagnostics.add(new UnsupportedRuntimeFeatureDiagnostic(runtime.name, new Error(`Multi-Step Checks are not supported.`)));
            }
        }
    }
    configDefaultsGetter(props) {
        return makeConfigDefaultsGetter(props.group?.getMultiStepCheckDefaults(), Session.multiStepCheckDefaults, props.group?.getCheckDefaults(), Session.checkDefaults);
    }
    applyConfigDefaults(props) {
        const config = super.applyConfigDefaults(props);
        const defaults = this.configDefaultsGetter(props);
        config.playwrightConfig ??= defaults('playwrightConfig');
        return config;
    }
    static async bundle(bundler, entry, runtimeId) {
        const runtime = Session.getRuntime(runtimeId);
        if (!runtime) {
            throw new Error(`${runtimeId} is not supported`);
        }
        const parser = Session.getParser(runtime);
        const parsed = await parser.parse(entry);
        // Maybe we can get the parsed deps with the content immediately
        const deps = [];
        for (const { filePath, content } of parsed.dependencies) {
            deps.push(Session.registerSharedFile({
                path: Session.relativePosixPath(filePath),
                content,
            }));
        }
        return {
            script: parsed.entrypoint.content,
            scriptPath: Session.relativePosixPath(parsed.entrypoint.filePath),
            dependencies: deps,
        };
    }
    getSourceFile() {
        return this.__checkFilePath;
    }
    async bundle(bundler) {
        return new MultiStepCheckBundle(this, await (async () => {
            if (isEntrypoint(this.code)) {
                const bundle = await MultiStepCheck.bundle(bundler, this.resolveContentFilePath(this.code.entrypoint), this.runtimeId);
                if (!bundle.script) {
                    throw new Error(`The "code" property must not point to an empty file.`);
                }
                return bundle;
            }
            const script = this.code.content;
            return {
                script,
            };
        })());
    }
    synthesize() {
        return {
            ...super.synthesize(),
            checkType: CheckTypes.MULTI_STEP,
            playwrightConfig: this.playwrightConfig,
        };
    }
}
