// checkly/dist/constructs/multi-step-check-codegen.js
import { expr, ident } from '../sourcegen/index.js';
import { valueForPlaywrightConfig } from './playwright-config-codegen.js';
import { buildRuntimeCheckProps } from './check-codegen.js';
import { Codegen } from './internal/codegen/index.js';
const construct = 'MultiStepCheck';
export class MultiStepCheckCodegen extends Codegen {
    describe(resource) {
        return `Multistep Check: ${resource.name}`;
    }
    gencode(logicalId, resource, context) {
        const filePath = context.filePath('resources/multi-step-checks', resource.name, {
            tags: resource.tags,
            isolate: true,
            unique: true,
        });
        const file = this.program.generatedConstructFile(filePath.fullPath);
        file.namedImport(construct, 'checkly/constructs');
        file.section(expr(ident(construct), builder => {
            builder.new(builder => {
                builder.string(logicalId);
                builder.object(builder => {
                    builder.object('code', builder => {
                        const scriptFile = this.program.staticSpecFile(filePath.extless, resource.script);
                        builder.string('entrypoint', file.relativePath(scriptFile));
                    });
                    if (resource.playwrightConfig) {
                        builder.value('playwrightConfig', valueForPlaywrightConfig(resource.playwrightConfig));
                    }
                    buildRuntimeCheckProps(this.program, file, builder, resource, context);
                });
            });
        }));
    }
}
