// checkly/dist/constructs/opsgenie-alert-channel.js
import { AlertChannel } from './alert-channel.js';
import { Session } from './session.js';
/**
 * Creates an Opsgenie Alert Channel
 *
 * @remarks
 *
 * This class make use of the Alert Channel endpoints.
 */
export class OpsgenieAlertChannel extends AlertChannel {
    name;
    apiKey;
    region;
    priority;
    /**
     * Constructs the Opsgenie Alert Channel instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props Opsgenie alert channel configuration properties
     *
     * {@link https://checklyhq.com/docs/cli/constructs-reference/#opsgeniealertchannel Read more in the docs}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.name = props.name;
        this.apiKey = props.apiKey;
        this.region = props.region;
        this.priority = props.priority;
        Session.registerConstruct(this);
    }
    describe() {
        return `OpsgenieAlertChannel:${this.logicalId}`;
    }
    synthesize() {
        return {
            ...super.synthesize(),
            type: 'OPSGENIE',
            config: {
                name: this.name,
                apiKey: this.apiKey,
                region: this.region,
                priority: this.priority,
            },
        };
    }
}
