// checkly/dist/constructs/pagerduty-alert-channel.js
import { AlertChannel } from './alert-channel.js';
import { Session } from './session.js';
/**
 * Creates an Pagerduty Alert Channel
 *
 * @remarks
 *
 * This class make use of the Alert Channel endpoints.
 */
export class PagerdutyAlertChannel extends AlertChannel {
    account;
    serviceName;
    serviceKey;
    /**
     * Constructs the Pagerduty Alert Channel instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props Pagerduty alert channel configuration properties
     *
     * {@link https://checklyhq.com/docs/cli/constructs-reference/#pagerdutyalertchannel Read more in the docs}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.account = props.account;
        this.serviceName = props.serviceName;
        this.serviceKey = props.serviceKey;
        Session.registerConstruct(this);
    }
    describe() {
        return `PagerdutyAlertChannel:${this.logicalId}`;
    }
    synthesize() {
        return {
            ...super.synthesize(),
            type: 'PAGERDUTY',
            config: {
                account: this.account,
                serviceName: this.serviceName,
                serviceKey: this.serviceKey,
            },
        };
    }
}
