// checkly/dist/constructs/phone-call-alert-channel.js
import { AlertChannel } from './alert-channel.js';
import { Session } from './session.js';
/**
 * Creates a Phone Call Alert Channel
 *
 * @remarks
 *
 * This class make use of the Alert Channel endpoints.
 */
export class PhoneCallAlertChannel extends AlertChannel {
    phoneNumber;
    name;
    /**
     * Constructs the Phone Call Alert Channel instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props Phone Call alert channel configuration properties
     *
     * {@link https://www.checklyhq.com/docs/constructs/phone-call-alert-channel/ Read more in the docs}
     * {@link https://www.checklyhq.com/docs/integrations/alerts/phone-calls/#supported-countries-and-regions | List of supported countries}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.phoneNumber = props.phoneNumber;
        this.name = props.name;
        Session.registerConstruct(this);
    }
    describe() {
        return `PhoneCallAlertChannel:${this.logicalId}`;
    }
    synthesize() {
        return {
            ...super.synthesize(),
            type: 'CALL',
            config: {
                number: this.phoneNumber,
                name: this.name,
            },
        };
    }
}
