// checkly/dist/constructs/playwright-check-bundle.js
export class PlaywrightCheckBundle {
    playwrightCheck;
    groupId;
    codeBundlePath;
    browsers;
    cacheHash;
    playwrightVersion;
    installCommand;
    testCommand;
    workingDir;
    constructor(playwrightCheck, props) {
        this.playwrightCheck = playwrightCheck;
        this.groupId = props.groupId;
        this.codeBundlePath = props.codeBundlePath;
        this.browsers = props.browsers;
        this.cacheHash = props.cacheHash;
        this.playwrightVersion = props.playwrightVersion;
        this.installCommand = props.installCommand;
        this.testCommand = props.testCommand;
        this.workingDir = props.workingDir;
    }
    synthesize() {
        return {
            ...this.playwrightCheck.synthesize(),
            groupId: this.groupId,
            codeBundlePath: this.codeBundlePath,
            browsers: this.browsers,
            cacheHash: this.cacheHash,
            playwrightVersion: this.playwrightVersion,
            installCommand: this.installCommand,
            testCommand: this.testCommand,
            workingDir: this.workingDir,
        };
    }
}
