// checkly/dist/constructs/private-location-group-assignment.js
import { Construct } from './construct.js';
import { Session } from './session.js';
/**
 * Creates an Private Location Group Assignment
 *
 * @remarks
 *
 * This class make use of the Private Location Group Assignment endpoints.
 */
export class PrivateLocationGroupAssignment extends Construct {
    privateLocationId;
    groupId;
    static __checklyType = 'private-location-group-assignment';
    /**
     * Constructs the Private Location GroupAssignment instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props private location group assignment configuration properties
     */
    constructor(logicalId, props) {
        super(PrivateLocationGroupAssignment.__checklyType, logicalId);
        this.privateLocationId = props.privateLocationId;
        this.groupId = props.groupId;
        Session.registerConstruct(this);
    }
    describe() {
        return `PrivateLocationGroupAssignment:${this.logicalId}`;
    }
    allowInChecklyConfig() {
        return true;
    }
    synthesize() {
        return {
            privateLocationId: this.privateLocationId,
            groupId: this.groupId,
        };
    }
}
