// checkly/dist/constructs/private-location.js
import { Construct } from './construct.js';
import { InvalidPropertyValueDiagnostic } from './construct-diagnostics.js';
import { validatePhysicalIdIsUuid } from './internal/common-diagnostics.js';
import { Session } from './session.js';
/**
 * Creates a reference to an existing Private Location.
 *
 * References link existing resources to a project without managing them.
 *
 * Use {@link PrivateLocation.fromId()} instead of instantiating this class directly.
 */
export class PrivateLocationRef extends Construct {
    constructor(logicalId, physicalId) {
        super(PrivateLocation.__checklyType, logicalId, physicalId, false);
        Session.registerConstruct(this);
    }
    describe() {
        return `PrivateLocationRef:${this.logicalId}`;
    }
    async validate(diagnostics) {
        await super.validate(diagnostics);
        await validatePhysicalIdIsUuid(diagnostics, 'PrivateLocation', this.physicalId);
    }
    allowInChecklyConfig() {
        return true;
    }
    synthesize() {
        return null;
    }
}
const RE_SLUG = /^((?!((us(-gov)?|ap|ca|cn|eu|sa|af|me)-(central|(north|south)?(east|west)?)-\d+))[a-zA-Z0-9-]{1,30})$/;
/**
 * Creates a Private Location
 *
 * @remarks
 *
 * This class make use of the Private Location endpoints.
 *
 * {@link https://www.checklyhq.com/docs/constructs/private-location/ Read more in the docs}
 */
export class PrivateLocation extends Construct {
    name;
    slugName;
    icon;
    proxyUrl;
    static __checklyType = 'private-location';
    /**
     * Constructs the Private Location instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props private location configuration properties
     */
    constructor(logicalId, props) {
        super(PrivateLocation.__checklyType, logicalId);
        this.name = props.name;
        this.slugName = props.slugName;
        this.icon = props.icon;
        this.proxyUrl = props.proxyUrl;
        Session.registerConstruct(this);
    }
    describe() {
        return `PrivateLocation:${this.logicalId}`;
    }
    async validate(diagnostics) {
        await super.validate(diagnostics);
        if (!RE_SLUG.test(this.slugName)) {
            diagnostics.add(new InvalidPropertyValueDiagnostic('slugName', new Error(`Value must not equal any AWS location.`)));
        }
    }
    /**
     * @param id - The UUID of the existing private location
     */
    static fromId(id) {
        return new PrivateLocationRef(`private-location-${id}`, id);
    }
    allowInChecklyConfig() {
        return true;
    }
    synthesize() {
        return {
            name: this.name,
            slugName: this.slugName,
            icon: this.icon,
            proxyUrl: this.proxyUrl,
        };
    }
}
