// checkly/dist/constructs/private-location-check-assignment.js
import { Construct } from './construct.js';
import { Session } from './session.js';
/**
 * Creates an Private Location Assignment
 *
 * @remarks
 *
 * This class make use of the Private Location Assignment endpoints.
 */
export class PrivateLocationCheckAssignment extends Construct {
    privateLocationId;
    checkId;
    static __checklyType = 'private-location-check-assignment';
    /**
     * Constructs the Private Location Assignment instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props private location assignment configuration properties
     */
    constructor(logicalId, props) {
        super(PrivateLocationCheckAssignment.__checklyType, logicalId);
        this.privateLocationId = props.privateLocationId;
        this.checkId = props.checkId;
        Session.registerConstruct(this);
    }
    describe() {
        return `PrivateLocationCheckAssignment:${this.logicalId}`;
    }
    synthesize() {
        return {
            privateLocationId: this.privateLocationId,
            checkId: this.checkId,
        };
    }
}
