// checkly/dist/constructs/private-location-codegen.js
import { Codegen } from './internal/codegen/index.js';
import { decl, expr, ident } from '../sourcegen/index.js';
const construct = 'PrivateLocation';
export function valueForPrivateLocationFromId(genfile, physicalId) {
    genfile.namedImport(construct, 'checkly/constructs');
    return expr(ident(construct), builder => {
        builder.member(ident('fromId'));
        builder.call(builder => {
            builder.string(physicalId);
        });
    });
}
export class PrivateLocationCodegen extends Codegen {
    describe(resource) {
        return `Private Location: ${resource.name}`;
    }
    prepare(logicalId, resource, context) {
        const filePath = context.filePath('resources/private-locations', resource.slugName, {
            unique: true,
        });
        context.registerPrivateLocation(resource.id, resource.name, this.program.generatedConstructFile(filePath.fullPath));
    }
    gencode(logicalId, resource, context) {
        const { id, file } = context.lookupPrivateLocation(resource.id);
        file.namedImport(construct, 'checkly/constructs');
        file.section(decl(id, builder => {
            builder.variable(expr(ident(construct), builder => {
                builder.new(builder => {
                    builder.string(logicalId);
                    builder.object(builder => {
                        builder.string('name', resource.name);
                        builder.string('slugName', resource.slugName);
                        if (resource.icon) {
                            builder.string('icon', resource.icon);
                        }
                        if (resource.proxyUrl) {
                            builder.string('proxyUrl', resource.proxyUrl);
                        }
                    });
                });
            }));
            builder.export();
        }));
    }
}
