// checkly/dist/constructs/project.js
import os from 'node:os';
import PQueue from 'p-queue';
import { Construct } from './construct.js';
import { Check, HeartbeatMonitor, } from './/index.js';
import { ConstructDiagnostics, InvalidPropertyValueDiagnostic } from './construct-diagnostics.js';
import { ProjectBundle } from './project-bundle.js';
import { Session } from './session.js';
// Cap how many constructs bundle concurrently. Bundling parses and resolves
// each check's dependency tree, so an unbounded fan-out over hundreds of checks
// can spike heap usage badly. Scale with CPU count but keep an upper bound so
// peak memory stays predictable on high-core machines.
const BUNDLE_CONCURRENCY = Math.max(4, Math.min(os.cpus().length, 16));
export class Project extends Construct {
    name;
    repoUrl;
    logicalId;
    testOnlyAllowed = false;
    data = {
        'check': {},
        'check-group': {},
        'alert-channel': {},
        'alert-channel-subscription': {},
        'maintenance-window': {},
        'private-location': {},
        'private-location-check-assignment': {},
        'private-location-group-assignment': {},
        'dashboard': {},
        'status-page': {},
        'status-page-service': {},
    };
    static __checklyType = 'project';
    /**
     * Constructs the Project instance
     *
     * @param logicalId unique project identifier
     * @param props project configuration properties
     */
    constructor(logicalId, props) {
        super(Project.__checklyType, logicalId);
        this.name = props.name;
        this.repoUrl = props.repoUrl;
        this.logicalId = logicalId;
    }
    describe() {
        return `Project:${this.logicalId}`;
    }
    async validate(diagnostics) {
        await super.validate(diagnostics);
        if (!this.name) {
            diagnostics.add(new InvalidPropertyValueDiagnostic('name', new Error(`Value must not be empty.`)));
        }
        const data = this.data;
        const constructDiagnostics = await Promise.all(Object.entries(data).flatMap(([, records]) => {
            return Object.values(records).map(async (construct) => {
                const diagnostics = new ConstructDiagnostics(construct);
                await construct.validate(diagnostics);
                return diagnostics;
            });
        }));
        diagnostics.extend(...constructDiagnostics);
    }
    allowTestOnly(enabled) {
        this.testOnlyAllowed = enabled;
    }
    addResource(type, logicalId, resource) {
        const existingResource = this.data[type][logicalId];
        if (existingResource) {
            // Non-member resources (i.e. references) can be used multiple times.
            // Behind the scenes, we'll create a single mapping for them, and the
            // referenced resource isn't managed by the project at all.
            if (!resource.member && !existingResource.member && existingResource.physicalId === resource.physicalId) {
                return;
            }
            throw new Error(`Resource of type '${type}' with logical id '${logicalId}' already exists.`);
        }
        this.data[type][logicalId] = resource;
    }
    async bundle(bundler) {
        const data = {
            ...this.data,
            // Filter out testOnly checks before bundling.
            check: Object.fromEntries(Object.entries(this.data.check)
                .filter(([, check]) => !check.testOnly || this.testOnlyAllowed)
                .filter(([, check]) => Session.checkFilter?.(check) ?? true)),
        };
        const dataBundle = Object.fromEntries(Object.entries(data).map(([type]) => {
            return [type, {}];
        }));
        // Bundle constructs through a bounded queue rather than an unbounded
        // Promise.all. Each task writes a distinct [type][logicalId] slot, so the
        // concurrent writes don't race; the final assembly is order-independent.
        const queue = new PQueue({ concurrency: BUNDLE_CONCURRENCY });
        await queue.addAll(Object.values(data).flatMap(records => {
            return Object.values(records).map(construct => async () => {
                const bundle = await construct.bundle(bundler);
                const { type, logicalId } = construct;
                dataBundle[type][logicalId] = { construct, bundle };
            });
        }));
        return new ProjectBundle(this, dataBundle);
    }
    synthesize() {
        const project = {
            logicalId: this.logicalId,
            name: this.name,
            repoUrl: this.repoUrl,
        };
        return {
            project,
            sharedFiles: Session.sharedFiles,
        };
    }
    getTestOnlyConstructs() {
        return Object
            .values(this.data)
            .flatMap((record) => Object
            .values(record)
            .filter((construct) => construct instanceof Check && construct.testOnly));
    }
    getHeartbeatLogicalIds() {
        return Object
            .values(this.data.check)
            .filter((construct) => construct instanceof HeartbeatMonitor)
            .map((construct) => construct.logicalId);
    }
}
