// checkly/dist/constructs/retry-strategy-codegen.js
import { expr, ident } from '../sourcegen/index.js';
export function valueForRetryStrategy(genfile, strategy) {
    genfile.namedImport('RetryStrategyBuilder', 'checkly/constructs');
    function buildBaseBackoffSecondsOption(options, builder) {
        if (options.baseBackoffSeconds) {
            builder.number('baseBackoffSeconds', options.baseBackoffSeconds);
        }
    }
    function buildMaxRetriesOption(options, builder) {
        if (options.maxRetries) {
            builder.number('maxRetries', options.maxRetries);
        }
    }
    function buildMaxDurationSecondsOption(options, builder) {
        if (options.maxDurationSeconds) {
            builder.number('maxDurationSeconds', options.maxDurationSeconds);
        }
    }
    function buildSameRegionOption(options, builder) {
        if (options.sameRegion !== undefined) {
            builder.boolean('sameRegion', options.sameRegion);
        }
    }
    function buildOnlyOnOption(options, builder) {
        if (options.onlyOn !== undefined) {
            const onlyOn = Array.isArray(options.onlyOn) ? options.onlyOn : [options.onlyOn];
            if (onlyOn.length === 1) {
                builder.string('onlyOn', onlyOn[0]);
            }
            else {
                builder.array('onlyOn', builder => {
                    for (const condition of onlyOn) {
                        builder.string(condition);
                    }
                });
            }
        }
    }
    function buildCommonOptions(options, builder) {
        buildBaseBackoffSecondsOption(options, builder);
        buildMaxRetriesOption(options, builder);
        buildMaxDurationSecondsOption(options, builder);
        buildSameRegionOption(options, builder);
        buildOnlyOnOption(options, builder);
    }
    if (strategy === null || strategy === undefined) {
        return expr(ident('RetryStrategyBuilder'), builder => {
            builder.member(ident('noRetries'));
            builder.call(builder => {
                builder.empty();
            });
        });
    }
    switch (strategy.type) {
        case 'FIXED':
            return expr(ident('RetryStrategyBuilder'), builder => {
                builder.member(ident('fixedStrategy'));
                builder.call(builder => {
                    builder.object(builder => {
                        buildCommonOptions(strategy, builder);
                    });
                });
            });
        case 'LINEAR':
            return expr(ident('RetryStrategyBuilder'), builder => {
                builder.member(ident('linearStrategy'));
                builder.call(builder => {
                    builder.object(builder => {
                        buildCommonOptions(strategy, builder);
                    });
                });
            });
        case 'EXPONENTIAL':
            return expr(ident('RetryStrategyBuilder'), builder => {
                builder.member(ident('exponentialStrategy'));
                builder.call(builder => {
                    builder.object(builder => {
                        buildCommonOptions(strategy, builder);
                    });
                });
            });
        case 'SINGLE_RETRY':
            return expr(ident('RetryStrategyBuilder'), builder => {
                builder.member(ident('singleRetry'));
                builder.call(builder => {
                    builder.object(builder => {
                        buildBaseBackoffSecondsOption(strategy, builder);
                        buildSameRegionOption(strategy, builder);
                        buildOnlyOnOption(strategy, builder);
                    });
                });
            });
        case 'NO_RETRIES':
            return expr(ident('RetryStrategyBuilder'), builder => {
                builder.member(ident('noRetries'));
                builder.call(builder => {
                    builder.empty();
                });
            });
        default:
            throw new Error(`Unsupported retry strategy type ${strategy.type}`);
    }
}
