// checkly/dist/constructs/session.js
import path from 'node:path';
import * as api from '../rest/api.js';
import { Parser } from '../services/check-parser/parser.js';
import { Construct } from './construct.js';
import { ValidationError } from './validator-error.js';
import { JitiFileLoader, MixedFileLoader, NativeFileLoader, UnsupportedFileLoaderError, } from '../loader/index.js';
import { pathToPosix } from '../services/util.js';
import { npmPackageManager } from '../services/check-parser/package-files/package-manager.js';
import { Err } from '../services/check-parser/package-files/result.js';
import { PlaywrightProjectBundler } from '../services/playwright-project-bundler.js';
import { PROJECT_CONSTRUCT_TYPE } from '../constants.js';
export class Session {
    static loader = new MixedFileLoader(new NativeFileLoader(), new JitiFileLoader());
    static project;
    static basePath;
    static contextPath;
    static checkDefaults;
    static checkFilter;
    static browserCheckDefaults;
    static multiStepCheckDefaults;
    static checkFilePath;
    static checkFileAbsolutePath;
    static availableRuntimes;
    static defaultRuntimeId;
    static verifyRuntimeDependencies = true;
    static loadingChecklyConfigFile;
    static checklyConfigFileConstructs;
    static privateLocations;
    static parsers = new Map();
    static playwrightProjectBundler;
    static constructExports = [];
    static ignoreDirectoriesMatch = [];
    static warnOnWebServerConfig;
    static packageManager = npmPackageManager;
    static workspace = Err(new Error(`Workspace support not initialized`));
    static detectedEnginePromise;
    static reset() {
        this.project = undefined;
        this.basePath = undefined;
        this.contextPath = undefined;
        this.checkDefaults = undefined;
        this.checkFilter = undefined;
        this.browserCheckDefaults = undefined;
        this.multiStepCheckDefaults = undefined;
        this.checkFilePath = undefined;
        this.checkFileAbsolutePath = undefined;
        this.availableRuntimes = {};
        this.defaultRuntimeId = undefined;
        this.verifyRuntimeDependencies = true;
        this.loadingChecklyConfigFile = false;
        this.checklyConfigFileConstructs = undefined;
        this.privateLocations = [];
        this.parsers = new Map();
        this.playwrightProjectBundler = undefined;
        this.constructExports = [];
        this.ignoreDirectoriesMatch = [];
        this.warnOnWebServerConfig = false;
        this.packageManager = npmPackageManager;
        this.workspace = Err(new Error(`Workspace support not initialized`));
        this.detectedEnginePromise = undefined;
        this.resetSharedFiles();
    }
    static async loadFile(filePath) {
        const loader = this.loader;
        if (loader === undefined) {
            throw new Error(`Session has no loader set`);
        }
        if (!loader.isAuthoritativeFor(filePath)) {
            throw new Error(`Unable to find a compatible loader for file '${filePath}'`);
        }
        try {
            const moduleExports = await loader.loadFile(filePath);
            // Register all exported constructs we find.
            for (const [exportName, value] of Object.entries(moduleExports ?? {})) {
                if (value instanceof Construct) {
                    this.constructExports.push({
                        type: value.type,
                        logicalId: value.logicalId,
                        filePath,
                        exportName,
                    });
                }
            }
            const defaultExport = moduleExports?.default ?? moduleExports;
            if (typeof defaultExport === 'function') {
                return await defaultExport();
            }
            return defaultExport;
        }
        catch (err) {
            if (err instanceof UnsupportedFileLoaderError && /\.[cm]?ts$/.test(filePath)) {
                throw new Error(`Unable to load the TypeScript file '${filePath}'.`, { cause: err });
            }
            throw new Error(`Error loading file '${filePath}'\n${err.stack}`, { cause: err });
        }
    }
    static registerConstruct(construct) {
        if (Session.project) {
            Session.project.addResource(construct.type, construct.logicalId, construct);
        }
        else if (Session.loadingChecklyConfigFile && construct.allowInChecklyConfig()) {
            Session.checklyConfigFileConstructs.push(construct);
        }
        else {
            throw new Error('Internal Error: Session is not properly configured for using a construct. Please contact Checkly support at support@checklyhq.com.');
        }
    }
    static validateCreateConstruct(construct) {
        if (typeof construct.logicalId !== 'string') {
            throw new ValidationError(`The "logicalId" of a ${construct.type} construct must be a string (logicalId=${construct.logicalId} [${typeof construct.logicalId}])`);
        }
        if (construct.type === PROJECT_CONSTRUCT_TYPE) {
            // Creating the construct is allowed - We're creating the project.
        }
        else if (Session.project) {
            // Creating the construct is allowed - We're in the process of parsing the project.
        }
        else if (Session.loadingChecklyConfigFile && construct.allowInChecklyConfig()) {
            // Creating the construct is allowed - We're in the process of parsing the Checkly config.
        }
        else if (Session.loadingChecklyConfigFile) {
            throw new Error(`Creating a ${construct.constructor.name} construct in the Checkly config file isn't supported.`);
        }
        else {
            throw new Error(`Unable to create a construct '${construct.constructor.name}' outside a Checkly CLI project.`);
        }
    }
    static async getPrivateLocations() {
        if (!Session.privateLocations) {
            const { data: privateLocations } = await api.privateLocations.getAll();
            Session.privateLocations = privateLocations;
        }
        return Session.privateLocations;
    }
    static getRuntime(runtimeId) {
        const effectiveRuntimeId = runtimeId ?? Session.defaultRuntimeId;
        if (effectiveRuntimeId === undefined) {
            throw new Error('Internal Error: Account default runtime is not set. Please contact Checkly support at support@checklyhq.com.');
        }
        return Session.availableRuntimes[effectiveRuntimeId];
    }
    static #getOrInitParser(cacheKey, init) {
        const existingParser = Session.parsers.get(cacheKey);
        if (existingParser !== undefined) {
            return existingParser;
        }
        const newParser = init();
        Session.parsers.set(cacheKey, newParser);
        return newParser;
    }
    static getParser(runtime) {
        return this.#getOrInitParser(`runtime:${runtime.name}`, () => {
            return new Parser({
                supportedNpmModules: Object.keys(runtime.dependencies),
                checkUnsupportedModules: Session.verifyRuntimeDependencies,
                workspace: Session.workspace.ok(),
            });
        });
    }
    static getPlaywrightParser() {
        return this.#getOrInitParser(`playwright`, () => {
            return new Parser({
                checkUnsupportedModules: false,
                workspace: Session.workspace.ok(),
                restricted: false,
            });
        });
    }
    static getPlaywrightProjectBundler() {
        if (this.playwrightProjectBundler === undefined) {
            this.playwrightProjectBundler = new PlaywrightProjectBundler();
        }
        return this.playwrightProjectBundler;
    }
    static relativePosixPath(filePath) {
        return pathToPosix(path.relative(Session.basePath, filePath));
    }
    static contextRelativePosixPath(filePath) {
        return pathToPosix(path.relative(Session.contextPath, filePath));
    }
    static sharedFileRefs = new Map();
    static sharedFiles = [];
    static registerSharedFile(file) {
        const ref = Session.sharedFileRefs.get(file.path);
        if (ref !== undefined) {
            return ref;
        }
        const newRef = Session.sharedFiles.push(file) - 1;
        Session.sharedFileRefs.set(file.path, newRef);
        return newRef;
    }
    static resetSharedFiles() {
        Session.sharedFileRefs.clear();
        Session.sharedFiles.splice(0);
    }
}
