// checkly/dist/constructs/slack-alert-channel.js
import { AlertChannel } from './alert-channel.js';
import { Session } from './session.js';
/**
 * Creates an Slack Alert Channel
 *
 * @remarks
 *
 * This class make use of the Alert Channel endpoints.
 */
export class SlackAlertChannel extends AlertChannel {
    url;
    channel;
    /**
     * Constructs the Slack Alert Channel instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props Slack alert channel configuration properties
     *
     * {@link https://www.checklyhq.com/docs/constructs/slack-alert-channel/ Read more in the docs}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.url = props.url;
        this.channel = props.channel;
        Session.registerConstruct(this);
    }
    describe() {
        return `SlackAlertChannel:${this.logicalId}`;
    }
    synthesize() {
        return {
            ...super.synthesize(),
            type: 'SLACK',
            config: {
                url: this.url,
                channel: this.channel,
            },
        };
    }
}
