// checkly/dist/constructs/sms-alert-channel.js
import { AlertChannel } from './alert-channel.js';
import { Session } from './session.js';
/**
 * Creates a SMS Alert Channel
 *
 * @remarks
 *
 * This class make use of the Alert Channel endpoints.
 */
export class SmsAlertChannel extends AlertChannel {
    phoneNumber;
    name;
    /**
     * Constructs the SMS Alert Channel instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props SMS alert channel configuration properties
     *
     * {@link https://www.checklyhq.com/docs/constructs/sms-alert-channel/ Read more in the docs}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.phoneNumber = props.phoneNumber;
        this.name = props.name;
        Session.registerConstruct(this);
    }
    describe() {
        return `SmsAlertChannel:${this.logicalId}`;
    }
    synthesize() {
        return {
            ...super.synthesize(),
            type: 'SMS',
            config: {
                number: this.phoneNumber,
                name: this.name,
            },
        };
    }
}
