// checkly/dist/constructs/status-page.js
import { Construct } from './construct.js';
import { Session } from './session.js';
import { Ref } from './ref.js';
/**
 * Creates a Status Page
 */
export class StatusPage extends Construct {
    name;
    cards;
    url;
    customDomain;
    logo;
    redirectTo;
    favicon;
    defaultTheme;
    static __checklyType = 'status-page';
    /**
     * Constructs the Status Page instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props status page configuration properties
     *
     * {@link https://www.checklyhq.com/docs/constructs/status-page/ Read more in the docs}
     */
    constructor(logicalId, props) {
        super(StatusPage.__checklyType, logicalId);
        this.name = props.name;
        this.url = props.url;
        this.cards = props.cards;
        this.customDomain = props.customDomain;
        this.logo = props.logo;
        this.redirectTo = props.redirectTo;
        this.favicon = props.favicon;
        this.defaultTheme = props.defaultTheme;
        Session.registerConstruct(this);
    }
    describe() {
        return `StatusPage:${this.logicalId}`;
    }
    synthesize() {
        return {
            name: this.name,
            url: this.url,
            customDomain: this.customDomain,
            logo: this.logo,
            redirectTo: this.redirectTo,
            favicon: this.favicon,
            defaultTheme: this.defaultTheme,
            cards: this.cards.map(card => ({
                name: card.name,
                services: card
                    .services
                    ?.map(service => Ref.from(service.logicalId)),
            })),
        };
    }
}
