// checkly/dist/constructs/status-page-service-codegen.js
import { Codegen } from './internal/codegen/index.js';
import { decl, expr, ident } from '../sourcegen/index.js';
const construct = 'StatusPageService';
export function valueForStatusPageServiceFromId(genfile, physicalId) {
    genfile.namedImport(construct, 'checkly/constructs');
    return expr(ident(construct), builder => {
        builder.member(ident('fromId'));
        builder.call(builder => {
            builder.string(physicalId);
        });
    });
}
export class StatusPageServiceCodegen extends Codegen {
    describe(resource) {
        return `Status Page Service: ${resource.name}`;
    }
    prepare(logicalId, resource, context) {
        const filePath = context.filePath('resources/status-pages/services', resource.name, {
            unique: true,
        });
        context.registerStatusPageService(resource.id, resource.name, this.program.generatedConstructFile(filePath.fullPath));
    }
    gencode(logicalId, resource, context) {
        const { id, file } = context.lookupStatusPageService(resource.id);
        file.namedImport(construct, 'checkly/constructs');
        file.section(decl(id, builder => {
            builder.variable(expr(ident(construct), builder => {
                builder.new(builder => {
                    builder.string(logicalId);
                    builder.object(builder => {
                        builder.string('name', resource.name);
                    });
                });
            }));
            builder.export();
        }));
    }
}
