// checkly/dist/constructs/tcp-monitor.js
import { Monitor } from './monitor.js';
import { Session } from './session.js';
import { NumericAssertionBuilder, GeneralAssertionBuilder } from './internal/assertion.js';
import { validateResponseTimes } from './internal/common-diagnostics.js';
/**
 * Builder class for creating TCP monitor assertions.
 * Provides methods to create assertions for TCP connection responses.
 *
 * @example
 * ```typescript
 * // Response time assertions
 * TcpAssertionBuilder.responseTime().lessThan(1000)
 * TcpAssertionBuilder.responseTime().greaterThan(100)
 *
 * // Response data assertions
 * TcpAssertionBuilder.responseData().contains('SMTP')
 * TcpAssertionBuilder.responseData().notContains('error')
 * ```
 */
export class TcpAssertionBuilder {
    /**
     * Creates an assertion builder for TCP response data.
     * @param property Optional property path for response data
     * @returns A general assertion builder for response data content
     */
    static responseData(property) {
        return new GeneralAssertionBuilder('RESPONSE_DATA', property);
    }
    /**
     * Creates an assertion builder for TCP response time.
     * @returns A numeric assertion builder for response time in milliseconds
     */
    static responseTime() {
        return new NumericAssertionBuilder('RESPONSE_TIME');
    }
}
/**
 * Creates a TCP Monitor
 */
export class TcpMonitor extends Monitor {
    request;
    degradedResponseTime;
    maxResponseTime;
    /**
     * Constructs the TCP Monitor instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props configuration properties
     *
     * {@link https://www.checklyhq.com/docs/constructs/tcp-monitor/ Read more in the docs}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.request = props.request;
        this.degradedResponseTime = props.degradedResponseTime;
        this.maxResponseTime = props.maxResponseTime;
        Session.registerConstruct(this);
        this.addSubscriptions();
        this.addPrivateLocationCheckAssignments();
    }
    describe() {
        return `TcpMonitor:${this.logicalId}`;
    }
    async validate(diagnostics) {
        await super.validate(diagnostics);
        await validateResponseTimes(diagnostics, this, {
            degradedResponseTime: 5_000,
            maxResponseTime: 5_000,
        });
    }
    synthesize() {
        return {
            ...super.synthesize(),
            checkType: 'TCP',
            request: this.request,
            degradedResponseTime: this.degradedResponseTime,
            maxResponseTime: this.maxResponseTime,
        };
    }
}
// Aliases for backwards compatibility.
export { TcpMonitor as TcpCheck, };
