// checkly/dist/constructs/webhook-alert-channel.js
import { AlertChannel } from './alert-channel.js';
import { Session } from './session.js';
/**
 * Creates an Webhook Alert Channel
 *
 * @remarks
 *
 * This class make use of the Alert Channel endpoints.
 */
export class WebhookAlertChannel extends AlertChannel {
    name;
    webhookType;
    url;
    template;
    method;
    headers;
    queryParameters;
    webhookSecret;
    /**
     * Constructs the Webhook Alert Channel instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props Webhook alert channel configuration properties
     *
     * {@link https://www.checklyhq.com/docs/constructs/webhook-alert-channel/ Read more in the docs}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.name = props.name;
        this.webhookType = props.webhookType;
        this.url = props.url;
        this.template = props.template;
        this.method = props.method;
        this.headers = props.headers;
        this.queryParameters = props.queryParameters;
        this.webhookSecret = props.webhookSecret;
        Session.registerConstruct(this);
    }
    describe() {
        return `WebhookAlertChannel:${this.logicalId}`;
    }
    synthesize() {
        return {
            ...super.synthesize(),
            type: 'WEBHOOK',
            config: {
                name: this.name,
                webhookType: this.webhookType,
                url: this.url,
                template: this.template,
                method: this.method,
                headers: this.headers,
                queryParameters: this.queryParameters,
                webhookSecret: this.webhookSecret,
            },
        };
    }
}
