// checkly/dist/services/check-parser/bundler.js
import { createReadStream, createWriteStream } from 'node:fs';
import fs from 'node:fs/promises';
import { tmpdir } from 'node:os';
import path from 'node:path';
import Debug from 'debug';
import * as uuid from 'uuid';
import { checklyStorage } from '../../rest/api.js';
import { computeWorkspaceCacheHash } from './cache-hash.js';
const debug = Debug('checkly:cli:services:check-parser:bundler');
export class BundleArchive {
    static TMPDIR_PREFIX = 'cli-';
    static DEFAULT_FILENAME = 'playwright-project.tar.gz';
    #tempDir;
    #archiveFile;
    #archiveFileWriteStream;
    #stripPrefix;
    #archive;
    constructor(options) {
        const { tempDir, archiveFile, archiveFileWriteStream, stripPrefix, archive, } = options;
        this.#tempDir = tempDir;
        this.#archiveFile = archiveFile;
        this.#archiveFileWriteStream = archiveFileWriteStream;
        this.#stripPrefix = stripPrefix;
        this.#archive = archive;
    }
    static async create(options) {
        debug(`Creating bundle archive`);
        const { tempDir: maybeTempDir, stripPrefix, } = options;
        const tempDir = maybeTempDir
            ? await fs.mkdir(maybeTempDir, { recursive: true }).then(() => maybeTempDir)
            // tmpdir() on macOS usually returns a path starting with /var which is
            // a symlink. Resolve the path so that we don't run into path mismatch
            // issues.
            : await fs.realpath(await fs.mkdtemp(path.join(tmpdir(), BundleArchive.TMPDIR_PREFIX)));
        debug(`Using temporary directory ${tempDir}`);
        const archiveFile = path.join(tempDir, BundleArchive.DEFAULT_FILENAME);
        debug(`Using archive file ${archiveFile}`);
        const archive = await createArchiver();
        const output = createWriteStream(archiveFile);
        archive.pipe(output);
        return new BundleArchive({
            tempDir,
            archiveFile,
            archive,
            archiveFileWriteStream: output,
            stripPrefix,
        });
    }
    // eslint-disable-next-line require-await
    async add(...files) {
        for (const file of files) {
            const name = this.#stripPrefix
                ? path.relative(this.#stripPrefix, file.filePath)
                : file.filePath;
            const entry = {
                mode: 0o755, // Default mode for files in the archive
                name,
            };
            if (file.physical) {
                this.#archive.file(file.filePath, entry);
            }
            else {
                this.#archive.append(file.content, entry);
            }
        }
    }
    async finalize() {
        await this.#archive.finalize();
        await new Promise((resolve, reject) => {
            this.#archiveFileWriteStream.on('close', resolve);
            this.#archiveFileWriteStream.on('error', reject);
        });
        return await FinalizedBundleArchive.create({
            archiveFile: this.#archiveFile,
        });
    }
    async destroy() {
        debug(`Destroying root ${this.#tempDir}`);
        await fs.rm(this.#tempDir, {
            recursive: true,
            force: true,
        });
    }
}
export class FinalizedBundleArchive {
    #archiveFile;
    constructor(options) {
        const { archiveFile, } = options;
        this.#archiveFile = archiveFile;
    }
    // eslint-disable-next-line require-await
    static async create(options) {
        return new FinalizedBundleArchive(options);
    }
    get archiveFile() {
        return this.#archiveFile;
    }
    async store() {
        const { data: { key, }, } = await this.#uploadCodeBundle(this.#archiveFile);
        return await RemoteBundleArchive.create({
            key,
        });
    }
    async #uploadCodeBundle(filePath) {
        const { size } = await fs.stat(filePath);
        const stream = createReadStream(filePath);
        stream.on('error', err => {
            throw new Error(`Failed to read Playwright project file: ${err.message}`);
        });
        return checklyStorage.uploadCodeBundle(stream, size);
    }
}
export class RemoteBundleArchive {
    #key;
    constructor(options) {
        const { key, } = options;
        this.#key = key;
    }
    // eslint-disable-next-line require-await
    static async create(options) {
        return new RemoteBundleArchive(options);
    }
    get key() {
        return this.#key;
    }
}
export class Bundler {
    #id;
    #marker;
    #cacheHash;
    #tempDir;
    #stripPrefix;
    #files = new Map();
    constructor(options) {
        const { tempDir, cacheHash, stripPrefix, } = options;
        this.#id = uuid.v4();
        this.#marker = new BundlePathMarker(`bundle:${this.#id}`);
        this.#cacheHash = cacheHash;
        this.#stripPrefix = stripPrefix;
        this.#tempDir = tempDir;
    }
    // eslint-disable-next-line require-await
    static async create(options) {
        debug(`Creating bundler`);
        return new Bundler(options);
    }
    static async createForWorkspace(workspace, options = {}) {
        debug(`Creating bundler for workspace`);
        const { tempDir, } = options;
        const cacheHash = await computeWorkspaceCacheHash(workspace);
        return new Bundler({
            tempDir,
            cacheHash,
            stripPrefix: workspace?.root.path,
        });
    }
    get marker() {
        return this.#marker;
    }
    updateMarker(newValue) {
        this.#marker.updateValue(newValue);
    }
    get cacheHash() {
        return this.#cacheHash;
    }
    registerFiles(...files) {
        for (const newFile of files) {
            const existingFile = this.#files.get(newFile.filePath);
            if (existingFile) {
                // Prefer physical files.
                if (existingFile.physical && !newFile.physical) {
                    continue;
                }
            }
            this.#files.set(newFile.filePath, newFile);
        }
    }
    async finalize() {
        const archive = await BundleArchive.create({
            tempDir: this.#tempDir,
            stripPrefix: this.#stripPrefix,
        });
        const files = Array.from(this.#files.values());
        files.sort((a, b) => {
            return a.filePath.localeCompare(b.filePath);
        });
        await archive.add(...files);
        return await archive.finalize();
    }
}
async function createArchiver() {
    // Dynamic import for CommonJs so it doesn't break when using checkly/playwright-reporter archiver
    // The custom Checkly fork of archiver exports TarArchive class instead of a default function
    const archiverModule = await import('archiver');
    if (archiverModule.TarArchive) {
        // Using Checkly's custom fork which exports TarArchive class
        return new archiverModule.TarArchive({ gzip: true, gzipOptions: { level: 9 } });
    }
    else if (archiverModule.default) {
        // Using standard archiver which has a default factory function
        return archiverModule.default('tar', { gzip: true, gzipOptions: { level: 9 } });
    }
    else {
        throw new Error('Unable to initialize archiver: neither TarArchive nor default export found');
    }
}
export class BundlePathMarker {
    #value;
    constructor(initialValue) {
        this.#value = initialValue;
    }
    updateValue(newValue) {
        this.#value = newValue;
    }
    toJSON() {
        return this.#value;
    }
}
