// checkly/dist/services/check-parser/collector.js
import { DependencyParseError } from './errors.js';
export class Collector {
    entrypoint;
    entrypointContent;
    missingFiles = [];
    parseErrors = [];
    unsupportedNpmDependencies = [];
    dependencies = new Map();
    constructor(entrypoint, entrypointContent) {
        this.entrypoint = entrypoint;
        this.entrypointContent = entrypointContent;
    }
    hasDependency(path) {
        return this.dependencies.has(path);
    }
    addDependency(path, content) {
        this.dependencies.set(path, content);
    }
    addUnsupportedNpmDependencies(file, unsupportedDependencies) {
        this.unsupportedNpmDependencies.push({ file, unsupportedDependencies });
    }
    addParsingError(file, message) {
        this.parseErrors.push({ file, error: message });
    }
    addMissingFile(filePath) {
        this.missingFiles.push(filePath);
    }
    validate() {
        if (this.missingFiles.length || this.parseErrors.length || this.unsupportedNpmDependencies.length) {
            throw new DependencyParseError(this.entrypoint, this.missingFiles, this.unsupportedNpmDependencies, this.parseErrors);
        }
    }
    getItems() {
        return {
            entrypoint: {
                filePath: this.entrypoint,
                content: this.entrypointContent,
            },
            dependencies: Array.from(this.dependencies.entries(), ([key, value]) => ({
                filePath: key,
                content: value,
            })),
        };
    }
}
