// checkly/dist/services/check-parser/parser.js
import fs from 'node:fs/promises';
import { createRequire } from 'node:module';
import path from 'node:path';
import * as acorn from 'acorn';
import * as walk from 'acorn-walk';
import Debug from 'debug';
import { Collector } from './collector.js';
import { DependencyParseError } from './errors.js';
import { PackageFilesResolver } from './package-files/resolver.js';
import { pathToPosix } from '../util.js';
import { isCoreExtension, isTSExtension } from './package-files/extension.js';
import { createFauxPackageFiles } from './faux-package.js';
import { PlaywrightConfigExpander } from './playwright-config-expander.js';
const debug = Debug('checkly:cli:services:check-parser:parser');
// Our custom configuration to handle walking errors
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const ignore = (_node, _st, _c) => { };
function isLegacySupportedFileExtension(value) {
    switch (value) {
        case '.js':
        case '.mjs':
        case '.ts':
            return true;
        default:
            return false;
    }
}
const supportedBuiltinModules = [
    'node:assert',
    'node:buffer',
    'node:crypto',
    'node:dns',
    'node:fs',
    'node:path',
    'node:querystring',
    'node:readline',
    'node:stream',
    'node:string_decoder',
    'node:timers',
    'node:tls',
    'node:url',
    'node:util',
    'node:zlib',
];
let tsParser;
function getTsParser() {
    if (tsParser) {
        return tsParser;
    }
    try {
        const require = createRequire(import.meta.url);
        tsParser = require('@typescript-eslint/typescript-estree');
        const AST_NODE_TYPES = tsParser.AST_NODE_TYPES;
        // Our custom configuration to handle walking errors
        Object.values(AST_NODE_TYPES).forEach(astType => {
            // Only handle the TS/JSX specific ones
            if (astType.startsWith('TS') || astType.startsWith('JSX')) {
                const base = walk.base;
                base[astType] = base[astType] ?? ignore;
            }
        });
        return tsParser;
    }
    catch (err) {
        if (err.code === 'ERR_MODULE_NOT_FOUND' || err.code === 'MODULE_NOT_FOUND') {
            throw new Error('Please install "typescript" to use TypeScript-code in check files', { cause: err });
        }
        throw err;
    }
}
class RawDependencyCollector {
    #dependencies = [];
    add(dependency) {
        this.#dependencies.push(dependency);
    }
    state() {
        return this.#dependencies;
    }
}
class NeighborDependencyCollector {
    #depended = new Set();
    #referenced = new Set();
    markDepended(...pkg) {
        for (const p of pkg) {
            this.#depended.add(p);
        }
    }
    markReferenced(...pkg) {
        for (const p of pkg) {
            this.markDepended(p);
            this.#referenced.add(p);
        }
    }
    unreferencedDependedPackages() {
        return Array.from(this.#depended).filter(pkg => !this.#referenced.has(pkg));
    }
}
const FILEOP_PARSE = 1;
const FILEOP_RESOLVE = 2;
export class Parser {
    supportedModules;
    checkUnsupportedModules;
    workspace;
    resolver;
    restricted;
    #parseCache = new Map();
    #resolveCache = new Map();
    #configExpander = new PlaywrightConfigExpander();
    // TODO: pass a npm matrix of supported npm modules
    // Maybe pass a cache so we don't have to fetch files separately all the time
    constructor(options) {
        this.restricted = options.restricted ?? true;
        this.supportedModules = new Set(supportedBuiltinModules.concat(options.supportedNpmModules ?? []));
        this.checkUnsupportedModules = options.checkUnsupportedModules ?? true;
        this.workspace = options.workspace;
        this.resolver = new PackageFilesResolver({
            workspace: options.workspace,
            restricted: this.restricted,
        });
    }
    supportsModule(importPath) {
        if (this.supportedModules.has(importPath)) {
            return true;
        }
        if (this.supportedModules.has('node:' + importPath)) {
            return true;
        }
        return false;
    }
    validateEntrypoint(filePath) {
        const ops = this.determineFileOps(filePath);
        if ((ops & FILEOP_PARSE) === 0) {
            throw new Error(`Unsupported file extension for ${filePath}`);
        }
    }
    async readFile(filePath) {
        debug(`Reading file ${filePath}`);
        try {
            const content = await fs.readFile(filePath, { encoding: 'utf-8' });
            return { filePath, content };
        }
        catch (err) {
            debug(`Failed to read file ${filePath}: ${err}`);
            throw new DependencyParseError(filePath, [filePath], [], []);
        }
    }
    determineFileOps(filePath) {
        const extension = path.extname(filePath);
        if (this.restricted) {
            if (isLegacySupportedFileExtension(extension)) {
                return FILEOP_RESOLVE | FILEOP_PARSE;
            }
            return 0;
        }
        if (isCoreExtension(extension)) {
            if (extension === '.json') {
                return FILEOP_RESOLVE;
            }
            return FILEOP_RESOLVE | FILEOP_PARSE;
        }
        if (isTSExtension(extension)) {
            return FILEOP_RESOLVE | FILEOP_PARSE;
        }
        return 0;
    }
    async getFilesAndDependencies(playwrightConfig) {
        const files = new Set(await this.#configExpander.findTestFiles(playwrightConfig));
        files.add(playwrightConfig.configFilePath);
        const errors = new Set();
        const missingFiles = new Set();
        const resultFileSet = new Set();
        const neighbors = new NeighborDependencyCollector();
        for (const filePath of files) {
            if (resultFileSet.has(filePath)) {
                continue;
            }
            const ops = this.determineFileOps(filePath);
            // Not necessary - just a minor optimization.
            if (ops === 0) {
                resultFileSet.add(filePath);
                continue;
            }
            let dependencies = [];
            if (ops & FILEOP_PARSE) {
                const { module, error } = await this.parseFile(filePath);
                if (error) {
                    errors.add(filePath);
                    continue;
                }
                dependencies = module.dependencies;
            }
            if (ops & FILEOP_RESOLVE) {
                const resolvedDependencies = await this.resolveDependencies(filePath, dependencies);
                for (const dep of resolvedDependencies.missing) {
                    missingFiles.add(dep.filePath);
                }
                neighbors.markDepended(...resolvedDependencies.neighbors.depends);
                neighbors.markReferenced(...resolvedDependencies.neighbors.references);
                for (const dep of resolvedDependencies.local) {
                    if (resultFileSet.has(dep.sourceFile.meta.filePath)) {
                        continue;
                    }
                    const filePath = dep.sourceFile.meta.filePath;
                    files.add(filePath);
                }
            }
            resultFileSet.add(filePath);
        }
        if (missingFiles.size) {
            throw new DependencyParseError([].join(', '), Array.from(missingFiles), [], []);
        }
        const outputFiles = [];
        for (const filePath of resultFileSet) {
            outputFiles.push({
                filePath: pathToPosix(filePath),
                physical: true,
            });
        }
        const neighborFiles = neighbors.unreferencedDependedPackages()
            .flatMap(pkg => createFauxPackageFiles(pkg));
        outputFiles.push(...neighborFiles);
        outputFiles.sort((a, b) => {
            return a.filePath.localeCompare(b.filePath);
        });
        return {
            files: outputFiles,
            errors: Array.from(errors),
        };
    }
    async parseFile(filePath) {
        const cached = this.#parseCache.get(filePath);
        if (cached !== undefined) {
            return await cached;
        }
        // Cache the in-flight read+parse promise (not the resolved value) so that
        // when many checks resolve the same shared file concurrently they share a
        // single read and a single AST instead of each allocating its own.
        const promise = this.readFile(filePath)
            .then(({ content }) => Parser.parseDependencies(filePath, content));
        this.#parseCache.set(filePath, promise);
        return await promise;
    }
    async parseFileContent(filePath, content) {
        const cached = this.#parseCache.get(filePath);
        if (cached !== undefined) {
            return await cached;
        }
        // The content is already in hand, so parse synchronously, but cache the
        // result as a promise to share the same #parseCache with parseFile.
        const promise = Promise.resolve(Parser.parseDependencies(filePath, content));
        this.#parseCache.set(filePath, promise);
        return await promise;
    }
    async resolveDependencies(filePath, dependencies) {
        const cached = this.#resolveCache.get(filePath);
        if (cached !== undefined) {
            return await cached;
        }
        // Cache the in-flight promise so concurrent callers for the same file
        // share a single dependency resolution instead of each walking the whole
        // tree and allocating a separate result.
        const promise = this.resolver.resolveDependenciesForFilePath(filePath, dependencies);
        this.#resolveCache.set(filePath, promise);
        return await promise;
    }
    async parse(entrypoint) {
        this.validateEntrypoint(entrypoint);
        const { content } = await this.readFile(entrypoint);
        /*
      * The importing of files forms a directed graph.
      * Vertices are source files and edges are from importing other files.
      * We can find all of the files we need to run the check by traversing this graph.
      * In this implementation, we use breadth first search.
      */
        const collector = new Collector(entrypoint, content);
        const neighbors = new NeighborDependencyCollector();
        const bfsQueue = [{ filePath: entrypoint, content }];
        while (bfsQueue.length > 0) {
            // Since we just checked the length, shift() will never return undefined.
            // We can add a not-null assertion operator (!).
            const item = bfsQueue.shift();
            const ops = this.determineFileOps(item.filePath);
            // Not necessary - just a minor optimization.
            if (ops === 0) {
                continue;
            }
            let dependencies = [];
            if (ops & FILEOP_PARSE) {
                const { module, error } = await this.parseFileContent(item.filePath, item.content);
                if (error) {
                    collector.addParsingError(item.filePath, error.message);
                    continue;
                }
                dependencies = module.dependencies;
            }
            if (ops & FILEOP_RESOLVE) {
                const resolvedDependencies = await this.resolveDependencies(item.filePath, dependencies);
                if (this.checkUnsupportedModules) {
                    const unsupportedDependencies = resolvedDependencies.external.flatMap(dep => {
                        if (!this.supportsModule(dep.importPath)) {
                            return [dep.importPath];
                        }
                        else {
                            return [];
                        }
                    });
                    if (unsupportedDependencies.length) {
                        collector.addUnsupportedNpmDependencies(item.filePath, unsupportedDependencies);
                    }
                }
                for (const dep of resolvedDependencies.missing) {
                    collector.addMissingFile(dep.filePath);
                }
                for (const dep of resolvedDependencies.local) {
                    const filePath = dep.sourceFile.meta.filePath;
                    if (collector.hasDependency(filePath)) {
                        continue;
                    }
                    collector.addDependency(filePath, dep.sourceFile.contents);
                    bfsQueue.push({ filePath, content: dep.sourceFile.contents });
                }
                neighbors.markDepended(...resolvedDependencies.neighbors.depends);
                neighbors.markReferenced(...resolvedDependencies.neighbors.references);
            }
        }
        for (const pkg of neighbors.unreferencedDependedPackages()) {
            for (const f of createFauxPackageFiles(pkg)) {
                collector.addDependency(f.filePath, f.content);
            }
        }
        collector.validate();
        return collector.getItems();
    }
    static parseDependencies(filePath, contents) {
        debug(`Parsing dependencies of ${filePath}`);
        const dependencies = new RawDependencyCollector();
        const extension = path.extname(filePath);
        try {
            if (isCoreExtension(extension) && extension !== '.json') {
                const ast = acorn.parse(contents, {
                    allowReturnOutsideFunction: true,
                    ecmaVersion: 'latest',
                    allowImportExportEverywhere: true,
                    allowAwaitOutsideFunction: true,
                });
                walk.simple(ast, Parser.jsNodeVisitor(dependencies));
            }
            else if (isTSExtension(extension)) {
                const tsParser = getTsParser();
                const ast = tsParser.parse(contents, {
                    // We must only enable jsx for tsx/jsx files. Otherwise type brackets
                    // may confuse the parser and cause an error.
                    jsx: extension.endsWith('x'),
                });
                // The AST from typescript-estree is slightly different from the type used by acorn-walk.
                // This doesn't actually cause problems (both are "ESTree's"), but we need to ignore type errors here.
                // @ts-ignore
                walk.simple(ast, Parser.tsNodeVisitor(tsParser, dependencies));
            }
        }
        catch (err) {
            debug(`Failed to parse dependencies of ${filePath}: ${err}`);
            return {
                module: {
                    dependencies: dependencies.state(),
                },
                error: err,
            };
        }
        return {
            module: {
                dependencies: dependencies.state(),
            },
        };
    }
    static jsNodeVisitor(dependencies) {
        return {
            CallExpression(node) {
                if (!Parser.isRequireExpression(node))
                    return;
                const requireStringArg = Parser.getRequireStringArg(node);
                Parser.registerDependency(requireStringArg, 'require', dependencies);
            },
            ImportDeclaration(node) {
                if (node.source.type !== 'Literal')
                    return;
                Parser.registerDependency(node.source.value, 'import', dependencies);
            },
            ExportNamedDeclaration(node) {
                if (node.source === null)
                    return;
                if (node.source.type !== 'Literal')
                    return;
                Parser.registerDependency(node.source.value, 'import', dependencies);
            },
            ExportAllDeclaration(node) {
                if (node.source === null)
                    return;
                if (node.source.type !== 'Literal')
                    return;
                Parser.registerDependency(node.source.value, 'import', dependencies);
            },
        };
    }
    static tsNodeVisitor(tsParser, dependencies) {
        return {
            // While rare, TypeScript files may also use require.
            CallExpression(node) {
                if (!Parser.isRequireExpression(node))
                    return;
                const requireStringArg = Parser.getRequireStringArg(node);
                Parser.registerDependency(requireStringArg, 'require', dependencies);
            },
            ImportDeclaration(node) {
                // For now, we only support literal strings in the import statement
                if (node.source.type !== tsParser.TSESTree.AST_NODE_TYPES.Literal)
                    return;
                Parser.registerDependency(node.source.value, 'import', dependencies);
            },
            ExportNamedDeclaration(node) {
                // The statement isn't importing another dependency
                if (node.source === null)
                    return;
                // For now, we only support literal strings in the export statement
                if (node.source.type !== tsParser.TSESTree.AST_NODE_TYPES.Literal)
                    return;
                Parser.registerDependency(node.source.value, 'import', dependencies);
            },
            ExportAllDeclaration(node) {
                if (node.source === null)
                    return;
                // For now, we only support literal strings in the export statement
                if (node.source.type !== tsParser.TSESTree.AST_NODE_TYPES.Literal)
                    return;
                Parser.registerDependency(node.source.value, 'import', dependencies);
            },
        };
    }
    static isRequireExpression(node) {
        if (node.type !== 'CallExpression') {
            // Ignore AST nodes that aren't call expressions
            return false;
        }
        else if (node.arguments.length === 0) {
            // Weird case of `require()` or `module.require()` without arguments
            return false;
        }
        else if (node.callee.type === 'Identifier') {
            // Handle the case of a simple call to `require('dependency')`
            return node.callee.name === 'require';
        }
        else if (node.callee.type === 'MemberExpression') {
            // Handle calls to `module.require('dependency')`
            const { object, property } = node.callee;
            return object.type === 'Identifier'
                && object.name === 'module'
                && property.type === 'Identifier'
                && property.name === 'require';
        }
        else {
            return false;
        }
    }
    static getRequireStringArg(node) {
        if (node.arguments[0].type === 'Literal') {
            return node.arguments[0].value;
        }
        else if (node.arguments[0].type === 'TemplateLiteral') {
            return node.arguments[0].quasis[0].value.cooked;
        }
        else {
            /*
            * It might be that `require` is called with a variable - `require(myPackage)`.
            * Unfortunately supporting that case would be complicated.
            * We just skip the dependency and hope that the check still works.
            */
            return null;
        }
    }
    static registerDependency(importPath, source, dependencies) {
        if (!importPath) {
            // If there's no importPath, don't register a dependency.
            return;
        }
        dependencies.add({
            importPath,
            source,
        });
    }
}
