// checkly/dist/services/check-parser/package-files/extension.js
import path from 'node:path';
export const CoreExtensions = ['.js', '.mjs', '.cjs', '.json'];
export function isCoreExtension(value) {
    return CoreExtensions.includes(value);
}
/**
 * Unlike TypeScript's native lookup order, our lookup order prefers
 * implementation files to declaration files.
 *
 * Why include declaration files at all? Some of our users use manually
 * created declaration files as essentially shared header files without a
 * corresponding implementation file, and the declaration is the only thing
 * we'll be able to find. Otherwise we'd complain about a missing dependency.
 *
 * @see https://www.typescriptlang.org/docs/handbook/modules/reference.html#file-extension-substitution
 */
export const tsCoreExtensionLookupOrder = {
    '.js': ['.ts', '.tsx', '.js', '.jsx', '.d.ts'],
    '.mjs': ['.mts', '.mjs', '.d.mts'],
    '.cjs': ['.cts', '.cjs', '.d.cts'],
    '.json': ['.json'],
};
export const TSExtensions = ['.ts', '.mts', '.cts', '.tsx'];
export function isTSExtension(value) {
    return TSExtensions.includes(value);
}
export const tsExtensionLookupOrder = {
    '.ts': ['.ts'],
    '.mts': ['.mts'],
    '.cts': ['.cts'],
    '.tsx': ['.tsx'],
};
function stripKnownSuffix(value, suffix) {
    return value.substring(0, value.length - suffix.length);
}
export class FileExtPath {
    filePath;
    ext;
    constructor(filePath, ext) {
        this.filePath = filePath;
        this.ext = ext;
    }
    static fromFilePath(filePath) {
        const ext = path.extname(filePath);
        return new FileExtPath(filePath, ext);
    }
    hasCoreExtension() {
        return this.ext in tsCoreExtensionLookupOrder;
    }
    hasTypeScriptExtension() {
        return this.ext in tsExtensionLookupOrder;
    }
    replaceExt(ext) {
        return stripKnownSuffix(this.filePath, this.ext) + ext;
    }
    appendExt(ext) {
        return this.filePath + ext;
    }
    resolve(...paths) {
        return path.resolve(this.filePath, ...paths);
    }
    self() {
        return this.filePath;
    }
}
