// checkly/dist/services/check-parser/package-files/lookup.js
import { PackageJsonFile } from './package-json-file.js';
import { FileExtPath, tsCoreExtensionLookupOrder, tsExtensionLookupOrder } from './extension.js';
export class LookupContext {
    #plainJs;
    constructor(options) {
        this.#plainJs = options.plainJs ?? false;
    }
    static forFilePath(filePath) {
        return new LookupContext({
            plainJs: FileExtPath.fromFilePath(filePath).hasCoreExtension(),
        });
    }
    collectLookupPaths(filePath) {
        const extPath = FileExtPath.fromFilePath(filePath);
        if (this.#plainJs) {
            if (extPath.hasCoreExtension()) {
                return [extPath.self()];
            }
            return this.extlessCoreLookupPaths(extPath);
        }
        if (extPath.hasCoreExtension()) {
            const extensions = tsCoreExtensionLookupOrder[extPath.ext];
            return extensions.map(ext => extPath.replaceExt(ext));
        }
        if (extPath.hasTypeScriptExtension()) {
            const extensions = tsExtensionLookupOrder[extPath.ext];
            return extensions.map(ext => extPath.replaceExt(ext));
        }
        return this.extlessTSLookupPaths(extPath);
    }
    extlessCoreLookupPaths(extPath) {
        return [
            extPath.appendExt('.js'),
            extPath.appendExt('.mjs'),
            extPath.appendExt('.cjs'),
            extPath.appendExt('.json'),
            extPath.resolve(PackageJsonFile.FILENAME),
            extPath.resolve('index.js'),
            extPath.resolve('index.mjs'),
            extPath.resolve('index.cjs'),
            extPath.resolve('index.json'), // Yes, this works.
        ];
    }
    extlessTSLookupPaths(extPath) {
        return this.extlessCoreLookupPaths(extPath).flatMap(filePath => {
            const extPath = FileExtPath.fromFilePath(filePath);
            const extensions = tsCoreExtensionLookupOrder[extPath.ext];
            return extensions.map(ext => extPath.replaceExt(ext));
        });
    }
}
