// checkly/dist/services/check-parser/package-files/package-json-file.js
import path from 'node:path';
import semver from 'semver';
import { JsonSourceFile } from './json-source-file.js';
import { FileMeta, SourceFile } from './source-file.js';
import { PathResolver } from './paths.js';
export class PackageJsonFile {
    static FILENAME = 'package.json';
    static #id = 0;
    id = ++PackageJsonFile.#id;
    jsonFile;
    basePath;
    mainPaths;
    constructor(jsonFile) {
        this.jsonFile = jsonFile;
        this.basePath = jsonFile.meta.dirname;
        const fallbackMainPath = path.resolve(this.basePath, 'index.js');
        this.mainPaths = jsonFile.data.main !== undefined
            ? [path.resolve(this.basePath, jsonFile.data.main), fallbackMainPath]
            : [fallbackMainPath];
    }
    hasExports() {
        return !!this.jsonFile.data.exports;
    }
    resolveExportPath(exportPath, conditions) {
        const resolver = PathResolver.createFromPaths(this.basePath, this.#resolveExports(this.jsonFile.data.exports ?? {}, conditions));
        // Normalize the export path to the canonical subpath form used by
        // `exports` keys. Per the Node.js spec, the root subpath is "." and
        // all other subpaths start with "./". An empty string means the bare
        // package import (`import x from 'foo'`) and maps to the root subpath.
        //
        // Previously the code blindly prepended "./" to everything, turning
        // "" into "./", which never matched the "." key and made bare imports
        // silently resolve to zero paths. Packages that also declared `main`
        // worked anyway because of the fallback in `resolveSourceFile`; other
        // packages did not.
        if (exportPath === '' || exportPath === '.') {
            exportPath = '.';
        }
        else if (!exportPath.startsWith('./')) {
            exportPath = `./${exportPath}`;
        }
        return resolver.resolve(exportPath);
    }
    #resolveExports(exports, conditions) {
        if (typeof exports === 'string') {
            return {
                '.': [exports],
            };
        }
        const resolved = {};
        for (const [from, target] of Object.entries(exports)) {
            const resolvedTarget = this.#resolveExportTarget(target, conditions);
            if (resolvedTarget !== undefined) {
                resolved[from] = [resolvedTarget];
            }
        }
        return resolved;
    }
    /**
     * Recursively resolves an export target against the provided conditions,
     * following the Node.js conditional exports resolution algorithm.
     *
     * Returns the first matching string target, or `undefined` if the target
     * resolves to `null` (blocked export) or cannot be resolved for the given
     * conditions.
     */
    #resolveExportTarget(target, conditions) {
        // null explicitly blocks an export.
        if (target === null) {
            return undefined;
        }
        // Direct string target — we're done.
        if (typeof target === 'string') {
            return target;
        }
        // Array — try each fallback in order until one resolves.
        if (Array.isArray(target)) {
            for (const item of target) {
                const resolvedItem = this.#resolveExportTarget(item, conditions);
                if (resolvedItem !== undefined) {
                    return resolvedItem;
                }
            }
            return undefined;
        }
        // Object — iterate conditions in insertion order (per the Node.js spec,
        // `default` is expected to be the last key in a condition object).
        if (typeof target === 'object') {
            for (const [condition, nested] of Object.entries(target)) {
                if (!conditions.includes(condition)) {
                    continue;
                }
                const resolvedNested = this.#resolveExportTarget(nested, conditions);
                if (resolvedNested !== undefined) {
                    return resolvedNested;
                }
            }
            return undefined;
        }
        return undefined;
    }
    get meta() {
        return this.jsonFile.meta;
    }
    get name() {
        return this.jsonFile.data.name;
    }
    get version() {
        return this.jsonFile.data.version;
    }
    get dependencies() {
        return this.jsonFile.data.dependencies;
    }
    get devDependencies() {
        return this.jsonFile.data.devDependencies;
    }
    get engines() {
        return this.jsonFile.data.engines;
    }
    get workspaces() {
        const workspaces = this.jsonFile.data.workspaces;
        if (Array.isArray(workspaces)) {
            return workspaces;
        }
        // Bun/Yarn object form: `{ packages: [...], catalogs/nohoist: ... }`.
        // Anything else (incl. a malformed value) yields no workspace patterns.
        if (typeof workspaces === 'object' && workspaces !== null && Array.isArray(workspaces.packages)) {
            return workspaces.packages;
        }
        return undefined;
    }
    supportsEngine(engine, version) {
        const requirements = this.engines?.[engine];
        if (requirements === undefined) {
            return {
                engine,
                incompatible: false,
            };
        }
        try {
            const ok = semver.satisfies(version, requirements);
            return {
                engine,
                requirements,
                incompatible: !ok,
            };
        }
        catch {
            return {
                engine,
                requirements,
                incompatible: false,
            };
        }
    }
    static make(filePath, data) {
        const contents = formatContents(data);
        const sourceFile = new SourceFile(FileMeta.fromFilePath(filePath), contents);
        const jsonFile = new JsonSourceFile(sourceFile, data);
        return new PackageJsonFile(jsonFile);
    }
    // eslint-disable-next-line require-await
    static async loadFromJsonSourceFile(jsonFile) {
        return new PackageJsonFile(jsonFile);
    }
    static async loadFromSourceFile(sourceFile) {
        const jsonSourceFile = await JsonSourceFile.loadFromSourceFile(sourceFile);
        if (!jsonSourceFile) {
            return;
        }
        return PackageJsonFile.loadFromJsonSourceFile(jsonSourceFile);
    }
    static async loadFromFilePath(filePath) {
        const sourceFile = await SourceFile.loadFromFilePath(filePath);
        if (!sourceFile) {
            return;
        }
        return PackageJsonFile.loadFromSourceFile(sourceFile);
    }
    static filePath(dirPath) {
        return path.join(dirPath, PackageJsonFile.FILENAME);
    }
    upsertDependencies(dependencies) {
        const result = updateDependencies(this.jsonFile.data.dependencies, dependencies);
        if (!result.changed) {
            return false;
        }
        this.jsonFile.data.dependencies = sortDependencies(result.dependencies);
        return true;
    }
    upsertDevDependencies(dependencies) {
        const result = updateDependencies(this.jsonFile.data.devDependencies, dependencies);
        if (!result.changed) {
            return false;
        }
        this.jsonFile.data.devDependencies = sortDependencies(result.dependencies);
        return true;
    }
    toJSON() {
        return formatContents(this.jsonFile.data);
    }
}
function updateDependencies(oldDependencies = {}, newDependencies = {}) {
    const result = {
        dependencies: { ...oldDependencies },
        changed: false,
    };
    const safeMinVersion = (version) => {
        try {
            return semver.minVersion(version);
        }
        catch {
            return null;
        }
    };
    for (const [name, newVersion] of Object.entries(newDependencies)) {
        const oldVersion = oldDependencies[name];
        if (oldVersion !== undefined) {
            const oldMinVersion = safeMinVersion(oldVersion);
            if (oldMinVersion === null) {
                // Assume it's something we shouldn't change.
                continue;
            }
            const newMinVersion = safeMinVersion(newVersion);
            if (newMinVersion === null) {
                // This should not happen.
                throw new Error(`Invalid new version '${newVersion}' for dependency '${name}'`);
            }
            if (oldMinVersion.compare(newMinVersion) >= 0) {
                // It's recent enough.
                continue;
            }
        }
        result.changed = true;
        result.dependencies[name] = newVersion;
    }
    return result;
}
function sortDependencies(dependencies) {
    return Object.fromEntries(Object.keys(dependencies).sort().map(name => {
        return [name, dependencies[name]];
    }));
}
function formatContents(data) {
    // NPM adds a trailing newline, so we do too.
    return JSON.stringify(data, null, 2) + '\n';
}
