// checkly/dist/services/check-parser/package-files/tsconfig-json-file.js
import path from 'node:path';
import { PathResolver } from './paths.js';
export class TSConfigFile {
    static FILENAME = 'tsconfig.json';
    static #id = 0;
    id = ++TSConfigFile.#id;
    jsonFile;
    basePath;
    moduleResolution;
    baseUrl;
    pathResolver;
    relatedSourceFiles = [];
    constructor(jsonFile) {
        this.jsonFile = jsonFile;
        this.basePath = jsonFile.meta.dirname;
        this.moduleResolution = jsonFile.data.compilerOptions?.moduleResolution?.toLocaleLowerCase() ?? 'unspecified';
        this.baseUrl = jsonFile.data.compilerOptions?.baseUrl !== undefined
            ? path.resolve(this.basePath, jsonFile.data.compilerOptions.baseUrl)
            : this.basePath;
        this.pathResolver = PathResolver.createFromPaths(this.baseUrl, jsonFile.data.compilerOptions?.paths ?? {});
    }
    get meta() {
        return this.jsonFile.meta;
    }
    // eslint-disable-next-line require-await
    static async loadFromJsonTextSourceFile(jsonFile) {
        return new TSConfigFile(jsonFile);
    }
    static filePath(dirPath) {
        return path.join(dirPath, TSConfigFile.FILENAME);
    }
    resolvePath(importPath) {
        return this.pathResolver.resolve(importPath);
    }
    collectLookupPaths(filePath) {
        let { 
        // eslint-disable-next-line prefer-const
        outDir, rootDir, 
        // eslint-disable-next-line prefer-const
        rootDirs, composite, } = this.jsonFile.data.compilerOptions ?? {};
        const candidates = [];
        if (outDir === undefined) {
            candidates.push(path.resolve(this.basePath, filePath));
            return candidates; // Nothing more we can do.
        }
        if (composite === undefined) {
            composite = false;
        }
        // Inferred rootDir is tsconfig directory if composite === true.
        if (rootDir === undefined && composite) {
            rootDir = '.';
        }
        // If we still don't have a root, we should calculate the longest common
        // path among input files, but that's a lot of effort. Assume tsconfig
        // directory and hope for the best.
        if (rootDir === undefined) {
            rootDir = '.';
        }
        const absoluteOutDir = path.resolve(this.basePath, outDir);
        const relativePath = path.relative(absoluteOutDir, filePath);
        // If the file is outside outDir, then assume we're looking for
        // something that wasn't compiled using this tsconfig (or at all), and
        // stop looking.
        if (relativePath.startsWith('..')) {
            candidates.push(path.resolve(this.basePath, filePath));
            return candidates;
        }
        candidates.push(path.resolve(this.basePath, rootDir, relativePath));
        // Assume that our inferred (or user specified) rootDir is enough to cover
        // the same conditions we'd have to infer rootDirs, and only add rootDirs
        // if they're actually set.
        for (const multiRootDir of rootDirs ?? []) {
            candidates.push(path.resolve(this.basePath, multiRootDir, relativePath));
        }
        return candidates;
    }
    registerRelatedSourceFile(file) {
        this.relatedSourceFiles.push(file);
    }
}
