// checkly/dist/services/check-parser/package-files/workspace.js
import { glob } from 'glob';
import { PackageJsonFile } from './package-json-file.js';
export class Package {
    name;
    path;
    workspaces;
    constructor({ name, path, workspaces }) {
        this.name = name;
        this.path = path;
        this.workspaces = workspaces;
    }
    get packageJsonPath() {
        return PackageJsonFile.filePath(this.path);
    }
    // eslint-disable-next-line require-await
    static async loadFromPackageJsonFile(packageJson) {
        const { name, workspaces } = packageJson;
        if (name === undefined) {
            return;
        }
        return new Package({
            name,
            path: packageJson.meta.dirname,
            workspaces,
        });
    }
    static async loadFromDirPath(dirPath) {
        const packageJson = await PackageJsonFile.loadFromFilePath(PackageJsonFile.filePath(dirPath));
        if (!packageJson) {
            return;
        }
        return await Package.loadFromPackageJsonFile(packageJson);
    }
}
export class Workspace {
    /**
     * The workspace root package.
     */
    root;
    /**
     * Packages that are a part of the workspace, excluding the root package.
     */
    packages;
    /**
     * The package manager specific lockfile of the workspace.
     */
    lockfile;
    /**
     * The package manager specific config file of the workspace.
     */
    configFile;
    #membersByName = new Map();
    #membersByPath = new Map();
    constructor(options) {
        this.root = options.root;
        this.packages = options.packages;
        this.#membersByName = [options.root, ...options.packages].reduce((map, pkg) => map.set(pkg.name, pkg), new Map());
        this.#membersByPath = [options.root, ...options.packages].reduce((map, pkg) => map.set(pkg.path, pkg), new Map());
        this.lockfile = options.lockfile;
        this.configFile = options.configFile;
    }
    memberByName(name) {
        return this.#membersByName.get(name);
    }
    memberByPath(path) {
        return this.#membersByPath.get(path);
    }
    /**
     * @param root An absolute path to the workspace root.
     * @param patterns Relative workspace patterns.
     * @returns Absolute paths to every package in the workspace.
     */
    static async resolvePatterns(root, patterns) {
        const lookup = patterns.map(pattern => {
            const joiner = pattern.endsWith('/') ? '' : '/';
            return [pattern, PackageJsonFile.FILENAME].join(joiner);
        });
        // Always ignore node_modules when resolving workspace patterns. npm
        // and yarn both skip node_modules internally when matching workspace
        // globs, and letting wide patterns like `**` pull in every installed
        // package causes the parser to walk into transitive dependencies
        // (including native bindings like fsevents) — see issue #1274.
        const results = await glob(lookup, {
            cwd: root,
            absolute: true,
            ignore: ['**/node_modules/**'],
        });
        const packages = [];
        for (const result of results) {
            const packageJson = await PackageJsonFile.loadFromFilePath(result);
            if (packageJson === undefined) {
                continue;
            }
            const workspacePackage = await Package.loadFromPackageJsonFile(packageJson);
            if (workspacePackage === undefined) {
                continue;
            }
            packages.push(workspacePackage);
        }
        return packages;
    }
}
