// checkly/dist/commands/incidents/list.js
import { Flags } from '@oclif/core';
import chalk from 'chalk';
import { AuthCommand } from '../authCommand.js';
import { outputFlag } from '../../helpers/flags.js';
import * as api from '../../rest/api.js';
import { formatIncidentsList } from '../../formatters/incidents.js';
import { filterIncidentsByStatus, flattenStatusPageServices, incidentListStatusOptions, } from '../../helpers/incidents.js';
export default class IncidentsList extends AuthCommand {
    static hidden = false;
    static readOnly = true;
    static idempotent = true;
    static description = 'List incidents, optionally filtered by status page or status.';
    static flags = {
        'limit': Flags.integer({
            char: 'l',
            description: 'Number of incidents to return (1-100).',
            default: 25,
        }),
        'status-page-id': Flags.string({
            description: 'Filter incidents by status page ID.',
        }),
        'status': Flags.string({
            description: 'Filter by incident status.',
            options: incidentListStatusOptions,
            default: 'open',
        }),
        'output': outputFlag({ default: 'table' }),
    };
    async run() {
        const { flags } = await this.parse(IncidentsList);
        this.style.outputFormat = flags.output;
        try {
            const page = await api.incidents.getAll({ limit: flags.limit });
            let incidents = filterIncidentsByStatus(page.entries, flags.status);
            if (flags['status-page-id']) {
                const statusPage = await api.statusPages.get(flags['status-page-id']);
                const statusPageServiceIds = new Set(flattenStatusPageServices(statusPage).map(service => service.id));
                incidents = incidents.filter(incident => incident.services.some(service => statusPageServiceIds.has(service.id)));
            }
            if (flags.output === 'json') {
                this.log(JSON.stringify({
                    data: incidents,
                    count: incidents.length,
                    ...(page.nextId ? { nextId: page.nextId } : {}),
                }, null, 2));
                return;
            }
            if (incidents.length === 0) {
                this.log('No incidents found.');
                return;
            }
            const fmt = flags.output === 'md' ? 'md' : 'terminal';
            if (fmt === 'md') {
                this.log(formatIncidentsList(incidents, fmt));
                return;
            }
            const output = [];
            output.push(formatIncidentsList(incidents, fmt));
            output.push('');
            const filters = [];
            if (flags.status !== 'all')
                filters.push(`status=${flags.status}`);
            if (flags['status-page-id'])
                filters.push(`status-page-id=${flags['status-page-id']}`);
            const summary = `${incidents.length} incident${incidents.length === 1 ? '' : 's'}`;
            if (filters.length > 0) {
                output.push(chalk.dim(`Showing ${summary} (${filters.join(', ')})`));
            }
            else {
                output.push(chalk.dim(`Showing ${summary}`));
            }
            this.log(output.join('\n'));
        }
        catch (err) {
            this.style.longError('Failed to list incidents.', err);
            process.exitCode = 1;
        }
    }
}
