// checkly/dist/helpers/command-preview.js
const OMITTED_FLAGS = new Set(['output', 'force', 'dry-run']);
export function buildConfirmCommand(command, flags, args) {
    const parts = ['checkly', command];
    if (args) {
        for (const value of Object.values(args)) {
            parts.push(String(value));
        }
    }
    for (const [key, value] of Object.entries(flags)) {
        if (OMITTED_FLAGS.has(key))
            continue;
        if (value === undefined || value === null)
            continue;
        if (Array.isArray(value)) {
            for (const item of value) {
                parts.push(`--${key}="${item}"`);
            }
        }
        else if (typeof value === 'boolean') {
            parts.push(value ? `--${key}` : `--no-${key}`);
        }
        else {
            parts.push(`--${key}="${value}"`);
        }
    }
    parts.push('--force');
    return parts.join(' ');
}
export function formatPreviewForAgent(preview, status) {
    return {
        status,
        command: preview.command,
        description: preview.description,
        classification: preview.classification,
        changes: preview.changes,
        confirmCommand: buildConfirmCommand(preview.command, preview.flags, preview.args),
    };
}
export function formatPreviewForTerminal(preview) {
    if (preview.changes.length === 1) {
        return `This will ${preview.changes[0]}`;
    }
    const lines = [];
    lines.push('This will:');
    for (const change of preview.changes) {
        lines.push(`  - ${change}`);
    }
    return lines.join('\n');
}
