// checkly/dist/helpers/command-style.js
import chalk from 'chalk';
import { ux } from '@oclif/core';
import { wrap } from './wrap.js';
import logSymbols from 'log-symbols';
const textWrapOptions = {
    prefix: '  ',
    length: 78,
};
const errorWrapOptions = {
    ...textWrapOptions,
    length: Infinity,
};
export class CommandStyle {
    c;
    outputFormat;
    constructor(command) {
        this.c = command;
    }
    actionStart(message) {
        if (this.c.fancy) {
            ux.action.start(message, undefined, { stdout: true });
        }
    }
    actionStatus(message) {
        if (this.c.fancy) {
            ux.action.status = message;
        }
    }
    actionSuccess() {
        if (this.c.fancy) {
            ux.action.stop(`✅`);
            this.c.log();
        }
    }
    actionFailure() {
        if (this.c.fancy) {
            ux.action.stop(`❌`);
            this.c.log();
        }
    }
    comment(message) {
        this.c.log(chalk.cyan(wrap(message, { ...textWrapOptions, prefix: '// ' })));
        this.c.log();
    }
    longSuccess(title, message) {
        this.c.log(`${logSymbols.success} ${title}`);
        this.c.log();
        this.c.log(wrap(message, textWrapOptions));
        this.c.log();
    }
    shortSuccess(message) {
        this.c.log(`${logSymbols.success} ${message}`);
        this.c.log();
    }
    longInfo(title, message) {
        this.c.log(`${logSymbols.info} ${title}`);
        this.c.log();
        this.c.log(wrap(message, textWrapOptions));
        this.c.log();
    }
    shortInfo(message) {
        this.c.log(`${logSymbols.info} ${message}`);
        this.c.log();
    }
    longWarning(title, message) {
        this.c.log(`${logSymbols.warning} ${title}`);
        this.c.log();
        this.c.log(chalk.yellow(this.#formatDescription(message)));
        this.c.log();
    }
    shortWarning(message) {
        this.c.log(`${logSymbols.warning} ${chalk.yellow(message)}`);
        this.c.log();
    }
    longError(title, message) {
        if (this.outputFormat === 'json') {
            this.c.log(JSON.stringify({ error: title, detail: this.#plainDescription(message) }));
            return;
        }
        this.c.log(`${logSymbols.error} ${title}`);
        this.c.log();
        this.c.log(chalk.red(this.#formatDescription(message)));
        this.c.log();
    }
    shortError(message) {
        if (this.outputFormat === 'json') {
            this.c.log(JSON.stringify({ error: message }));
            return;
        }
        this.c.log(`${logSymbols.error} ${chalk.red(message)}`);
        this.c.log();
    }
    fatal(message) {
        this.c.log(chalk.red(message));
        this.c.log();
    }
    #formatDescription(description) {
        if (typeof description === 'string') {
            return wrap(description, textWrapOptions);
        }
        const { message } = description;
        return wrap(message, errorWrapOptions);
    }
    #plainDescription(description) {
        return typeof description === 'string' ? description : description.message;
    }
}
