// checkly/dist/helpers/incidents.js
export const incidentSeverityOptions = ['minor', 'medium', 'major', 'critical'];
export const incidentProgressStatusOptions = ['investigating', 'identified', 'monitoring'];
export const incidentListStatusOptions = ['open', 'resolved', 'all'];
export const DEFAULT_CREATE_INCIDENT_MESSAGE = 'We are investigating this incident.';
export const DEFAULT_RESOLVE_INCIDENT_MESSAGE = 'Incident resolved.';
export function toIncidentSeverity(value) {
    return value.toUpperCase();
}
export function toIncidentProgressStatus(value) {
    return value.toUpperCase();
}
export function flattenStatusPageServices(statusPage) {
    return statusPage.cards.flatMap(card => card.services);
}
export function resolveIncidentServices(statusPageServices, selectedServiceIds) {
    if (statusPageServices.length === 0) {
        throw new Error('This status page has no services. Add services to the status page first.');
    }
    let selectedServices;
    if (!selectedServiceIds || selectedServiceIds.length === 0) {
        selectedServices = statusPageServices;
    }
    else {
        const serviceMap = new Map(statusPageServices.map(service => [service.id, service]));
        const missing = selectedServiceIds.filter(id => !serviceMap.has(id));
        if (missing.length > 0) {
            throw new Error(`Unknown service IDs for this status page: ${missing.join(', ')}`);
        }
        selectedServices = selectedServiceIds.map(id => serviceMap.get(id));
    }
    return selectedServices.map(service => {
        if (!service.accountId) {
            throw new Error(`Service "${service.name}" (${service.id}) is missing account metadata from the API response.`);
        }
        return {
            id: service.id,
            name: service.name,
            accountId: service.accountId,
        };
    });
}
export function filterIncidentsByStatus(incidents, status) {
    if (status === 'all')
        return incidents;
    if (status === 'resolved') {
        return incidents.filter(incident => incident.lastUpdateStatus === 'RESOLVED');
    }
    return incidents.filter(incident => incident.lastUpdateStatus !== 'RESOLVED');
}
