// checkly/dist/services/checkly-config-loader.js
import * as path from 'path';
import fs from 'node:fs/promises';
import { findPlaywrightConfigPath, getDefaultChecklyConfig, writeChecklyConfigFile } from './util.js';
import { Session } from '../constructs/index.js';
function isString(obj) {
    return (Object.prototype.toString.call(obj) === '[object String]');
}
export async function getChecklyConfigFile() {
    const filenames = [
        'checkly.config.ts',
        'checkly.config.mts',
        'checkly.config.cts',
        'checkly.config.js',
        'checkly.config.mjs',
        'checkly.config.cjs',
    ];
    let config;
    for (const configFile of filenames) {
        const dir = path.resolve(path.dirname(configFile));
        const configFilePath = path.resolve(dir, configFile);
        try {
            await fs.access(configFilePath, fs.constants.R_OK);
        }
        catch {
            continue;
        }
        const file = await fs.readFile(configFilePath);
        if (file) {
            config = {
                checklyConfig: file.toString(),
                fileName: configFile,
            };
            break;
        }
    }
    return config;
}
export class ConfigNotFoundError extends Error {
    searchPaths;
    configFiles;
    constructor(searchPaths, configFiles, options) {
        const message = `Unable to detect a Checkly configuration file in any of the following paths:`
            + `\n\n`
            + `${searchPaths.map(searchPath => `  ${searchPath}`).join('\n')}`
            + `\n\n`
            + `Configuration files we looked for:`
            + `\n\n`
            + `${configFiles.map(lockfile => `  ${lockfile}`).join('\n')}`;
        super(message, options);
        this.name = 'ConfigNotFoundError';
        this.searchPaths = searchPaths;
        this.configFiles = configFiles;
    }
}
export const defaultFilenames = [
    'checkly.config.ts',
    'checkly.config.mts',
    'checkly.config.cts',
    'checkly.config.js',
    'checkly.config.mjs',
    'checkly.config.cjs',
];
export async function loadChecklyConfig(dir, filenames = defaultFilenames, writeChecklyConfig = true, playwrightConfigPath) {
    Session.loadingChecklyConfigFile = true;
    try {
        let config;
        Session.checklyConfigFileConstructs = [];
        for (const filename of filenames) {
            const filePath = path.join(dir, filename);
            try {
                await fs.access(filePath, fs.constants.R_OK);
            }
            catch {
                continue;
            }
            config = await Session.loadFile(filePath);
            break;
        }
        if (!config) {
            config = await handleMissingConfig(dir, filenames, writeChecklyConfig, playwrightConfigPath);
        }
        validateConfigFields(config, ['logicalId', 'projectName']);
        const constructs = Session.checklyConfigFileConstructs;
        Session.checklyConfigFileConstructs = [];
        if (config.cli?.loader) {
            Session.loader = config.cli.loader;
        }
        return { config, constructs };
    }
    finally {
        Session.loadingChecklyConfigFile = false;
    }
}
async function handleMissingConfig(dir, filenames, shouldWriteConfig = true, pwPath) {
    const baseName = path.basename(dir);
    const playwrightConfigPath = pwPath ?? findPlaywrightConfigPath(dir);
    if (playwrightConfigPath) {
        const checklyConfig = getDefaultChecklyConfig(baseName, `./${path.relative(dir, playwrightConfigPath)}`);
        if (shouldWriteConfig) {
            await writeChecklyConfigFile(dir, checklyConfig);
        }
        return checklyConfig;
    }
    throw new ConfigNotFoundError([dir], filenames);
}
function validateConfigFields(config, fields) {
    for (const field of fields) {
        if (!config?.[field] || !isString(config[field])) {
            throw new Error(`Config object missing a ${field} as type string`);
        }
    }
}
