// checkly/dist/services/config.js
import Conf from 'conf';
const dataSchema = {
    accountId: { type: 'string' },
    accountName: { type: 'string' },
};
const authSchema = {
    apiKey: { type: 'string' },
};
const projectSuffix = process.env.CHECKLY_ENV ?? '';
// eslint-disable-next-line no-restricted-syntax
var Env;
(function (Env) {
    Env["production"] = "production";
    Env["staging"] = "staging";
    Env["development"] = "development";
    Env["local"] = "local";
})(Env || (Env = {}));
class ChecklyConfig {
    _auth;
    _data;
    // Accessing auth or data will cause a config file to be created.
    // We should avoid doing this unless absolutely necessary, since this operation can fail due to file permissions.
    get auth() {
        // Create this._auth lazily
        return this._auth ?? (this._auth = new Conf({
            projectName: '@checkly/cli',
            configName: 'auth',
            projectSuffix,
            // @ts-ignore
            schema: authSchema,
        }));
    }
    get data() {
        // Create this._data lazily
        return this._data ?? (this._data = new Conf({
            projectName: '@checkly/cli',
            configName: 'config',
            projectSuffix,
            // @ts-ignore
            schema: dataSchema,
        }));
    }
    clear() {
        this.auth.clear();
        this.data.clear();
    }
    getEnv() {
        const environments = ['production', 'development', 'staging', 'local'];
        const env = process.env.CHECKLY_ENV || environments[0];
        if (!(env in Env)) {
            throw new Error('Invalid CHECKLY_ENV');
        }
        return env;
    }
    getApiKey() {
        return process.env.CHECKLY_API_KEY || this.auth.get('apiKey') || '';
    }
    getAccountId() {
        return process.env.CHECKLY_ACCOUNT_ID || this.data.get('accountId') || '';
    }
    hasEnvVarsConfigured() {
        const apiKey = process.env.CHECKLY_API_KEY || '';
        const accoundId = process.env.CHECKLY_ACCOUNT_ID || '';
        return apiKey !== '' || accoundId !== '';
    }
    getApiUrl() {
        const environments = {
            local: process.env.CHECKLY_API_URL || 'http://127.0.0.1:3000',
            development: 'https://api-dev.checklyhq.com',
            staging: 'https://api-test.checklyhq.com',
            production: 'https://api.checklyhq.com',
        };
        return environments[this.getEnv()];
    }
    getMqttUrl() {
        const environments = {
            local: 'wss://events-local.checklyhq.com',
            development: 'wss://events-dev.checklyhq.com',
            staging: 'wss://events-test.checklyhq.com',
            production: 'wss://events.checklyhq.com',
        };
        return environments[this.getEnv()];
    }
    hasValidCredentials() {
        if (this.getApiKey() !== '' && this.getAccountId() !== '') {
            return true;
        }
        return false;
    }
}
const config = new ChecklyConfig();
export default config;
