// checkly/dist/services/playwright-config.js
import * as path from 'node:path';
function toAbsolutePath(dir, file) {
    return path.resolve(dir, file);
}
function parseBrowsers(config) {
    const browsers = new Set();
    const browserKeywords = ['browserName', 'defaultBrowserType', 'channel'];
    for (const browserKeyword of browserKeywords) {
        if (config?.use?.[browserKeyword]) {
            browsers.add(config?.use[browserKeyword]);
        }
    }
    return browsers;
}
function buildSnapshotTemplates(config, filePath) {
    const fileRelativePath = path.relative(config.testDir, filePath);
    const parsed = path.parse(fileRelativePath);
    return Array.from(config.snapshotTemplates).map(template => {
        return template
            .replace(/\{(.)?testDir\}/g, '$1' + config.testDir)
            .replace(/\{(.)?snapshotDir\}/g, '$1' + config.snapshotDir)
            .replace(/\{(.)?testFileDir\}/g, '$1' + parsed.dir)
            .replace(/\{(.)?platform\}/g, '$1' + config.platform)
            .replace(/\{(.)?projectName\}/g, config.projectName)
            .replace(/\{(.)?testName\}/g, '$1' + '*')
            .replace(/\{(.)?testFileName\}/g, '$1' + parsed.base)
            .replace(/\{(.)?testFilePath\}/g, '$1' + fileRelativePath)
            .replace(/\{(.)?arg\}/g, '$1' + '*')
            .replace(/\{(.)?ext\}/g, '$1' + '.*');
    });
}
const DEFAULT_SNAPSHOT_TEMPLATE = '{snapshotDir}/{testFilePath}-snapshots/{arg}{ext}';
export class PlaywrightConfig {
    projectName;
    configFilePath;
    platform;
    testDir;
    snapshotDir;
    testMatch;
    snapshotTemplates;
    browsers;
    projects;
    files;
    constructor(filePath, playwrightConfig) {
        const dir = path.dirname(filePath);
        this.projectName = '';
        this.platform = 'linux';
        this.testDir = playwrightConfig.testDir ? toAbsolutePath(dir, playwrightConfig.testDir) : dir;
        this.snapshotDir = playwrightConfig.snapshotDir ? toAbsolutePath(dir, playwrightConfig.snapshotDir) : this.testDir;
        this.files = new Set();
        this.snapshotTemplates = new Set();
        const testMatch = playwrightConfig.testMatch ?? ['**/*.@(spec|test).?(c|m)[jt]s?(x)'];
        this.testMatch = new Set(Array.isArray(testMatch) ? testMatch : [testMatch]);
        this.configFilePath = filePath;
        const fileDefinitions = ['tsconfig', 'globalSetup', 'globalTeardown'];
        for (const fileDefinition of fileDefinitions) {
            const definition = playwrightConfig[fileDefinition];
            if (!definition) {
                continue;
            }
            if (Array.isArray(definition)) {
                definition.forEach((file) => this.files.add(toAbsolutePath(dir, file)));
            }
            else {
                this.files.add(toAbsolutePath(dir, definition));
            }
        }
        if (playwrightConfig.snapshotPathTemplate) {
            this.snapshotTemplates.add(playwrightConfig.snapshotPathTemplate);
        }
        const expect = playwrightConfig.expect;
        if (expect?.toHaveScreenshot?.pathTemplate) {
            this.snapshotTemplates.add(expect.toHaveScreenshot.pathTemplate);
        }
        if (expect?.toMatchAriaSnapshot?.pathTemplate) {
            this.snapshotTemplates.add(expect.toMatchAriaSnapshot.pathTemplate);
        }
        if (this.snapshotTemplates.size === 0) {
            this.snapshotTemplates.add(DEFAULT_SNAPSHOT_TEMPLATE);
        }
        this.browsers = parseBrowsers(playwrightConfig);
        if (playwrightConfig.projects?.length) {
            this.projects = playwrightConfig.projects.map((project) => new PlaywrightProject(dir, this, project));
        }
    }
    getBrowsers() {
        const browsers = new Set(this.browsers);
        this.projects?.forEach(project => project.browsers.forEach(browser => browsers.add(browser)));
        if (browsers.size === 0) {
            // Add the default browser
            browsers.add('chromium');
        }
        return Array.from(browsers);
    }
    getSnapshotPath(filePath) {
        return buildSnapshotTemplates(this, filePath);
    }
}
export class PlaywrightProject {
    projectName;
    platform;
    testDir;
    snapshotDir;
    testMatch;
    expect;
    snapshotTemplates;
    browsers;
    constructor(dir, playwrightConfig, playwrightProject) {
        this.projectName = playwrightProject.name;
        this.platform = 'linux';
        this.testDir = playwrightProject.testDir ? toAbsolutePath(dir, playwrightProject.testDir) : playwrightConfig.testDir;
        this.snapshotDir = playwrightProject.snapshotDir
            ? toAbsolutePath(dir, playwrightProject.snapshotDir)
            : (playwrightConfig.snapshotDir ?? this.testDir);
        this.snapshotTemplates = new Set();
        const testMatch = playwrightProject.testMatch ?? Array.from(playwrightConfig.testMatch);
        this.testMatch = new Set(Array.isArray(testMatch) ? testMatch : [testMatch]);
        const snapshotPathTemplate = playwrightProject.snapshotPathTemplate ?? playwrightConfig.snapshotPathTemplate;
        if (snapshotPathTemplate) {
            this.snapshotTemplates.add(snapshotPathTemplate);
        }
        // Check if the project overrides the global expect field
        const expect = playwrightProject.expect ?? playwrightConfig.expect;
        if (expect?.toHaveScreenshot?.pathTemplate) {
            this.snapshotTemplates.add(expect.toHaveScreenshot.pathTemplate);
        }
        if (expect?.toMatchAriaSnapshot?.pathTemplate) {
            this.snapshotTemplates.add(expect.toMatchAriaSnapshot.pathTemplate);
        }
        if (this.snapshotTemplates.size === 0) {
            this.snapshotTemplates.add(DEFAULT_SNAPSHOT_TEMPLATE);
        }
        this.browsers = parseBrowsers(playwrightProject);
    }
    getSnapshotPath(filePath) {
        return buildSnapshotTemplates(this, filePath);
    }
}
