// checkly/dist/services/test-runner.js
import * as uuid from 'uuid';
import { testSessions } from '../rest/api.js';
import AbstractCheckRunner from './abstract-check-runner.js';
import { pullSnapshots } from '../services/snapshot-service.js';
import { PlaywrightCheckBundle } from '../constructs/playwright-check-bundle.js';
export default class TestRunner extends AbstractCheckRunner {
    projectBundle;
    checkBundles;
    sharedFiles;
    location;
    shouldRecord;
    repoInfo;
    environment;
    updateSnapshots;
    baseDirectory;
    testRetryStrategy;
    streamLogs;
    refreshCache;
    constructor(accountId, projectBundle, checkBundles, sharedFiles, location, timeout, verbose, shouldRecord, repoInfo, environment, updateSnapshots, baseDirectory, testRetryStrategy, streamLogs, refreshCache, detach) {
        super(accountId, timeout, verbose, detach ?? false);
        this.projectBundle = projectBundle;
        this.checkBundles = checkBundles;
        this.sharedFiles = sharedFiles;
        this.location = location;
        this.shouldRecord = shouldRecord;
        this.repoInfo = repoInfo;
        this.environment = environment;
        this.updateSnapshots = updateSnapshots;
        this.baseDirectory = baseDirectory;
        this.testRetryStrategy = testRetryStrategy;
        this.streamLogs = streamLogs ?? false;
        this.refreshCache = refreshCache ?? false;
    }
    async scheduleChecks(checkRunSuiteId) {
        const checkRunJobs = this.checkBundles.map(({ construct: check, bundle }) => {
            // Get the group reference - Playwright checks store it in the bundle,
            // other checks store it in the check construct (as groupId)
            const groupId = bundle instanceof PlaywrightCheckBundle
                ? bundle.groupId
                : check.groupId;
            return {
                ...bundle.synthesize(),
                testRetryStrategy: this.testRetryStrategy,
                group: groupId ? this.projectBundle.data['check-group'][groupId.ref].bundle.synthesize() : undefined,
                sourceInfo: {
                    checkRunSuiteId,
                    checkRunId: uuid.v4(),
                    updateSnapshots: this.updateSnapshots,
                },
                logicalId: check.logicalId,
                filePath: check.getSourceFile(),
            };
        });
        if (!checkRunJobs.length) {
            throw new Error('Unable to find checks to run.');
        }
        const { data } = await testSessions.run({
            name: this.projectBundle.project.name,
            checkRunJobs,
            project: { logicalId: this.projectBundle.project.logicalId },
            sharedFiles: this.sharedFiles,
            runLocation: this.location,
            repoInfo: this.repoInfo,
            environment: this.environment,
            shouldRecord: this.shouldRecord,
            streamLogs: this.streamLogs,
            refreshCache: this.refreshCache,
        });
        const { testSessionId, sequenceIds } = data;
        const checks = this.checkBundles.map(({ construct: check }) => {
            return { check, sequenceId: sequenceIds?.[check.logicalId] };
        });
        return { testSessionId, checks };
    }
    async processCheckResult(result) {
        await super.processCheckResult(result);
        if (this.updateSnapshots) {
            await pullSnapshots(this.baseDirectory, result.assets?.snapshots);
        }
    }
}
