// checkly/dist/services/trigger-runner.js
import { testSessions } from '../rest/api.js';
import AbstractCheckRunner from './abstract-check-runner.js';
export default class TriggerRunner extends AbstractCheckRunner {
    shouldRecord;
    location;
    targetTags;
    targetCheckIds;
    envVars;
    repoInfo;
    environment;
    testSessionName;
    testRetryStrategy;
    refreshCache;
    constructor(accountId, timeout, verbose, shouldRecord, location, targetTags, targetCheckIds, envVars, repoInfo, environment, testSessionName, testRetryStrategy, refreshCache, detach) {
        super(accountId, timeout, verbose, detach ?? false);
        this.shouldRecord = shouldRecord;
        this.location = location;
        this.targetTags = targetTags;
        this.targetCheckIds = targetCheckIds;
        this.envVars = envVars;
        this.repoInfo = repoInfo;
        this.environment = environment;
        this.testSessionName = testSessionName;
        this.testRetryStrategy = testRetryStrategy;
        this.refreshCache = refreshCache ?? false;
    }
    async scheduleChecks(checkRunSuiteId) {
        const { data } = await testSessions.trigger({
            name: this.testSessionName ?? 'Triggered Session',
            shouldRecord: this.shouldRecord,
            runLocation: this.location,
            checkRunSuiteId,
            targetTags: this.targetTags,
            checkId: this.targetCheckIds.length > 0 ? this.targetCheckIds : undefined,
            environmentVariables: this.envVars,
            repoInfo: this.repoInfo,
            environment: this.environment,
            testRetryStrategy: this.testRetryStrategy,
            refreshCache: this.refreshCache,
        });
        const { checks, testSessionId, sequenceIds, } = data;
        const augmentedChecks = checks.map(check => ({
            check,
            sequenceId: sequenceIds?.[check.id],
        }));
        return { checks: augmentedChecks, testSessionId };
    }
}
