// checkly/dist/loader/jiti.js
import Debug from 'debug';
import { fileURLToPath } from 'node:url';
import { FileLoader, UnsupportedFileLoaderError } from './loader.js';
import { FileMatch } from './match.js';
import { preferenceDelta } from './config.js';
const __filename = fileURLToPath(import.meta.url);
const debug = Debug('checkly:cli:loader:jiti');
export class UninitializedJitiFileLoaderState extends FileLoader {
    static init;
    async loadFile(filePath) {
        UninitializedJitiFileLoaderState.init ??= (async () => {
            debug('Initializing loader');
            try {
                const jitiExports = await import('jiti');
                const jiti = jitiExports.createJiti(__filename, {
                    tsx: true,
                    tsconfigPaths: true,
                });
                debug(`Successfully initialized loader`);
                JitiFileLoader.state = new InitializedJitiFileLoaderState(jiti);
            }
            catch (err) {
                debug(`Failed to initialize loader: ${err}`);
                JitiFileLoader.state = new FailedJitiFileLoaderState(err);
            }
        })();
        await UninitializedJitiFileLoaderState.init;
        return await JitiFileLoader.state.loadFile(filePath);
    }
}
export class FailedJitiFileLoaderState extends FileLoader {
    error;
    constructor(error) {
        super();
        this.error = error;
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars, require-await
    async loadFile(filePath) {
        throw new UnsupportedFileLoaderError('JitiFileLoader is not supported', {
            cause: this.error,
        });
    }
}
export class InitializedJitiFileLoaderState extends FileLoader {
    jiti;
    constructor(jiti) {
        super();
        this.jiti = jiti;
    }
    async loadFile(filePath) {
        debug(`Loading file ${filePath}`);
        try {
            const moduleExports = await this.jiti.import(filePath);
            return moduleExports;
        }
        catch (err) {
            debug(`Failed to load file ${filePath}: ${err}`);
            throw err;
        }
    }
}
export class JitiFileLoader extends FileLoader {
    static DEFAULT_PRIORITY = 500 + preferenceDelta('jiti');
    static state = new UninitializedJitiFileLoaderState();
    constructor(options) {
        super({
            match: FileMatch.standardFiles().complement(),
            priority: JitiFileLoader.DEFAULT_PRIORITY,
            ...options,
        });
    }
    async loadFile(filePath) {
        return await JitiFileLoader.state.loadFile(filePath);
    }
}
