// checkly/dist/loader/loader.js
import { FileMatch } from './match.js';
export class FileLoader {
    static DEFAULT_PRIORITY = 1000;
    fileMatcher;
    #priority;
    constructor(options) {
        this.fileMatcher = options?.match ?? FileMatch.any();
        this.#priority = options?.priority ?? FileLoader.DEFAULT_PRIORITY;
    }
    /**
     * The priority of the loader, used to resolve load order if multiple
     * loaders are available.
     */
    get priority() {
        return this.#priority;
    }
    /**
     * Checks whether the FileLoader can be used for a file path.
     *
     * @param filePath The file path to evaluate.
     * @returns Whether the FileLoader is authoritative for the file path.
     */
    isAuthoritativeFor(filePath) {
        return this.fileMatcher.match(filePath);
    }
}
/**
 * Error thrown when a FileLoader is authoritative for a file path but
 * fails to load it.
 */
export class UnsupportedFileLoaderError extends Error {
    constructor(message = 'File cannot be loaded by this loader', options) {
        super(message, options);
        this.name = 'UnsupportedFileLoaderError';
    }
}
