// checkly/dist/loader/match.js
import { extname } from 'node:path';
export class FileMatch {
    match;
    constructor(match) {
        this.match = match;
    }
    complement() {
        return new FileMatch(filePath => !this.match(filePath));
    }
    union(matcher) {
        return new FileMatch(filePath => {
            return this.match(filePath) || matcher.match(filePath);
        });
    }
    intersection(matcher) {
        return new FileMatch(filePath => {
            return this.match(filePath) && matcher.match(filePath);
        });
    }
    difference(matcher) {
        return this.intersection(matcher.complement());
    }
    static none() {
        return new FileMatch(() => false);
    }
    static any() {
        return new FileMatch(() => true);
    }
    static pattern(pattern) {
        return new FileMatch(filePath => pattern.test(filePath));
    }
    static extension(...extensions) {
        const set = new Set(extensions);
        return new FileMatch(filePath => set.has(extname(filePath)));
    }
    static custom(match) {
        return new FileMatch(match);
    }
    static standardFiles() {
        return FileMatch.extension('.js', '.mjs', '.cjs', '.json');
    }
}
