// checkly/dist/constructs/internal/codegen/snippet.js
import { Parser } from '../../../services/check-parser/parser.js';
export class UnsupportedScriptError extends Error {
}
export function isSafeSnippetFilename(name) {
    // These are the characters the backend supports.
    return /^[\w-]+$/.test(name);
}
export function isSnippet(resource) {
    if (typeof resource.id !== 'number') {
        return false;
    }
    if (typeof resource.name !== 'string') {
        return false;
    }
    if (typeof resource.script !== 'string') {
        return false;
    }
    validateScript(resource.script);
    return true;
}
export function snippetUsesHandlebarsSyntax(script) {
    return script.includes('{{>') || script.includes('{{#');
}
export function validateScript(content) {
    if (snippetUsesHandlebarsSyntax(content)) {
        throw new UnsupportedScriptError(`Conversion from legacy handlebars syntax is not supported.`);
    }
}
const SNIPPET_PATH_PREFIX = './snippets/';
export function parseSnippetDependencies(content) {
    const { module: { dependencies }, error, } = Parser.parseDependencies('__placeholder.ts', content);
    if (error) {
        throw new Error(`Failed to parse '${SNIPPET_PATH_PREFIX}...' dependencies: ${error}`, { cause: error });
    }
    return dependencies
        .filter(({ importPath }) => importPath.startsWith(SNIPPET_PATH_PREFIX))
        .map(({ importPath }) => importPath.slice(SNIPPET_PATH_PREFIX.length));
}
