// checkly/dist/sourcegen/arraybuilder.js
import { object } from './objectbuilder.js';
import { ArrayValue } from './array.js';
import { BooleanValue } from './boolean.js';
import { NullValue } from './null.js';
import { NumberValue } from './number.js';
import { StringValue } from './string.js';
import { expr } from './exprbuilder.js';
import { UndefinedValue } from './undefined.js';
import { IdentifierValue } from './identifier.js';
export function array(build) {
    const builder = new ArrayValueBuilder();
    build(builder);
    return builder.build();
}
export class ArrayValueBuilder {
    #elements = [];
    empty() {
        this.#elements = [];
        return this;
    }
    string(value) {
        try {
            return this.value(new StringValue(value));
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (string): ${cause}`, { cause });
        }
    }
    boolean(value) {
        try {
            return this.value(new BooleanValue(value));
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (boolean): ${cause}`, { cause });
        }
    }
    number(value) {
        try {
            return this.value(new NumberValue(value));
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (number): ${cause}`, { cause });
        }
    }
    null() {
        try {
            return this.value(new NullValue());
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (null): ${cause}`, { cause });
        }
    }
    undefined() {
        try {
            return this.value(new UndefinedValue());
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (undefined): ${cause}`, { cause });
        }
    }
    ident(value) {
        try {
            return this.value(new IdentifierValue(value));
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (identifier): ${cause}`, { cause });
        }
    }
    array(build) {
        try {
            return this.value(array(build));
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (array): ${cause}`, { cause });
        }
    }
    object(build) {
        try {
            return this.value(object(build));
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (object): ${cause}`, { cause });
        }
    }
    expr(context, build) {
        try {
            return this.value(expr(context, build));
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (expr): ${cause}`, { cause });
        }
    }
    value(value) {
        this.#elements.push(value);
        return this;
    }
    get elements() {
        return this.#elements;
    }
    build() {
        return new ArrayValue([...this.#elements]);
    }
}
