// checkly/dist/sourcegen/declbuilder.js
import { IdentifierDeclaration, VariableDeclaration, ExportDeclaration, LeadingCommentDeclaration, TrailingCommentDeclaration } from './decl.js';
export function decl(identifier, build) {
    const builder = new DeclarationBuilder(identifier);
    build(builder);
    return builder.build();
}
export class DeclarationBuilder {
    #identifier;
    #decl;
    constructor(identifier) {
        this.#identifier = identifier;
        this.#decl = new IdentifierDeclaration(identifier);
    }
    variable(value, options) {
        this.#decl = new VariableDeclaration(this.#identifier, value, options);
        return this;
    }
    export(options) {
        this.#decl = new ExportDeclaration(this.#decl, options);
        return this;
    }
    leadingComment(comment) {
        this.#decl = new LeadingCommentDeclaration(comment, this.#decl);
        return this;
    }
    trailingComment(comment) {
        this.#decl = new TrailingCommentDeclaration(comment, this.#decl);
        return this;
    }
    build() {
        return this.#decl;
    }
}
