// checkly/dist/sourcegen/expr.js
import { Value } from './value.js';
export class ExpressionValue extends Value {
}
export class NewExpressionValue extends ExpressionValue {
    callee;
    args;
    constructor(callee, args) {
        super();
        this.callee = callee;
        this.args = args;
    }
    render(output) {
        output.append('new');
        output.significantWhitespace();
        this.callee.render(output);
        this.args.render(output);
    }
}
export class CallExpressionValue extends ExpressionValue {
    callee;
    args;
    constructor(callee, args) {
        super();
        this.callee = callee;
        this.args = args;
    }
    render(output) {
        this.callee.render(output);
        this.args.render(output);
    }
}
export class MemberExpressionValue extends ExpressionValue {
    object;
    property;
    constructor(object, property) {
        super();
        this.object = object;
        this.property = property;
    }
    render(output) {
        this.object.render(output);
        output.append('.');
        this.property.render(output); // @TODO: . vs ['']
    }
}
export class BinaryExpressionValue extends ExpressionValue {
    left;
    right;
    op;
    constructor(left, right, op) {
        super();
        this.left = left;
        this.right = right;
        this.op = op;
    }
    render(output) {
        this.left.render(output);
        output.cosmeticWhitespace();
        output.append(this.op);
        output.cosmeticWhitespace();
        this.right.render(output);
    }
}
export class UnaryExpressionValue extends ExpressionValue {
    value;
    op;
    constructor(value, op) {
        super();
        this.value = value;
        this.op = op;
    }
    render(output) {
        output.append(this.op);
        this.value.render(output);
    }
}
