// checkly/dist/sourcegen/exprbuilder.js
import { ArgumentsValueBuilder } from './argsbuilder.js';
import { CallExpressionValue, MemberExpressionValue, NewExpressionValue } from './expr.js';
export function expr(context, build) {
    const builder = new ExpressionValueBuilder(context);
    build(builder);
    return builder.build();
}
export class ExpressionValueBuilder {
    #context;
    constructor(value) {
        this.#context = value;
    }
    new(build) {
        try {
            const builder = new ArgumentsValueBuilder();
            build(builder);
            return this.context(new NewExpressionValue(this.#context, builder.build()));
        }
        catch (cause) {
            throw new Error(`Failed to create 'new' expression: ${cause}`, { cause });
        }
    }
    call(build) {
        try {
            const builder = new ArgumentsValueBuilder();
            build(builder);
            return this.context(new CallExpressionValue(this.#context, builder.build()));
        }
        catch (cause) {
            throw new Error(`Failed to create 'call' expression: ${cause}`, { cause });
        }
    }
    member(property) {
        try {
            return this.context(new MemberExpressionValue(this.#context, property));
        }
        catch (cause) {
            throw new Error(`Failed to create 'member' expression: ${cause}`, { cause });
        }
    }
    context(value) {
        this.#context = value;
        return this;
    }
    build() {
        return this.#context;
    }
}
