// checkly/dist/sourcegen/objectbuilder.js
import { array } from './arraybuilder.js';
import { BooleanValue } from './boolean.js';
import { expr } from './exprbuilder.js';
import { IdentifierValue } from './identifier.js';
import { NullValue } from './null.js';
import { NumberValue } from './number.js';
import { ObjectProperty, ObjectValue } from './object.js';
import { StringValue } from './string.js';
import { UndefinedValue } from './undefined.js';
export function object(build, options) {
    const builder = new ObjectValueBuilder(options);
    build(builder);
    return builder.build();
}
export class ObjectValueBuilder {
    #properties = new Map();
    #valueOptions;
    #implicitOrder = true;
    #implicitOrderValue = 0;
    #implicitOrderStep = 100;
    constructor(options) {
        this.#valueOptions = {
            ...options,
        };
        if (options?.implicitOrder !== undefined) {
            this.#implicitOrder = options.implicitOrder;
        }
        if (options?.implicitOrderStart !== undefined) {
            this.#implicitOrderValue = options.implicitOrderStart;
        }
        if (options?.implicitOrderStep !== undefined) {
            this.#implicitOrderStep = options.implicitOrderStep;
        }
    }
    string(name, value, options) {
        try {
            return this.value(name, new StringValue(value), options);
        }
        catch (cause) {
            throw new Error(`Failed to set object property '${name}' (string): ${cause}`, { cause });
        }
    }
    boolean(name, value, options) {
        try {
            return this.value(name, new BooleanValue(value), options);
        }
        catch (cause) {
            throw new Error(`Failed to set object property '${name}' (boolean): ${cause}`, { cause });
        }
    }
    number(name, value, options) {
        try {
            return this.value(name, new NumberValue(value), options);
        }
        catch (cause) {
            throw new Error(`Failed to set object property '${name}' (number): ${cause}`, { cause });
        }
    }
    null(name, options) {
        try {
            return this.value(name, new NullValue(), options);
        }
        catch (cause) {
            throw new Error(`Failed to set object property '${name}' (null): ${cause}`, { cause });
        }
    }
    undefined(name, options) {
        try {
            return this.value(name, new UndefinedValue(), options);
        }
        catch (cause) {
            throw new Error(`Failed to set object property '${name}' (undefined): ${cause}`, { cause });
        }
    }
    ident(name, value, options) {
        try {
            return this.value(name, new IdentifierValue(value), options);
        }
        catch (cause) {
            throw new Error(`Failed to set object property '${name}' (identifier): ${cause}`, { cause });
        }
    }
    array(name, build, options) {
        try {
            return this.value(name, array(build), options);
        }
        catch (cause) {
            throw new Error(`Failed to set object property '${name}' (array): ${cause}`, { cause });
        }
    }
    object(name, build, options) {
        try {
            return this.value(name, object(build), options);
        }
        catch (cause) {
            throw new Error(`Failed to set object property '${name}' (object): ${cause}`, { cause });
        }
    }
    expr(name, context, build, options) {
        try {
            return this.value(name, expr(context, build), options);
        }
        catch (cause) {
            throw new Error(`Failed to set object property '${name}' (expr): ${cause}`, { cause });
        }
    }
    value(name, value, options) {
        if (options?.order === undefined && this.#implicitOrder) {
            options = {
                ...options,
                order: this.#implicitOrderValue,
            };
            this.#implicitOrderValue += this.#implicitOrderStep;
        }
        this.#properties.set(name, new ObjectProperty(name, value, options));
        return this;
    }
    build() {
        return new ObjectValue([...this.#properties.values()], this.#valueOptions);
    }
}
