// checkly/dist/rest/assets.js
import axios from 'axios';
// eslint-disable-next-line no-restricted-syntax
var AssetType;
(function (AssetType) {
    AssetType["LOG"] = "log";
    AssetType["CHECK_RUN_DATA"] = "check-run-data";
    AssetType["SCREENSHOT"] = "screenshot";
})(AssetType || (AssetType = {}));
export function isCheckResultAssetsV1(assets) {
    return typeof assets === 'object'
        && assets !== null
        && 'version' in assets === false;
}
export function isCheckResultAssetsV2(assets) {
    return typeof assets === 'object'
        && assets !== null
        && 'version' in assets
        && assets.version === 2;
}
export function isCheckResultAssetsV3(assets) {
    return typeof assets === 'object'
        && assets !== null
        && 'version' in assets
        && assets.version === 3;
}
export default class Assets {
    api;
    constructor(api) {
        this.api = api;
    }
    getLogs(asset) {
        return this.getAssets(AssetType.LOG, asset);
    }
    getCheckRunData(asset) {
        return this.getAssets(AssetType.CHECK_RUN_DATA, asset);
    }
    async getAssets(assetType, asset) {
        const region = asset.region;
        let key;
        switch (assetType) {
            case AssetType.LOG: {
                if (isCheckResultAssetsV1(asset) && asset.logPath) {
                    key = asset.logPath;
                }
                else if (isCheckResultAssetsV2(asset)) {
                    return this.getAssetsV2(assetType, asset);
                }
                else if (isCheckResultAssetsV3(asset) && asset.logPath) {
                    key = asset.clickHouseId;
                }
                break;
            }
            case AssetType.CHECK_RUN_DATA: {
                if (isCheckResultAssetsV1(asset) && asset.checkRunDataPath) {
                    key = asset.checkRunDataPath;
                }
                else if (isCheckResultAssetsV2(asset)) {
                    return this.getAssetsV2(assetType, asset);
                }
                else if (isCheckResultAssetsV3(asset)) {
                    key = asset.clickHouseId;
                    break;
                }
            }
        }
        if (!key) {
            return;
        }
        const response = await this.api.get(`/next/assets/${assetType}/${region}/${encodeURIComponent(key)}`);
        return response.data;
    }
    async getAssetsV2(assetType, asset) {
        const entryName = assetType === AssetType.LOG ? 'logs.txt' : 'check-run-data.json';
        const entry = asset.entries.find(e => e.name === entryName);
        if (!entry) {
            return;
        }
        const presignedUrl = await this.getAssetsLink(asset.region, asset.key);
        const response = await axios.get(presignedUrl, {
            headers: { Range: `bytes=${entry.start}-${entry.end}` },
            responseType: 'text',
        });
        return JSON.parse(response.data);
    }
    async getAssetsLink(region, key) {
        const response = await this.api.get(`/next/assets/${AssetType.CHECK_RUN_DATA}/${region}/${encodeURIComponent(key)}/redirect?link=true`);
        return response.data;
    }
}
