// checkly/dist/rest/checks.js
import { parseTotalFromContentRange } from '../helpers/content-range.js';
class Checks {
    api;
    constructor(api) {
        this.api = api;
    }
    getAll(params = {}) {
        return this.api.get('/v1/checks', { params });
    }
    async getAllPaginated(params = {}) {
        const response = await this.api.get('/v1/checks', { params });
        const total = parseTotalFromContentRange(response.headers['content-range']) ?? response.data.length;
        return {
            checks: response.data,
            total,
            page: params.page ?? 1,
            limit: params.limit ?? 10,
        };
    }
    async fetchAll(params = {}) {
        const pageSize = 100;
        const first = await this.getAllPaginated({ ...params, page: 1, limit: pageSize });
        const allChecks = [...first.checks];
        if (first.total <= pageSize)
            return allChecks;
        const totalPages = Math.ceil(first.total / pageSize);
        const remaining = await Promise.all(Array.from({ length: totalPages - 1 }, (_, i) => this.api.get('/v1/checks', {
            params: { ...params, page: i + 2, limit: pageSize },
        }).then(r => r.data)));
        for (const page of remaining) {
            allChecks.push(...page);
        }
        return allChecks;
    }
    get(id) {
        return this.api.get(`/v1/checks/${id}`);
    }
    delete(id) {
        return this.api.delete(`/v1/checks/${id}`);
    }
}
export default Checks;
