// checkly/dist/rest/environment-variables.js
import { ConflictError, NotFoundError } from './errors.js';
/**
 * Error thrown when the requested environment variable does not exist.
 */
export class EnvironmentVariableNotFoundError extends Error {
    environmentVariable;
    constructor(environmentVariable, options) {
        super(`Environment variable "${environmentVariable}" does not exist.`, options);
        this.name = 'EnvironmentVariableNotFoundError';
        this.environmentVariable = environmentVariable;
    }
}
/**
 * Error thrown when the requested environment variable already exists.
 */
export class EnvironmentVariableAlreadyExistsError extends Error {
    environmentVariable;
    constructor(environmentVariable, options) {
        super(`Environment variable "${environmentVariable}" already exists.`, options);
        this.name = 'EnvironmentVariableAlreadyExistsError';
        this.environmentVariable = environmentVariable;
    }
}
class EnvironmentVariables {
    api;
    constructor(api) {
        this.api = api;
    }
    getAll() {
        return this.api.get('/v1/variables');
    }
    /**
     * @throws {EnvironmentVariableNotFoundError} If the environment variable does not exist.
     */
    async delete(environmentVariableKey) {
        try {
            return await this.api.delete(`/v1/variables/${environmentVariableKey}`);
        }
        catch (err) {
            if (err instanceof NotFoundError) {
                throw new EnvironmentVariableNotFoundError(environmentVariableKey);
            }
            throw err;
        }
    }
    /**
     * @throws {EnvironmentVariableAlreadyExistsError} If the environment variable already exists.
     */
    async add(environmentVariableKey, environmentVariableValue, locked = false, secret = false) {
        try {
            return await this.api.post('/v1/variables', { key: environmentVariableKey, value: environmentVariableValue, locked, secret });
        }
        catch (err) {
            if (err instanceof ConflictError) {
                throw new EnvironmentVariableAlreadyExistsError(environmentVariableKey);
            }
            throw err;
        }
    }
    /**
     * @throws {EnvironmentVariableNotFoundError} If the environment variable does not exist.
     */
    async get(environmentVariableKey) {
        try {
            return await this.api.get(`/v1/variables/${environmentVariableKey}`);
        }
        catch (err) {
            if (err instanceof NotFoundError) {
                throw new EnvironmentVariableNotFoundError(environmentVariableKey);
            }
            throw err;
        }
    }
    /**
     * @throws {EnvironmentVariableNotFoundError} If the environment variable does not exist.
    */
    // update environment variable with default locked value false
    async update(environmentVariableKey, environmentVariableValue, locked = false, secret = false) {
        try {
            return await this.api.put(`/v1/variables/${environmentVariableKey}`, { value: environmentVariableValue, locked, secret });
        }
        catch (err) {
            if (err instanceof NotFoundError) {
                throw new EnvironmentVariableNotFoundError(environmentVariableKey);
            }
            throw err;
        }
    }
}
export default EnvironmentVariables;
