// checkly/dist/rest/test-sessions.js
import { compressJSONPayload } from './util.js';
import { ForbiddenError, RequestTimeoutError } from './errors.js';
const COMPLETION_MAX_WAIT_SECONDS = 30;
export class NoMatchingChecksError extends Error {
    constructor(options) {
        super(`No matching checks found.`, options);
        this.name = 'NoMatchingChecksError';
    }
}
class TestSessions {
    api;
    constructor(api) {
        this.api = api;
    }
    async run(payload) {
        return await this.api.post('/next/test-sessions/run', payload, {
            transformRequest: compressJSONPayload,
        });
    }
    /**
     * @throws {NoMatchingChecksError} If no checks matched the request.
     */
    async trigger(payload) {
        try {
            const resp = await this.api.post('/next/test-sessions/trigger', payload);
            if (resp.data.checks.length === 0) {
                // Currently the BE will never return an empty `checks` array, it returns a 403 instead.
                // This is added to make the old CLI versions compatible if we ever change this, though.
                throw new NoMatchingChecksError();
            }
            return resp;
        }
        catch (err) {
            if (err instanceof ForbiddenError && err.data.errorCode === 'ERR_NO_MATCHING_CHECKS') {
                throw new NoMatchingChecksError();
            }
            throw err;
        }
    }
    get(id) {
        return this.api.get(`/v1/test-sessions/${id}`);
    }
    getResult(id, resultId) {
        return this.api.get(`/v1/test-sessions/${id}/results/${resultId}`);
    }
    getCompletion(id, maxWaitSeconds = COMPLETION_MAX_WAIT_SECONDS) {
        return this.api.get(`/v1/test-sessions/${id}/completion`, {
            params: { maxWaitSeconds },
        });
    }
    async pollUntilComplete(id) {
        while (true) {
            try {
                const { data } = await this.getCompletion(id);
                return data;
            }
            catch (err) {
                if (err instanceof RequestTimeoutError) {
                    continue;
                }
                throw err;
            }
        }
    }
    list(params = {}) {
        return this.api.get('/v1/test-sessions', { params });
    }
    getShortLink(id) {
        return this.api.get(`/next/test-sessions/${id}/link`);
    }
    getResultShortLinks(testSessionId, testResultId) {
        return this.api.get(`/next/test-sessions/${testSessionId}/results/${testResultId}/links`);
    }
}
export default TestSessions;
